From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Thu, 19 Jun 2025 00:49:24 +0800
Subject: [PATCH] Vanilla Fluid Pushing


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 529be219452fe64398f91af77ea47c5cbfe8a1df..3c9b659aa473dcc65da4aaaecb4a8132a8206b81 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4622,8 +4622,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
                                 final Vec3 flow = fluidState.getFlow(world, mutablePos);
 
-                                if (diff < 0.4) {
-                                    pushVector = pushVector.add(flow.scale(diff));
+                                // Leaves start - Vanilla Fluid Pushing
+                                if (maxHeightDiff < 0.4) {
+                                    pushVector = pushVector.add(flow.scale(maxHeightDiff));
+                                // Leaves end - Vanilla Fluid Pushing
                                 } else {
                                     pushVector = pushVector.add(flow);
                                 }
