From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:42:17 +0800
Subject: [PATCH] Old zombie reinforcement


diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index 39b65970a48568c95ff482b9636e7391f300ffa8..50417ae5883374113cf820c064570b6aeda4fbb8 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -342,7 +342,7 @@ public class Zombie extends Monster {
                 int floor = Mth.floor(this.getX());
                 int floor1 = Mth.floor(this.getY());
                 int floor2 = Mth.floor(this.getZ());
-                EntityType<? extends Zombie> type = this.getType();
+                EntityType<? extends Zombie> type = org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldZombieReinforcement ? EntityType.ZOMBIE : this.getType(); // Leaves end - old zombie reinforcement
                 Zombie zombie = type.create(level, EntitySpawnReason.REINFORCEMENT);
                 if (zombie == null) {
                     return true;
