From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Thu, 19 Jun 2025 00:49:24 +0800
Subject: [PATCH] Vanilla Fluid Pushing


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 0ec5603097a984713bc3ae7a03a7535e552cf851..57baab29543c745007543f77b5a2bd27f006a904 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4618,8 +4618,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
                                 final Vec3 flow = fluidState.getFlow(world, mutablePos);
 
-                                if (diff < 0.4) {
-                                    pushVector = pushVector.add(flow.scale(diff));
+                                // Leaves start - Vanilla Fluid Pushing
+                                if (maxHeightDiff < 0.4) {
+                                    pushVector = pushVector.add(flow.scale(maxHeightDiff));
+                                // Leaves end - Vanilla Fluid Pushing
                                 } else {
                                     pushVector = pushVector.add(flow);
                                 }
