--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -93,19 +_,33 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
 }
 
-val generatedApiPath: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedApiPath: java.nio.file.Path = rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath() // Leaves - build change
 idea {
     module {
         generatedSourceDirs.add(generatedApiPath.toFile())
     }
 }
+// Leaves start - build change
 sourceSets {
     main {
         java {
             srcDir(generatedApiPath)
+            srcDir(file("../paper-api/src/main/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/main/resources"))
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
         }
     }
 }
+// Leaves end - build change
 
 val outgoingVariants = arrayOf("runtimeElements", "apiElements", "sourcesElements", "javadocElements")
 val mainCapability = "${project.group}:${project.name}:${project.version}"
@@ -150,6 +_,15 @@
     }
 }
 
+// Leaves start - hide irrelevant compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Leaves end - hide irrelevant compilation warnings
+
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
         into("META-INF/maven/${project.group}/${project.name}")
@@ -169,7 +_,7 @@
 
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
-    options.overview = "src/main/javadoc/overview.html"
+    options.overview = "../paper-api/src/main/javadoc/overview.html" // Leaves - build change
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -202,16 +_,17 @@
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../paper-api/src/main/javadoc").withPropertyName("javadoc-sourceset") // Leaves - build change
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../paper-api/src/main/javadoc") { // Leaves - build change
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
         }
     }
+    options.addStringOption("Xdoclint:none", "-quiet") // Leaves - hide irrelevant compilation warnings
 }
 
 tasks.test {
