From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 7 Feb 2025 14:23:43 +0800
Subject: [PATCH] Xaero Map Protocol


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 7f21903070ab733249441e796a1faa5adf873ffb..e1ce47f1f507d7d1ec8606b02815b0d8fb970734 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1202,6 +1202,7 @@ public abstract class PlayerList {
         player.connection.send(new ClientboundSetChunkCacheRadiusPacket(level.spigotConfig.viewDistance));
         player.connection.send(new ClientboundSetSimulationDistancePacket(level.spigotConfig.simulationDistance));
         // Paper end
+        org.leavesmc.leaves.protocol.XaeroMapProtocol.onSendWorldInfo(player); // Leaves - xaero map protocol
         if (level.isRaining()) {
             // CraftBukkit start - handle player weather
             // player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
