From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 16:42:23 +0800
Subject: [PATCH] Fix trapdoor feature


diff --git a/net/minecraft/world/level/block/TrapDoorBlock.java b/net/minecraft/world/level/block/TrapDoorBlock.java
index 46c23990e78ce77b188052672778bdef560a0d13..c6aad768cad0f0eccbe5405f41c4fa34583970ba 100644
--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -160,6 +160,8 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                     hasNeighborSignal = eventRedstone.getNewCurrent() > 0;
                 }
                 // CraftBukkit end
+
+                /* Leaves - it is feature, not bug!
                 // Paper start - break redstone on trapdoors early
                 boolean open = state.getValue(TrapDoorBlock.OPEN) != hasNeighborSignal;
                 // note: this must run before any state for this block/its neighborus are written to the world
@@ -178,8 +180,9 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                         }
                     }
                 }
-                if (open) {
-                // Paper end - break redstone on trapdoors early
+                */
+
+                if (state.getValue(TrapDoorBlock.OPEN) != hasNeighborSignal) {
                     state = state.setValue(OPEN, Boolean.valueOf(hasNeighborSignal));
                     this.playSound(null, level, pos, hasNeighborSignal);
                 }
