From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 14 Jun 2023 12:07:07 +0800
Subject: [PATCH] Redstone wire dont connect if on trapdoor


diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 12c9d60314c99fb65e640d255a2d0c6b7790ad4d..ff29671f78a4e896e70daf89f440f3b9425b0760 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -204,7 +204,7 @@ public class RedStoneWireBlock extends Block {
         RandomSource random
     ) {
         if (direction == Direction.DOWN) {
-            return !this.canSurviveOn(level, neighborPos, neighborState) ? Blocks.AIR.defaultBlockState() : state;
+            return org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneDontCantOnTrapDoor ? state : !this.canSurviveOn(level, neighborPos, neighborState) ? Blocks.AIR.defaultBlockState() : state; // Leaves - behavior to 1.19
         } else if (direction == Direction.UP) {
             return this.getConnectionState(level, state, pos);
         } else {
@@ -263,7 +263,7 @@ public class RedStoneWireBlock extends Block {
         BlockPos blockPos = pos.relative(direction);
         BlockState blockState = level.getBlockState(blockPos);
         if (nonNormalCubeAbove) {
-            boolean flag = blockState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, blockPos, blockState);
+            boolean flag = (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneDontCantOnTrapDoor && blockState.getBlock() instanceof TrapDoorBlock) || this.canSurviveOn(level, blockPos, blockState); // Leaves - behavior to 1.19
             if (flag && shouldConnectTo(level.getBlockState(blockPos.above()))) {
                 if (blockState.isFaceSturdy(level, blockPos, direction.getOpposite())) {
                     return RedstoneSide.UP;
