From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 28 Sep 2023 17:07:02 +0800
Subject: [PATCH] CCE update suppression


diff --git a/net/minecraft/world/level/block/ShulkerBoxBlock.java b/net/minecraft/world/level/block/ShulkerBoxBlock.java
index 7794e66211af34ecca8524d3ca7e0d928f0ab0eb..fca50f68625050daabcae3a3b615cf88cce41111 100644
--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -236,7 +236,9 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
     protected int getAnalogOutputSignal(BlockState blockState, Level level, BlockPos pos) {
         // Leaves start - fix update suppression crash
         try {
-            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
+            return org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.cceUpdateSuppression ?
+                    AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) level.getBlockEntity(pos)) : // Leaves - make cce happy(?)
+                    AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
         } catch (ClassCastException ex) {
             if (org.leavesmc.leaves.LeavesConfig.modify.updateSuppressionCrashFix) {
                 throw new org.leavesmc.leaves.util.UpdateSuppressionException(pos, this);
@@ -244,7 +246,6 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
                 throw ex;
             }
         }
-        // Leaves end - fix update suppression crash
     }
 
     public static Block getBlockByColor(@Nullable DyeColor color) {
