From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 30 Oct 2023 10:43:44 +0800
Subject: [PATCH] Make Item tick vanilla


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 4f5db8203189a413f410fe7648e4995bae171918..e3cdb823b6d76e62725b6428651dacfe20fc4154 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -133,6 +133,9 @@ public class ItemEntity extends Entity implements TraceableEntity {
     // Paper start - EAR 2
     @Override
     public void inactiveTick() {
+        // Leaves start - vanilla
+        this.tick();
+        /*
         super.inactiveTick();
         if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
             this.pickupDelay--;
@@ -150,6 +153,8 @@ public class ItemEntity extends Entity implements TraceableEntity {
             // CraftBukkit end
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
+         */
+        // Leaves end - vanilla
     }
     // Paper end - EAR 2
 
