From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 5 Feb 2025 23:11:32 +0800
Subject: [PATCH] Servux Protocol


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 439bc4a93deb6fe134a9b1de4ca73a8c785d46db..dd7df1616d215020a65ce5d048f1b6f01469c86f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2192,6 +2192,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         }
 
         this.lastSpawnChunkRadius = i;
+        org.leavesmc.leaves.protocol.servux.ServuxHudDataProtocol.refreshSpawnMetadata = true; // Leaves - servux
     }
 
     public LongSet getForcedChunks() {
