From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bluemangoo <chenfy2006@qq.com>
Date: Fri, 19 Jul 2024 15:04:03 +0800
Subject: [PATCH] Disable crystal-portal proximity check


diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index 1bc4c8be71b445f64134548a85fd81442298c0f1..a1d8b38b20efd7ef0fa6a8d92d218fada66d446e 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -27,7 +27,7 @@ public class EndCrystalItem extends Item {
         if (!blockState.is(Blocks.OBSIDIAN) && !blockState.is(Blocks.BEDROCK)) {
             return InteractionResult.FAIL;
         } else {
-            BlockPos blockPos = clickedPos.above(); final BlockPos aboveBlockPosition = blockPos; // Paper - OBFHELPER
+            BlockPos blockPos = clickedPos.above(); // final BlockPos aboveBlockPosition = blockPos; // Paper - OBFHELPER // Leaves
             if (!level.isEmptyBlock(blockPos)) {
                 return InteractionResult.FAIL;
             } else {
@@ -51,7 +51,7 @@ public class EndCrystalItem extends Item {
                         level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, blockPos);
                         EndDragonFight dragonFight = ((ServerLevel)level).getDragonFight();
                         if (dragonFight != null) {
-                            dragonFight.tryRespawn(aboveBlockPosition); // Paper - Perf: Do crystal-portal proximity check before entity lookup
+                            dragonFight.tryRespawn(); // dragonFight.tryRespawn(aboveBlockPosition); // Paper - Perf: Do crystal-portal proximity check before entity lookup // Leaves
                         }
                     }
 
diff --git a/net/minecraft/world/level/dimension/end/EndDragonFight.java b/net/minecraft/world/level/dimension/end/EndDragonFight.java
index 5a28a821bb401f8f1465b085c9ffda52ba9a0a9f..b03c5fde5cc62a26f8bb7632b1a684755a6e5649 100644
--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -575,12 +575,14 @@ public class EndDragonFight {
     }
 
     public boolean tryRespawn() { // CraftBukkit - return boolean
+        /* Leaves
         // Paper start - Perf: Do crystal-portal proximity check before entity lookup
         return this.tryRespawn(null);
     }
 
     public boolean tryRespawn(@Nullable BlockPos placedEndCrystalPos) { // placedEndCrystalPos is null if the tryRespawn() call was not caused by a placed end crystal
         // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+         */
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos blockPos = this.portalLocation;
             if (blockPos == null) {
@@ -595,6 +597,7 @@ public class EndDragonFight {
 
                 blockPos = this.portalLocation;
             }
+            /* Leaves
             // Paper start - Perf: Do crystal-portal proximity check before entity lookup
             if (placedEndCrystalPos != null) {
                 // The end crystal must be 0 or 1 higher than the portal origin
@@ -610,7 +613,7 @@ public class EndDragonFight {
                 }
             }
             // Paper end - Perf: Do crystal-portal proximity check before entity lookup
-
+             */
 
             List<EndCrystal> list = Lists.newArrayList();
             BlockPos blockPos1 = blockPos.above(1);
