From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Sun, 27 Oct 2024 15:57:37 +0800
Subject: [PATCH] Disable vault blacklist


diff --git a/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index 9bf95455f93a17c53bf457d90fd30a3595227569..f580b2431edf8768a60c57de84da64682d017d8f 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -61,7 +61,7 @@ public class VaultServerData {
     }
 
     boolean hasRewardedPlayer(Player player) {
-        return this.rewardedPlayers.contains(player.getUUID());
+        return !org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist && this.rewardedPlayers.contains(player.getUUID()); // Leaves - disable vault blacklist
     }
 
     @VisibleForTesting
@@ -70,6 +70,7 @@ public class VaultServerData {
         addToRewardedPlayers(player.getUUID());
     }
     public boolean addToRewardedPlayers(final java.util.UUID player) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist) return false; // Leaves - disable vault blacklist
         final boolean removed = this.rewardedPlayers.add(player);
     // Paper end - Vault API
         if (this.rewardedPlayers.size() > 128) {
diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 50d6ff126a35ce8613175b550dac50cd82c43f9d..13ba4407b639ca64224348dd898c30d1256b13fa 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -68,8 +68,8 @@ public class VaultSharedData {
         Set<UUID> set = config.playerDetector()
             .detect(level, config.entitySelector(), pos, deactivationRange, false)
             .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+            .filter(uuid -> org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist || !serverData.getRewardedPlayers().contains(uuid))
+            .collect(Collectors.toSet()); // Leaves - disable vault blacklist
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
