From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 30 Mar 2022 08:58:32 +0000
Subject: [PATCH] Add isShrink to EntityResurrectEvent


diff --git a/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java b/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
index 2a7426c157f6a06f2491d95d67df95a0fb809926..8fe7d559ad6cbb8f15584c202067250dd55a8ede 100644
--- a/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
@@ -16,20 +16,44 @@ import org.jetbrains.annotations.Nullable;
 public class EntityResurrectEvent extends EntityEvent implements Cancellable {
 
     private static final HandlerList handlers = new HandlerList();
-    //
+
     private boolean cancelled;
+    private boolean shrink;
 
     private final EquipmentSlot hand;
 
     @ApiStatus.Internal
-    public EntityResurrectEvent(@NotNull LivingEntity what, @Nullable EquipmentSlot hand) {
+    public EntityResurrectEvent(@NotNull LivingEntity what, @Nullable EquipmentSlot hand, boolean shrink) {
         super(what);
+        this.shrink = shrink;
         this.hand = hand;
     }
 
+    public EntityResurrectEvent(@NotNull LivingEntity what, @Nullable EquipmentSlot hand) {
+        this(what, hand, true);
+    }
+
     @Deprecated(since = "1.19.2", forRemoval = true)
     public EntityResurrectEvent(@NotNull LivingEntity what) {
-        this(what, null);
+        this(what, null, true);
+    }
+
+    /**
+     * Get is shrink item
+     *
+     * @return is shrink item
+     */
+    public boolean isShrink() {
+        return shrink;
+    }
+
+    /**
+     * Set is shrink item
+     *
+     * @param shrink is shrink item
+     */
+    public void setShrink(boolean shrink) {
+        this.shrink = shrink;
     }
 
     @NotNull
