From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 3 Aug 2023 14:21:55 +0800
Subject: [PATCH] Force peaceful mode switch


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 015d852d5a0c01042a2153a6916d408660356c59..41c3c561d3d23136fbfa19c23d5e82ab1f2f06f3 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4346,6 +4346,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     void setSendViewDistance(int viewDistance);
     // Paper end - view distance api
 
+    // Leaves start - unsupported settings
+    public void setPeacefulModeSwitchTick(int tick);
+
+    public int getPeacefulModeSwitchTick();
+    // Leaves end - unsupported settings
+
     /**
      * Gets all generated structures that intersect the chunk at the given
      * coordinates. <br>
