From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 17 May 2025 17:36:22 +0800
Subject: [PATCH] Replay Mod API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 3b0513fbbf46e7cbb1790264eebaed9862d48eb5..44ee8036e2f813e947c2c166e1876030cabf86ff 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3014,4 +3014,10 @@ public final class Bukkit {
         return server.getBotManager();
     }
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    public static @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Leaves end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index bc9bdaadf6b6bb7cc7a8cf2d4e9b8e7724f36445..712c76385c3a038eaa252743e228705a99a7ebbe 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2770,4 +2770,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.leavesmc.leaves.entity.bot.BotManager getBotManager();
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager();
+    // Leaves end - Photographer API
 }
