From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 26 Jan 2025 01:39:16 -0500
Subject: [PATCH] Replay Mod API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 699c71f6749352aaf8cb82a51a82f86c3a8d3abc..8018a4e88c9392fd68c3bd03f13e7443e6a3966d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3009,4 +3009,10 @@ public final class Bukkit {
         return server.getBotManager();
     }
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    public static @NotNull org.leavesmc.leaves.entity.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Leaves end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 68c45204d1d6b8c33ab47a2414158fc273873414..f1519b243b676017185308145fe972702073c222 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -67,6 +67,7 @@ import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.leavesmc.leaves.entity.BotManager;
+import org.leavesmc.leaves.entity.PhotographerManager;
 
 /**
  * Represents a server implementation.
@@ -2706,4 +2707,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull BotManager getBotManager();
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    @NotNull PhotographerManager getPhotographerManager();
+    // Leaves end - Photographer API
 }
