From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 26 Jan 2025 01:39:16 -0500
Subject: [PATCH] Replay Mod API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index fe30de1f0a7fc7112466b6eb2e5813f39259c3b6..23f3ab31bbc9d197f9a82f4ed5003e6de814fad7 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3018,4 +3018,10 @@ public final class Bukkit {
         return server.getBotManager();
     }
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    public static @NotNull org.leavesmc.leaves.entity.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Leaves end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index cbec1d17b11872dc6f07a5a07936917706947ed1..858fe35d55b834a090b12c2e26b0fca08405cf04 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -68,6 +68,7 @@ import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.leavesmc.leaves.entity.BotManager;
+import org.leavesmc.leaves.entity.PhotographerManager;
 
 /**
  * Represents a server implementation.
@@ -2726,4 +2727,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull BotManager getBotManager();
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    @NotNull PhotographerManager getPhotographerManager();
+    // Leaves end - Photographer API
 }
