From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 26 Jan 2025 01:39:16 -0500
Subject: [PATCH] Replay Mod API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index aa06af1e53e556a5d37aeeb7ac7103bdc6eff472..d45d8d4eb673526b3d22cf9106a7d15dbdd94500 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3008,4 +3008,10 @@ public final class Bukkit {
         return server.getBotManager();
     }
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    public static @NotNull org.leavesmc.leaves.entity.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Leaves end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 41474dd9ea63e00efde93cff3df2c2e0c956e7ba..2b9cf972d18b43531bdbeb04eb5a8e5cdf909bbd 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -67,6 +67,7 @@ import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.leavesmc.leaves.entity.BotManager;
+import org.leavesmc.leaves.entity.PhotographerManager;
 
 /**
  * Represents a server implementation.
@@ -2715,4 +2716,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull BotManager getBotManager();
     // Leaves end - Bot API
+
+    // Leaves start - Photographer API
+    @NotNull PhotographerManager getPhotographerManager();
+    // Leaves end - Photographer API
 }
