From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Wed, 22 May 2024 10:12:04 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 23f3ab31bbc9d197f9a82f4ed5003e6de814fad7..30488f80b98d1dba485e53601edb5cb0e23b7591 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3024,4 +3024,10 @@ public final class Bukkit {
         return server.getPhotographerManager();
     }
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    public static org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager() {
+        return server.getBytebufManager();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 858fe35d55b834a090b12c2e26b0fca08405cf04..e794223bb697362c20fcbeb6595b95b11dafb694 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2731,4 +2731,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Photographer API
     @NotNull PhotographerManager getPhotographerManager();
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager();
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ac341838b6a2d2600eec8ae26791346f65ac93ae..7febcef116ae4027843169143dd90d651055a608 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3845,6 +3845,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // Leaves start - Bytebuf API
+    void sendPacket(org.leavesmc.leaves.bytebuf.packet.Packet packet);
+
+    void sendPacket(org.leavesmc.leaves.bytebuf.Bytebuf buf, org.leavesmc.leaves.bytebuf.packet.PacketType type);
+    // Leaves end - Bytebuf API
+
     @Override
     Spigot spigot();
     // Spigot end
