From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Thu, 6 Feb 2025 12:49:03 +0800
Subject: [PATCH] Revert raid changes


diff --git a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
index 61ff80a91c392e9e3c8eb393356b40721d28b6ec..57a24231962883affbcc5b46c0b952470398acb8 100644
--- a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
@@ -225,9 +225,7 @@ public class EntityPotionEffectEvent extends EntityEvent implements Cancellable
         /**
          * When a player gets bad omen after killing a patrol captain.
          *
-         * @deprecated no longer used, player now gets an ominous bottle instead
          */
-        @Deprecated(since = "1.21", forRemoval = true)
         PATROL_CAPTAIN,
         /**
          * When a potion effect is modified through the plugin methods.
