From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Wed, 22 May 2024 10:12:04 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d45d8d4eb673526b3d22cf9106a7d15dbdd94500..28ac7ef197dd6a568c01f762a787aa1ca396017b 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3014,4 +3014,10 @@ public final class Bukkit {
         return server.getPhotographerManager();
     }
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    public static org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager() {
+        return server.getBytebufManager();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2b9cf972d18b43531bdbeb04eb5a8e5cdf909bbd..079825a5ed671eac5f4f529393882a32d4b557c5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2720,4 +2720,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Photographer API
     @NotNull PhotographerManager getPhotographerManager();
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager();
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 7b9324d121d39ef75dc4d709831472598c1c50a0..7349663d3dd6d4d45dc12f2dbfb5337bcd0efbf2 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3829,6 +3829,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // Leaves start - Bytebuf API
+    void sendPacket(org.leavesmc.leaves.bytebuf.packet.Packet packet);
+
+    void sendPacket(org.leavesmc.leaves.bytebuf.Bytebuf buf, org.leavesmc.leaves.bytebuf.packet.PacketType type);
+    // Leaves end - Bytebuf API
+
     @Override
     Spigot spigot();
     // Spigot end
