From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Thu, 6 Feb 2025 12:49:03 +0800
Subject: [PATCH] Revert raid changes


diff --git a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
index 22e14ba522510d659c191d72536cde895458d9ed..253f937dbd130fac064963b4af7572e1302fee8f 100644
--- a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
@@ -221,9 +221,7 @@ public class EntityPotionEffectEvent extends EntityEvent implements Cancellable
         /**
          * When a player gets bad omen after killing a patrol captain.
          *
-         * @deprecated no longer used, player now gets an ominous bottle instead
          */
-        @Deprecated(since = "1.21", forRemoval = true)
         PATROL_CAPTAIN,
         /**
          * When a potion effect is modified through the plugin methods.
