From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:59:11 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 29570b8bcf34806af2f8d3dfdd656319176a61e5..786546ba9503a6fc7762589059b998afc62d8e22 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3026,4 +3026,10 @@ public final class Bukkit {
         return server.getBytebufAllocator();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d7252c4c094152733adff9fbdb3588c86c1cb167..fc87df91320ce2e5b3bac1dea51f484b416054c4 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2774,4 +2774,6 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager(); // Leaves end - Photographer API
 
     org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator(); // Leaves - Bytebuf API
+
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig(); // Leaves - Config API
 }
