From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:59:11 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 0149af5bd4234f925e12d78251766be5a16e5060..3c019280bc751e0710f1d34478b56986bc6000fc 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3039,4 +3039,10 @@ public final class Bukkit {
         return server.getBytebufManager();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e63fb4e0c55929f2721e16f69e0c0a4b455477fa..37ff5eabaf0fd3ab72eb37a1f21904b5a1adebe0 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2740,4 +2740,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Bytebuf API
     org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager();
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig();
+    // Leaves End - Config API
 }
