From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:59:11 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c25729882f87a904e59680967f3d20b02df50eae..5cc913f9e903ea5ef51c2e2b8cc18fdad8838c1e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3026,4 +3026,10 @@ public final class Bukkit {
         return server.getBytebufManager();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 06988e9586868ab3d5cd941c0ce9a40eec64eecc..0a52d9cffa58c922079bf97deb32dd4f057da097 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2778,4 +2778,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Bytebuf API
     org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager();
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig();
+    // Leaves End - Config API
 }
