From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 15 Feb 2026 01:35:58 -0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e065fa76649cd37cfd5e54634fc841232d1d4558..583ae21839f447f5112ac8abb06a161ed85fe988 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3026,4 +3026,10 @@ public final class Bukkit {
         return server.getBytebufManager();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9a3d62c61ecb5ee9f7607fa919409fcc4404ab60..670c2f51ec1199ae03ce1b41119903a7665fda95 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2774,4 +2774,6 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager(); // Leaves end - Photographer API
 
     org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufManager(); // Leaves - Bytebuf API
+
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig(); // Leaves - Config API
 }
