From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Sat, 2 Aug 2025 09:46:11 +0800
Subject: [PATCH] Leaves Plugin

This patch is licensed under the MIT license.

diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 3f5962acf1b7df1b4147e9fd1a91fc2ab5f8d817..3a9c06e71d6a3e429e256e291fdf594c994b419b 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3004,6 +3004,12 @@ public final class Bukkit {
         server.restart();
     }
 
+    // Leaves start - Feature API
+    public static org.leavesmc.leaves.plugin.FeatureManager getFeatureManager() {
+        return server.getFeatureManager();
+    }
+    // Leaves end - Feature API
+
     // Leaves start - Bot API
     /**
      * Returns a bot manager.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index c73d0a80153b3a5206159574dba553975a7daea9..ca445ab0892a9ac4f8ef3228d8b28a9d0d12667e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2781,6 +2781,10 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
 
+    // Leaves start - Feature API
+    org.leavesmc.leaves.plugin.FeatureManager getFeatureManager();
+    // Leaves End - Feature API
+
     // Leaves start - Bot API
     /**
      * Returns a bot manager.
