From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 28 Jan 2022 18:34:29 +0800
Subject: [PATCH] Make snowball and egg can knockback player


diff --git a/net/minecraft/world/entity/projectile/Snowball.java b/net/minecraft/world/entity/projectile/Snowball.java
index c57bbdc13221d2ce349f3f1d894193f80ff1e24b..51e412e814523c4acccee9caf39af729e1e950d8 100644
--- a/net/minecraft/world/entity/projectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/Snowball.java
@@ -54,6 +54,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = result.getEntity();
         int i = entity instanceof Blaze ? 3 : 0;
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), i);
+        // Leaves start - make snowball can knockback player
+        if (org.leavesmc.leaves.LeavesConfig.modify.snowballAndEggCanKnockback && entity instanceof net.minecraft.server.level.ServerPlayer player) {
+            player.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            player.knockback(0.4000000059604645D, this.getX() - player.getX(), this.getZ() - player.getZ(), this, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.DAMAGE);
+        }
+        // Leaves end - make snowball can knockback player
     }
 
     @Override
diff --git a/net/minecraft/world/entity/projectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/ThrownEgg.java
index 76481c0e77fc3a2e4be8eeb9de8d1e6de5507c64..14aeb330627e6fa06834736ab747b0fe64666bec 100644
--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -53,6 +53,12 @@ public class ThrownEgg extends ThrowableItemProjectile {
     protected void onHitEntity(EntityHitResult result) {
         super.onHitEntity(result);
         result.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Leaves start - make egg can knockback player
+        if (org.leavesmc.leaves.LeavesConfig.modify.snowballAndEggCanKnockback && result.getEntity() instanceof net.minecraft.server.level.ServerPlayer player) {
+            player.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            player.knockback(0.4000000059604645D, this.getX() - player.getX(), this.getZ() - player.getZ(), this, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.DAMAGE);
+        }
+        // Leaves end - make egg can knockback player
     }
 
     @Override
