From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 27 Mar 2022 12:48:40 +0800
Subject: [PATCH] Redstone Shears Wrench


diff --git a/net/minecraft/world/item/BlockItem.java b/net/minecraft/world/item/BlockItem.java
index 6db566adf2d0df1d26221eda04aa01738df6d3d2..24ebfdf98e343a7892af621d032c35819156deec 100644
--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -56,6 +56,14 @@ public class BlockItem extends Item {
             if (blockPlaceContext == null) {
                 return InteractionResult.FAIL;
             } else {
+                // Leaves start - shears wrench
+                if (org.leavesmc.leaves.util.ShearsWrenchUtil.shouldSkipPlace(context)) {
+                    if (context.getPlayer() != null) {
+                        context.getPlayer().containerMenu.forceHeldSlot(blockPlaceContext.getHand());
+                    }
+                    return InteractionResult.FAIL;
+                }
+                // Leaves end - shears wrench
                 BlockState placementState = this.getPlacementState(blockPlaceContext);
                 // CraftBukkit start - special case for handling block placement with water lilies and snow buckets
                 org.bukkit.block.BlockState bukkitState = null;
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index 8cf3e51e12f9cf98836657e722edb23943f9e866..5fd5e9fd4e1a79dd1a9d62a5cf0c308805979420 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -80,7 +80,10 @@ public class ShearsItem extends Item {
 
             return InteractionResult.SUCCESS;
         } else {
-            return super.useOn(context);
+            // Leaves start - shears wrench
+            InteractionResult result = org.leavesmc.leaves.util.ShearsWrenchUtil.tryApplyRotate(context, blockState);
+            return result == null ? super.useOn(context) : result;
+            // Leaves end - shears wrench
         }
     }
 }
