From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 30 Mar 2022 08:58:45 +0000
Subject: [PATCH] Add isShrink to EntityResurrectEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 8968ca0dffc42bd74c5159e5d602361dd016ef88..96b9e04867a44a2f92120ab69562b956c8c021b7 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1752,14 +1752,14 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             final org.bukkit.inventory.EquipmentSlot handSlot = (hand != null) ? org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand) : null;
-            final EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot);
+            final EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot, true); // Leaves - can dont shrink
             event.setCancelled(itemStack == null);
             this.level().getCraftServer().getPluginManager().callEvent(event);
             if (event.isCancelled()) {
                 // Set death protection to null as the event was cancelled. Prevent any attempt at resurrection.
                 deathProtection = null;
             } else {
-                if (!itemInHand.isEmpty() && itemStack != null) { // Paper - only reduce item if actual totem was found
+                if (!itemInHand.isEmpty() && itemStack != null && event.isShrink()) { // Paper - only reduce item if actual totem was found // Leaves - can dont shrink
                     itemInHand.shrink(1);
                 }
                 // Paper start - fix NPE when pre-cancelled EntityResurrectEvent is uncancelled
