From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 5 Feb 2025 23:11:32 +0800
Subject: [PATCH] Servux Protocol


diff --git a/net/minecraft/server/ServerTickRateManager.java b/net/minecraft/server/ServerTickRateManager.java
index 8d0f810100a4eac65831913feeb9a9eeb0e6006a..a714017fd7b69c0c1d71724cd4524f46c8d1cba8 100644
--- a/net/minecraft/server/ServerTickRateManager.java
+++ b/net/minecraft/server/ServerTickRateManager.java
@@ -133,4 +133,10 @@ public class ServerTickRateManager extends TickRateManager {
         player.connection.send(ClientboundTickingStatePacket.from(this));
         player.connection.send(ClientboundTickingStepPacket.from(this));
     }
+
+    // Leaves start - servux
+    public long getRemainingSprintTicks() {
+        return remainingSprintTicks;
+    }
+    // Leaves end - servux
 }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3f02da2014ed8e4169f4bb819b4d9e61154b7b06..c2531d2b08fa3d4179d5b30ca02d9d3f08a11dcc 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2342,9 +2342,21 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         // Paper end
     }
 
+
     @Override
     public LevelData.RespawnData getRespawnData() {
         return this.getServer().getRespawnData();
+        // TODO: 1.21.9 removed spawn chunk, should we keep this?
+        // int i = this.getGameRules().getInt(GameRules.RULE_SPAWN_CHUNK_RADIUS) + 1;
+        // if (i > 1) {
+        //     // Paper start - allow disabling gamerule limits
+        //     for (ChunkPos chunkPos : io.papermc.paper.util.MCUtil.getSpiralOutChunks(pos, i - 2)) {
+        //         this.getChunkSource().addTicketAtLevel(TicketType.START, chunkPos, ChunkLevel.ENTITY_TICKING_LEVEL);
+        //     }
+        //     // Paper end - allow disabling gamerule limits
+        // }
+        // this.lastSpawnChunkRadius = i;
+        // org.leavesmc.leaves.protocol.servux.ServuxHudDataProtocol.refreshSpawnMetadata = true; // Leaves - servux
     }
 
     public LongSet getForceLoadedChunks() {
