From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH] Stick can change ArmorStand arm status


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index a3cc0001a949597e345d7919c3f6109fa4a949ad..dfae614a67476d649be2a8c2cd8258abe12842e3 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -341,6 +341,12 @@ public class ArmorStand extends LivingEntity {
                     return InteractionResult.SUCCESS_SERVER;
                 }
             } else {
+                // Leaves start - stick can change ArmorStand arm status
+                if (org.leavesmc.leaves.LeavesConfig.modify.stickChangeArmorStandArmStatus && itemInHand.is(Items.STICK) && player.isShiftKeyDown()) {
+                    setShowArms(!showArms());
+                }
+                // Leaves end - stick can change ArmorStand arm status
+
                 if (this.isDisabled(equipmentSlotForItem)) {
                     return InteractionResult.FAIL;
                 }
