From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 22 Jul 2024 18:30:42 +0800
Subject: [PATCH] Can disable LivingEntity aiStep alive check


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 6a0ebbdfcc0bc07915574c1cf87463e82213ca7e..b87576911f99c04f473e29265ede981502bbba8f 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3433,7 +3433,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
         }
 
-        if (!this.isRemoved()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableLivingEntityAiStepAliveCheck || !this.isRemoved()) { // Leaves - can disable check
             this.aiStep();
         }
 
