From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH] Stick can change ArmorStand arm status


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index 46369e554be8c72e6fc2e6d58374dbfc77d94879..d7725b5ca689e3d5b512baab04e113be77c0b2ee 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -228,6 +228,13 @@ public class ArmorStand extends LivingEntity {
                     return InteractionResult.SUCCESS_SERVER;
                 }
             } else {
+                // Leaves start - stick can change ArmorStand arm status
+                if (org.leavesmc.leaves.LeavesConfig.modify.stickChangeArmorStandArmStatus && itemInHand.is(Items.STICK) && player.isShiftKeyDown()) {
+                    setShowArms(!showArms());
+                    return InteractionResult.FAIL;
+                }
+                // Leaves end - stick can change ArmorStand arm status
+
                 if (this.isDisabled(equipmentSlotForItem)) {
                     return InteractionResult.FAIL;
                 }
