From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 10 May 2025 00:49:51 +0800
Subject: [PATCH] Vanilla creative pickup behavior


diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index da8bab81e135a9b240fbeac0e9e1d2aa12c6212d..0ca8888289a91fc966e6fb49d1b1f73ef76ad345 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -175,6 +175,13 @@ public class Inventory implements Container, Nameable {
     // CraftBukkit start - Watch method above! :D
     public int canHold(ItemStack itemStack) {
         int remains = itemStack.getCount();
+
+        // Leaves start - vanilla creative pickup behavior
+        if (player.hasInfiniteMaterials()) {
+            return remains;
+        }
+        // Leaves end - vanilla creative pickup behavior
+
         for (int slot = 0; slot < this.items.size(); ++slot) {
             ItemStack itemInSlot = this.getItem(slot);
             if (itemInSlot.isEmpty()) {
