From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH] Stick can change ArmorStand arm status


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index 79ae039d1908830e81912d230b3540a4920e2017..a35415dced4c15ece829cea74010f42f0cb00441 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -229,6 +229,13 @@ public class ArmorStand extends LivingEntity {
                     return InteractionResult.SUCCESS_SERVER;
                 }
             } else {
+                // Leaves start - stick can change ArmorStand arm status
+                if (org.leavesmc.leaves.LeavesConfig.modify.stickChangeArmorStandArmStatus && itemInHand.is(Items.STICK) && player.isShiftKeyDown()) {
+                    setShowArms(!showArms());
+                    return InteractionResult.FAIL;
+                }
+                // Leaves end - stick can change ArmorStand arm status
+
                 if (this.isDisabled(equipmentSlotForItem)) {
                     return InteractionResult.FAIL;
                 }
