From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 18:04:35 +0800
Subject: [PATCH] Spectator dont get Advancement


diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 658db8460122db271bbc3c2c0844574b8d75f044..c5f99ca1167fd912c7fb46c9835f4e4766c152aa 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -168,6 +168,11 @@ public class PlayerAdvancements {
     }
 
     public boolean award(final AdvancementHolder holder, final String criterion) {
+        // Leaves start - spectator don't get advancement
+        if (org.leavesmc.leaves.LeavesConfig.modify.spectatorDontGetAdvancement && player.gameMode.getGameModeForPlayer() == net.minecraft.world.level.GameType.SPECTATOR) {
+            return false;
+        }
+        // Leaves end - spectator don't get advancement
         boolean result = false;
         AdvancementProgress progress = this.getOrStartProgress(holder);
         boolean wasDone = progress.isDone();
