From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 4 Sep 2023 22:09:10 +0800
Subject: [PATCH] RNG Fishing


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index bbfd9db9363e2fa35a1be95c466e1b758d6efb61..5f8711b3d6ffcbd9ecfedec90ff09e6e0ddfa228 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -522,7 +522,7 @@ public class FishingHook extends Projectile {
                     .withLuck(this.luck + owner.getLuck())
                     .create(LootContextParamSets.FISHING);
                 LootTable lootTable = this.level().getServer().reloadableRegistries().getLootTable(BuiltInLootTables.FISHING);
-                List<ItemStack> items = lootTable.getRandomItems(params);
+                List<ItemStack> items = org.leavesmc.leaves.LeavesConfig.modify.oldMC.rngFishing ? lootTable.getRandomItems(params, this.random) : lootTable.getRandomItems(params); // Leaves - world random
                 CriteriaTriggers.FISHING_ROD_HOOKED.trigger((ServerPlayer)owner, rod, this, items);
 
                 for (ItemStack itemStack : items) {
