From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Thu, 14 Aug 2025 15:47:58 +0800
Subject: [PATCH 5/5] Do not tick Arrow life regardless


diff --git a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
index 6ac6e7cb688ce85f731a0a3be81be4beb38767cb..c5664d824a276e43715733af4172dcb6dad35d14 100644
--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -229,8 +229,10 @@ public abstract class AbstractArrow extends Projectile {
             }
         } else {
             // Paper start - tick life regardless after X seconds
-            final io.papermc.paper.configuration.type.number.IntOr.Disabled maxArrowDespawnInvulnerability = this.level().paperConfig().entities.spawning.maxArrowDespawnInvulnerability;
-            if (maxArrowDespawnInvulnerability.enabled() && this.tickCount > maxArrowDespawnInvulnerability.intValue()) this.tickDespawn();
+            // Leaves start - do not tick arrow life regardless
+            // final io.papermc.paper.configuration.type.number.IntOr.Disabled maxArrowDespawnInvulnerability = this.level().paperConfig().entities.spawning.maxArrowDespawnInvulnerability;
+            // if (maxArrowDespawnInvulnerability.enabled() && this.tickCount > maxArrowDespawnInvulnerability.intValue()) this.tickDespawn();
+            // Leaves end - do not tick arrow life regardless
             // Paper end - tick life regardless after X seconds
             this.inGroundTime = 0;
             Vec3 originalPosition = this.position();
