From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 23 May 2023 17:10:00 +0800
Subject: [PATCH 1/5] Disable distance check for UseItemOnPacket


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 404b88d30778d07745368b14e739e91fb4a95cda..5a49f2c8b46adbdf547708cdd0e59d053bada273 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2139,7 +2139,7 @@ public class ServerGamePacketListenerImpl
                 if (this.player.isWithinBlockInteractionRange(pos, 1.0)) {
                     Vec3 distance = location.subtract(Vec3.atCenterOf(pos));
                     double limit = 1.0000001;
-                    if (Math.abs(distance.x()) < 1.0000001 && Math.abs(distance.y()) < 1.0000001 && Math.abs(distance.z()) < 1.0000001) {
+                    if (org.leavesmc.leaves.LeavesConfig.modify.disableDistanceCheckForUseItem ||Math.abs(distance.x()) < 1.0000001 && Math.abs(distance.y()) < 1.0000001 && Math.abs(distance.z()) < 1.0000001) { // Leaves - disableDistanceCheckForUseItem
                         Direction direction = blockHit.getDirection();
                         this.player.resetLastActionTime();
                         int maxY = level.getMaxY();
