From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 26 Aug 2025 03:14:04 +0800
Subject: [PATCH] Modify end void rings generation


diff --git a/net/minecraft/world/level/levelgen/DensityFunctions.java b/net/minecraft/world/level/levelgen/DensityFunctions.java
index 985bb77d065c96cc11e9191bace21cbed15f14f1..965e945a45fca12d49581155b30a374891e6f266 100644
--- a/net/minecraft/world/level/levelgen/DensityFunctions.java
+++ b/net/minecraft/world/level/levelgen/DensityFunctions.java
@@ -554,7 +554,7 @@ public final class DensityFunctions {
             int chunkZ = sectionZ / 2;
             int subSectionX = sectionX % 2;
             int subSectionZ = sectionZ % 2;
-            float doffs = 100.0F - (ca.spottedleaf.moonrise.common.PlatformHooks.get().configFixMC159283() ? Mth.sqrt((long) sectionX * (long) sectionX + (long) sectionZ * (long) sectionZ) : Mth.sqrt(sectionX * sectionX + sectionZ * sectionZ)) * 8.0F; // Paper - cast ints to long when MC-159283 fix enabled
+            float doffs = 100.0F - (!org.leavesmc.leaves.LeavesConfig.fix.vanillaEndVoidRings && ca.spottedleaf.moonrise.common.PlatformHooks.get().configFixMC159283() ? Mth.sqrt((long) sectionX * (long) sectionX + (long) sectionZ * (long) sectionZ) : Mth.sqrt(sectionX * sectionX + sectionZ * sectionZ)) * 8.0F; // Paper - cast ints to long when MC-159283 fix enabled // Leaves - customize end void rings generation
             doffs = Mth.clamp(doffs, -100.0F, 80.0F);
 
             NoiseCache cache = noiseCache.get().computeIfAbsent(islandNoise, noiseKey -> new NoiseCache()); // Paper - Perf: Optimize end generation
