From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Sat, 20 Jan 2024 02:50:56 +0800
Subject: [PATCH 1/5] Placing locked hopper no longer send NC updates


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 7034abfaf58bd685e185025a30b24754fd067734..ece30dfc6c31d4ebb0248e3c6885ab18d1c865b4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1159,7 +1159,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
 
             if ((flags & Block.UPDATE_NEIGHBORS) != 0) {
-                this.updateNeighborsAt(pos, blockState.getBlock());
+                // Leaves start - Placing locked hopper doesn't send NC updates.
+                if (!(blockState.getBlock() instanceof net.minecraft.world.level.block.HopperBlock) || blockState.getValue(net.minecraft.world.level.block.HopperBlock.ENABLED)) {
+                    this.updateNeighborsAt(pos, blockState.getBlock());
+                }
+                // Leaves end - Placing locked hopper doesn't send NC updates.
                 if (!this.isClientSide() && state.hasAnalogOutputSignal()) {
                     this.updateNeighbourForOutputSignal(pos, newState.getBlock());
                 }
