From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Sat, 20 Jan 2024 02:50:56 +0800
Subject: [PATCH] Placing locked hopper no longer send NC updates


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index e34e901dce06c6bc6060418593010dcfb38e9396..077992e18217dc8f68f4e15fe8e71fe8ce59cb2c 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1160,7 +1160,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
 
             if ((flags & Block.UPDATE_NEIGHBORS) != 0) {
-                this.updateNeighborsAt(pos, blockState.getBlock());
+                // Leaves start - Placing locked hopper doesn't send NC updates.
+                if (!(blockState.getBlock() instanceof net.minecraft.world.level.block.HopperBlock) || blockState.getValue(net.minecraft.world.level.block.HopperBlock.ENABLED)) {
+                    this.updateNeighborsAt(pos, blockState.getBlock());
+                }
+                // Leaves end - Placing locked hopper doesn't send NC updates.
                 if (!this.isClientSide() && state.hasAnalogOutputSignal()) {
                     this.updateNeighbourForOutputSignal(pos, newState.getBlock());
                 }
