From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH 3/5] Stick can change ArmorStand arm status


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index fc4fce0b1d57a62c1e5c80a5b91c2b1a77d4bace..3dd6f93bf623a1e3f7429d57ca175585ab7fbc04 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -231,6 +231,13 @@ public class ArmorStand extends LivingEntity {
                     return InteractionResult.SUCCESS_SERVER;
                 }
             } else {
+                // Leaves start - stick can change ArmorStand arm status
+                if (org.leavesmc.leaves.LeavesConfig.modify.stickChangeArmorStandArmStatus && itemStack.is(Items.STICK) && player.isShiftKeyDown()) {
+                    setShowArms(!showArms());
+                    return InteractionResult.FAIL;
+                }
+                // Leaves end - stick can change ArmorStand arm status
+
                 if (this.isDisabled(itemInHandSlot)) {
                     return InteractionResult.FAIL;
                 }
