From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 25 May 2023 16:37:06 +0800
Subject: [PATCH] No feather falling trample


diff --git a/net/minecraft/world/level/block/FarmBlock.java b/net/minecraft/world/level/block/FarmBlock.java
index da822a89e134e98696308315799d872944e7712f..9d613f5ff53609f05025ae46f7f3758c0d556626 100644
--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -116,6 +116,13 @@ public class FarmBlock extends Block {
             && entity instanceof LivingEntity
             && (entity instanceof Player || serverLevel.getGameRules().get(GameRules.MOB_GRIEFING))
             && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
+                // Leaves start - noFeatherFallingTrample
+                if (org.leavesmc.leaves.LeavesConfig.modify.noFeatherFallingTrample) {
+                    if (net.minecraft.world.item.enchantment.EnchantmentHelper.getEnchantmentLevel(level.registryAccess().lookupOrThrow(net.minecraft.core.registries.Registries.ENCHANTMENT).getOrThrow(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING), (LivingEntity) entity) > 0) {
+                    return;
+                    }
+                }
+                // Leaves end - noFeatherFallingTrample
                 // CraftBukkit start - Interact soil
                 org.bukkit.event.Cancellable cancellable;
                 if (entity instanceof Player) {
