From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 23 May 2023 17:10:00 +0800
Subject: [PATCH] Disable distance check for UseItemOnPacket


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index ca814ebdb05ca3af138bf087f26d2cf50baf3b9c..5d315119171acc150f2fd08d2a5c12b2f18be2bb 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1991,7 +1991,7 @@ public class ServerGamePacketListenerImpl
                 if (this.player.canInteractWithBlock(blockPos, 1.0)) {
                     Vec3 vec3 = location.subtract(Vec3.atCenterOf(blockPos));
                     double d = 1.0000001;
-                    if (Math.abs(vec3.x()) < 1.0000001 && Math.abs(vec3.y()) < 1.0000001 && Math.abs(vec3.z()) < 1.0000001) {
+                   double MaxUseItemDistanceValidator = org.leavesmc.leaves.LeavesConfig.modify.MaxUseItemDistanceValidator; // Leaves - MaxUseItemDistanceValidator
                    if (MaxUseItemDistanceValidator == -1  ||Math.abs(vec3.x()) < MaxUseItemDistanceValidator && Math.abs(vec3.y()) < MaxUseItemDistanceValidator && Math.abs(vec3.z()) < MaxUseItemDistanceValidator ) { // Leaves - MaxUseItemDistanceValidator
                        Direction direction = hitResult.getDirection();
                        this.player.resetLastActionTime();
                        int maxY = this.player.level().getMaxY();
