From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:22 +0800
Subject: [PATCH] Bytebuf API


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 9218322801c41657c00928ef90b0de99df1670f6..1d31d63406c2d25221a1f1c0313ee96979d46ed2 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -177,6 +177,13 @@ public abstract class PlayerList {
         boolean immediateRespawn = gameRules.get(GameRules.IMMEDIATE_RESPAWN);
         boolean reducedDebugInfo = gameRules.get(GameRules.REDUCED_DEBUG_INFO);
         boolean doLimitedCrafting = gameRules.get(GameRules.LIMITED_CRAFTING);
+
+            // Leaves start - Bytebuf API
+            if (!(player instanceof org.leavesmc.leaves.bot.ServerBot) && !(player instanceof org.leavesmc.leaves.replay.ServerPhotographer)) {
+                this.cserver.getBytebufHandler().injectPlayer(player);
+            }
+            // Leaves end - Bytebuf API
+
         playerConnection.send(
             new ClientboundLoginPacket(
                 player.getId(),
