From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 1 Jul 2025 00:19:23 +0800
Subject: [PATCH 3/5] Configurable item damage check


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 54d67c847c1de9ee216cedb0527136f50223ea2d..63c25557dde82fb22dc6997953beeea8e72b8f44 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -592,11 +592,11 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
     }
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
+        return org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableItemDamageCheck ? this.getOrDefault(DataComponents.DAMAGE, 0) : Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage()); // Leaves - disable check
     }
 
     public void setDamageValue(final int value) {
-        this.set(DataComponents.DAMAGE, Mth.clamp(value, 0, this.getMaxDamage()));
+        this.set(DataComponents.DAMAGE, org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableItemDamageCheck ? value : Mth.clamp(value, 0, this.getMaxDamage())); // Leaves - disable check
     }
 
     public int getMaxDamage() {
