From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 30 Mar 2022 08:58:45 +0000
Subject: [PATCH] Add isShrink to EntityResurrectEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 428982842a0c18aa3894c50af9e90d76d6ab416e..fa99a55c302391f6f9c35bb62b409867168c893d 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1748,14 +1748,14 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             final org.bukkit.inventory.EquipmentSlot handSlot = (usedHand != null) ? org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(usedHand) : null;
-            final EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot);
+            final EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot, true); // Leaves - can dont shrink
             event.setCancelled(protectionItem == null);
             this.level().getCraftServer().getPluginManager().callEvent(event);
             if (event.isCancelled()) {
                 // Set death protection to null as the event was cancelled. Prevent any attempt at resurrection.
                 protection = null;
             } else {
-                if (!itemStack.isEmpty() && protectionItem != null) { // Paper - only reduce item if actual totem was found
+                if (!itemStack.isEmpty() && protectionItem != null && event.isShrink()) { // Paper - only reduce item if actual totem was found // Leaves - can dont shrink
                     itemStack.shrink(1);
                 }
                 // Paper start - fix NPE when pre-cancelled EntityResurrectEvent is uncancelled
