From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 29 Nov 2025 21:01:21 +0800
Subject: [PATCH] Fix stacked container destroyed drop


diff --git a/net/minecraft/world/item/ItemUtils.java b/net/minecraft/world/item/ItemUtils.java
index 1b90cf3476899951bbdc5b47b142df80d68c54e0..ee028691972cd7d9c8dd22635635b6644a1eeeb3 100644
--- a/net/minecraft/world/item/ItemUtils.java
+++ b/net/minecraft/world/item/ItemUtils.java
@@ -42,6 +42,17 @@ public class ItemUtils {
     }
 
     public static void onContainerDestroyed(final ItemEntity container, final Stream<ItemStack> contents) {
+        if (org.leavesmc.leaves.LeavesConfig.fix.stackedContainerDestroyedDrop) {
+            java.util.List<ItemStack> list = contents.toList();
+            for (int i = 0; i < container.getItem().getCount(); i++) {
+                createDroppedItem(container, list.stream());
+            }
+        } else {
+            createDroppedItem(container, contents);
+        }
+    }
+
+    private static void createDroppedItem(final ItemEntity container, final Stream<ItemStack> contents) {
         Level level = container.level();
         if (!level.isClientSide()) {
             // Paper start - call EntityDropItemEvent
