From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 9 Jul 2023 16:51:47 +0800
Subject: [PATCH] Raider die skip self raid check


diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index 8270d76a753bfd26a4c8ef6610bee5c24ee59cfe..72551960faaedbf43d1c65ad061f6d366028a3e9 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -115,7 +115,7 @@ public abstract class Raider extends PatrollingMonster {
             Entity entity = cause.getEntity();
             Raid currentRaid = this.getCurrentRaid();
             if (currentRaid != null) {
-                if (this.isPatrolLeader()) {
+                if (!org.leavesmc.leaves.LeavesConfig.modify.skipSelfRaidCheck && this.isPatrolLeader()) { // Leaves - skip self raid check
                     currentRaid.removeLeader(this.getWave());
                 }
 
