From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 19:16:16 +0800
Subject: [PATCH] No block update command


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index 9cce442c16ced8d9320a5760580ff13f02cbf8f1..6c67c3b9e7c70366c2889932fd4c0265218656d4 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -381,7 +381,7 @@ public class ServerPlayerGameMode {
                 org.bukkit.block.BlockState state = bblock.getState();
                 this.level.captureDrops = new java.util.ArrayList<>();
                 // CraftBukkit end
-                BlockState blockState1 = block.playerWillDestroy(this.level, pos, blockState, this.player);
+                BlockState blockState1 = org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate() ? blockState : block.playerWillDestroy(this.level, pos, blockState, this.player); // Leaves - no block update
                 boolean flag = this.level.removeBlock(pos, false);
                 if (flag) {
                     block.destroy(this.level, pos, blockState1);
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 6202ea6f6f8e6b751434848fb6551c61162d076b..a8c20ef3be53bd462e4c352a08f4b7201a8af373 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1089,6 +1089,11 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
 
     @Override
     public boolean setBlock(BlockPos pos, BlockState state, int flags, int recursionLeft) {
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            flags = flags & ~1 | net.minecraft.world.level.block.Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
@@ -1190,6 +1195,11 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
         BlockState state = newState;
         BlockState blockState = oldState;
         BlockState blockState1 = currentState;
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            flags = flags & ~1 | net.minecraft.world.level.block.Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         if (blockState1 == state) {
             if (blockState != blockState1) {
                 this.setBlocksDirty(pos, blockState, blockState1);
@@ -1208,7 +1218,12 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
 
             if ((flags & 16) == 0 && recursionLeft > 0) {
                 int i = flags & -34;
-
+                // Leaves start - no block update
+                if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+                    this.updatePOIOnBlockStateChange(pos, blockState, blockState1);
+                    return;
+                }
+                // Leaves end - no block update
                 // CraftBukkit start
                 blockState.updateIndirectNeighbourShapes(this, pos, i, recursionLeft - 1); // Don't call an event for the old block to limit event spam
                 boolean cancelledUpdates = false; // Paper - Fix block place logic
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 16aa9f5996dc6eda95541fddb01e00e41305357a..6bd7d937bab8baf7c657e689e8a1b4069c207ff9 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -105,6 +105,7 @@ public class PistonBaseBlock extends DirectionalBlock {
     }
 
     private void checkIfExtend(Level level, BlockPos pos, BlockState state) {
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) return; // Leaves - no block update
         Direction direction = state.getValue(FACING);
         boolean neighborSignal = this.getNeighborSignal(level, pos, direction);
         if (neighborSignal && !state.getValue(EXTENDED)) {
diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 8942aa2ca4796c7c36c0955141627ea905e0ec64..2d2b7f3284171517d1e68cef802ae197e2e6f8bb 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -34,6 +34,11 @@ public interface NeighborUpdater {
     static void executeShapeUpdate(
         LevelAccessor level, Direction direction, BlockPos pos, BlockPos neighborPos, BlockState neighborState, int flags, int recursionLeft
     ) {
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            flags = flags & ~1 | net.minecraft.world.level.block.Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         BlockState blockState = level.getBlockState(pos);
         if ((flags & 128) == 0 || !blockState.is(Blocks.REDSTONE_WIRE)) {
             BlockState blockState1 = blockState.updateShape(level, level, pos, direction, neighborPos, neighborState, level.getRandom());
@@ -48,6 +53,7 @@ public interface NeighborUpdater {
 
     static void executeUpdate(Level level, BlockState state, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston, BlockPos sourcePos) {
         // Paper end - Add source block to BlockPhysicsEvent
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) return; // Leaves - no block update
         try {
             // CraftBukkit start
             org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(state), org.bukkit.craftbukkit.block.CraftBlock.at(level, sourcePos)); // Paper - Add source block to BlockPhysicsEvent
