From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 30 Jul 2022 01:11:30 +0800
Subject: [PATCH] Configurable MC-59471


diff --git a/net/minecraft/world/level/block/TripWireHookBlock.java b/net/minecraft/world/level/block/TripWireHookBlock.java
index 2c8a4febdbc0a8010e786293f5b110236757aa60..0187bc6660243c2af344bc3ecf013f0bb87f310e 100644
--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -210,10 +210,17 @@ public class TripWireHookBlock extends Block {
                     BlockPos testPosx = pos.relative(direction, i);
                     BlockState wireData = wireStates[i];
                     if (wireData != null) {
-                        BlockState testPosState = level.getBlockState(testPosx);
-                        if (testPosState.is(Blocks.TRIPWIRE) || testPosState.is(Blocks.TRIPWIRE_HOOK)) {
-                            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !testPosState.is(Blocks.TRIPWIRE)) level.setBlock(testPosx, wireData.trySetValue(ATTACHED, attached), Block.UPDATE_ALL); // Paper - prevent tripwire from updating
+                        // Leaves start - MC-59471
+                        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.tripwire.stringTripwireHookDuplicate) {
+                            level.setBlock(testPosx, wireData.trySetValue(ATTACHED, attached), Block.UPDATE_ALL);
+                            level.getBlockState(testPosx);
+                        } else {
+                            BlockState testPosState = level.getBlockState(testPosx);
+                            if (testPosState.is(Blocks.TRIPWIRE) || testPosState.is(Blocks.TRIPWIRE_HOOK)) {
+                                if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !testPosState.is(Blocks.TRIPWIRE)) level.setBlock(testPosx, wireData.trySetValue(ATTACHED, attached), Block.UPDATE_ALL); // Paper - prevent tripwire from updating
+                            }
                         }
+                        // Leaves end - MC-59471
                     }
                 }
             }
