From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 9 Jul 2023 16:51:47 +0800
Subject: [PATCH] Raider die skip self raid check


diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index a495789b2d21fa9a24d5dca4ecfa196ddce49466..7b697f1dfe8131296ef55cd88a8b9d7cd884eb44 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -156,7 +156,7 @@ public abstract class Raider extends PatrollingMonster {
     }
 
     public boolean hasRaid() {
-        return this.level() instanceof ServerLevel serverLevel && (this.getCurrentRaid() != null || serverLevel.getRaidAt(this.blockPosition()) != null);
+        return !org.leavesmc.leaves.LeavesConfig.modify.skipSelfRaidCheck && (this.level() instanceof ServerLevel serverLevel && (this.getCurrentRaid() != null || serverLevel.getRaidAt(this.blockPosition()) != null));  // Leaves - skip self raid check
     }
 
     public boolean hasActiveRaid() {
