From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 22 Jan 2024 11:41:13 +0800
Subject: [PATCH 4/5] Renewable sponges


diff --git a/net/minecraft/world/entity/monster/Guardian.java b/net/minecraft/world/entity/monster/Guardian.java
index b13aa1d5d175a05a856d55dc09b3c7c4265b1834..a34852bf64623cf94f6e788853a44d8579204279 100644
--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -442,6 +442,28 @@ public class Guardian extends Monster {
         }
     }
 
+    // Leaves start - renewable sponges
+    @Override
+    public void thunderHit(net.minecraft.server.level.ServerLevel level, net.minecraft.world.entity.LightningBolt lightning) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.renewableSponges && !this.isRemoved() && !(this instanceof ElderGuardian)) {
+            ElderGuardian elderGuardian = new ElderGuardian(EntityType.ELDER_GUARDIAN ,this.level());
+            elderGuardian.snapTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
+            elderGuardian.finalizeSpawn(level ,this.level().getCurrentDifficultyAt(elderGuardian.blockPosition()), EntitySpawnReason.CONVERSION, null);
+            elderGuardian.setNoAi(this.isNoAi());
+
+            if (this.hasCustomName()) {
+                elderGuardian.setCustomName(this.getCustomName());
+                elderGuardian.setCustomNameVisible(this.isCustomNameVisible());
+            }
+
+            this.level().addFreshEntity(elderGuardian);
+            this.discard();
+        } else {
+            super.thunderHit(level, lightning);
+        }
+    }
+    // Leaves end - renewable sponges
+
     private static class GuardianMoveControl extends MoveControl {
         private final Guardian guardian;
 
