From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Mon, 3 Feb 2025 23:11:33 -0500
Subject: [PATCH] Check frozen ticks before landing block


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 5e7068bf9b5364c382fccfb110ed2a26a5ac1731..0cd9a4ff2c480774487bd82531996a35db81ae14 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -572,11 +572,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+        if (org.leavesmc.leaves.LeavesConfig.performance.checkFrozenTicksBeforeLandingBlock || !this.getBlockStateOnLegacy().isAir()) { // Leaves - check frozen ticks before landing block
             int ticksFrozen = this.getTicksFrozen();
             if (ticksFrozen > 0) {
-                AttributeInstance attribute = this.getAttribute(Attributes.MOVEMENT_SPEED);
-                if (attribute == null) {
+                AttributeInstance attribute = !org.leavesmc.leaves.LeavesConfig.performance.checkFrozenTicksBeforeLandingBlock || !this.getBlockStateOnLegacy().isAir() ? this.getAttribute(Attributes.MOVEMENT_SPEED) : null; // Leaves - check frozen ticks before landing block
+                          ute == null) {
                     return;
                 }
 
