From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Sat, 17 May 2025 11:38:56 +0000
Subject: [PATCH 3/5] Do not reset placed block on exception


diff --git a/net/minecraft/world/item/BlockItem.java b/net/minecraft/world/item/BlockItem.java
index 3167be4b7c2c61ef725af4a832d759ab6e061af0..43e1c40d4c724d4ec0457e819f4b54f310ee0459 100644
--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -69,20 +69,22 @@ public class BlockItem extends Item {
                     BlockState placedState = level.getBlockState(pos);
                     if (placedState.is(placementState.getBlock())) {
                         placedState = this.updateBlockStateFromTag(pos, level, itemStack, placedState);
+                        // Leaves start - we do not need this
                         // Paper start - Reset placed block on exception
-                        try {
+                        // try {
                         this.updateCustomBlockEntityTag(pos, level, player, itemStack, placedState);
                         updateBlockEntityComponents(level, pos, itemStack);
-                        } catch (Exception ex) {
-                            ((org.bukkit.craftbukkit.block.CraftBlockState) previousState).revertPlace();
-                            if (player instanceof ServerPlayer serverPlayer) {
-                                net.minecraft.server.MinecraftServer.LOGGER.warn("Player {} tried placing invalid block", player.getScoreboardName(), ex);
-                                serverPlayer.connection.disconnect(net.minecraft.network.chat.Component.literal("Packet processing error"));
-                                return InteractionResult.FAIL;
-                            }
-                            throw ex; // Rethrow exception if not placed by a player
-                        }
+                        // } catch (Exception ex) {
+                        //     ((org.bukkit.craftbukkit.block.CraftBlockState) previousState).revertPlace();
+                        //     if (player instanceof ServerPlayer serverPlayer) {
+                        //         net.minecraft.server.MinecraftServer.LOGGER.warn("Player {} tried placing invalid block", player.getScoreboardName(), ex);
+                        //         serverPlayer.connection.disconnect(net.minecraft.network.chat.Component.literal("Packet processing error"));
+                        //         return InteractionResult.FAIL;
+                        //     }
+                        //     throw ex; // Rethrow exception if not placed by a player
+                        // }
                         // Paper end - Reset placed block on exception
+                        // Leaves end - we dot not need this
                         placedState.getBlock().setPlacedBy(level, pos, placedState, player, itemStack);
                         // CraftBukkit start - special case for handling block placement with water lilies, frog spawn and snow buckets
                         if (player != null && (this instanceof PlaceOnWaterBlockItem || this instanceof SolidBucketItem)) {
