From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 30 Jul 2023 12:20:16 +0800
Subject: [PATCH] Zero tick plants


diff --git a/net/minecraft/world/level/block/BambooStalkBlock.java b/net/minecraft/world/level/block/BambooStalkBlock.java
index 81e2a279d4c29f5fe52387875489239515a8c82b..66ef754fa72626886b14ea2941a851c847094efd 100644
--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -108,6 +108,10 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+            // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index 9662828596de15e5b855f12d2a3b014763b3fc2e..67490a7db2db521863bcc4d923ed435de9d6e961 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -47,6 +47,10 @@ public class CactusBlock extends Block {
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+            // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/ChorusFlowerBlock.java b/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 184c971852a9eb387484e4ce89a26a69567cbeda..fe668a8c4c7449731901dc3bee8996b7baeaed5c 100644
--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -49,6 +49,9 @@ public class ChorusFlowerBlock extends Block {
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+            // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
index bac7f990282fd7c676c2f8c40d7fd87badb1e284..876f2e24c5e62eaf3b355ea313d7a9a1b473ffc8 100644
--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -155,4 +155,15 @@ public abstract class GrowingPlantHeadBlock extends GrowingPlantBlock implements
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    // Leaves start - zero tick plants
+    @Override
+    public void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        if (!state.canSurvive(world, pos)) {
+            world.destroyBlock(pos, true);
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+        }
+    }
+    // Leaves end - zero tick plants
 }
diff --git a/net/minecraft/world/level/block/SugarCaneBlock.java b/net/minecraft/world/level/block/SugarCaneBlock.java
index 4437fd5724e094e8d2edb97ca9bf494ef57f5b79..1060c57dde837481a898d9a5f45c7f1fd7174c4e 100644
--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -43,6 +43,10 @@ public class SugarCaneBlock extends Block {
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+            // Leaves end - zero tick plants
         }
     }
 
