From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 12:00:59 +0800
Subject: [PATCH] Fix villagers dont release memory


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 09d357035d2b0a75a70710bf15b34cfae2501a43..591e413dba3459667cf2643e26684c8fae8d4c82 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3805,7 +3805,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this;
     }
 
-    private Entity teleportCrossDimension(ServerLevel level, TeleportTransition teleportTransition) {
+    protected Entity teleportCrossDimension(ServerLevel level, TeleportTransition teleportTransition) { // Leaves - private -> protected
         List<Entity> passengers = this.getPassengers();
         List<Entity> list = new ArrayList<>(passengers.size());
         this.ejectPassengers();
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index e0e0d2ea7fc60e3142c675404d152eca60263240..956b949b42a7041290838dde816d3b79e1f9e0d4 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -1033,4 +1033,19 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             return super.applyImplicitComponent(component, value);
         }
     }
+
+    // Leaves start - fixes a memory leak when villagers get moved to another world
+    @Override
+    public Entity teleportCrossDimension(ServerLevel world, net.minecraft.world.level.portal.TeleportTransition transition) {
+        if (org.leavesmc.leaves.LeavesConfig.performance.villagersDontReleaseMemoryFix) {
+            this.releaseAllPois();
+            this.getBrain().eraseMemory(MemoryModuleType.HOME);
+            this.getBrain().eraseMemory(MemoryModuleType.JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.MEETING_POINT);
+            this.refreshBrain(transition.newLevel());
+        }
+        return super.teleportCrossDimension(world, transition);
+    }
+    // Leaves end - fixes a memory leak when villagers get moved to another world
 }
