From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 16:42:23 +0800
Subject: [PATCH 1/5] Fix trapdoor feature


diff --git a/net/minecraft/world/level/block/TrapDoorBlock.java b/net/minecraft/world/level/block/TrapDoorBlock.java
index 84404387841e8c720a016d8aa72b9ea344b883d7..e2752934dcedc5ad7ab3a794663f51af52495324 100644
--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -144,6 +144,7 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                     signal = eventRedstone.getNewCurrent() > 0;
                 }
                 // CraftBukkit end
+                /* Leaves - it is feature, not bug!
                 // Paper start - break redstone on trapdoors early
                 boolean open = state.getValue(TrapDoorBlock.OPEN) != signal;
                 // note: this must run before any state for this block/its neighbours are written to the world
@@ -162,8 +163,9 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                         }
                     }
                 }
-                if (open) {
-                    // Paper end - break redstone on trapdoors early
+                */
+
+                if (state.getValue(TrapDoorBlock.OPEN) != signal) {
                     state = state.setValue(OPEN, signal);
                     this.playSound(null, level, pos, signal);
                 }
