From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 4 Sep 2023 22:09:10 +0800
Subject: [PATCH] RNG Fishing


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index 30e3c19adcdc55fa468257037ca49275c09de324..c6e87d453bfccbd7706ec45b105e469f8a27ba99 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -531,7 +531,7 @@ public class FishingHook extends Projectile {
                     .withLuck(this.luck + playerOwner.getLuck())
                     .create(LootContextParamSets.FISHING);
                 LootTable lootTable = this.level().getServer().reloadableRegistries().getLootTable(BuiltInLootTables.FISHING);
-                List<ItemStack> randomItems = lootTable.getRandomItems(lootParams);
+                List<ItemStack> randomItems = org.leavesmc.leaves.LeavesConfig.modify.oldMC.rngFishing ? lootTable.getRandomItems(lootParams, this.random) : lootTable.getRandomItems(lootParams); // Leaves - world random
                 CriteriaTriggers.FISHING_ROD_HOOKED.trigger((ServerPlayer)playerOwner, stack, this, randomItems);
 
                 for (ItemStack itemStack : randomItems) {
