From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 16 Aug 2023 13:10:58 +0800
Subject: [PATCH] Fix minecraft hopper not work without player


diff --git a/net/minecraft/world/entity/vehicle/MinecartHopper.java b/net/minecraft/world/entity/vehicle/MinecartHopper.java
index 41a6ec508a10a49a37539d2f10171d15c233b280..ae6246635b7961b91a217b7f4f9d9d0f41560207 100644
--- a/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -101,6 +101,13 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
         }
     }
 
+    // Leaves start - tick every
+    @Override
+    public void inactiveTick() {
+        this.tick();
+    }
+    // Leaves start - tick every
+
     public boolean suckInItems() {
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             this.immunize(); // Paper
