From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 5 Sep 2023 08:49:01 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/net/minecraft/world/entity/monster/skeleton/Skeleton.java b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
index 94a01d35457053bb4265acb5542ae2f83671d9bb..f1638ebacf61d4baffea7eff60605c56ca03bb2e 100644
--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -130,4 +130,25 @@ public class Skeleton extends AbstractSkeleton {
     SoundEvent getStepSound() {
         return SoundEvents.SKELETON_STEP;
     }
+
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropCustomDeathLoot(net.minecraft.server.level.ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
+        super.dropCustomDeathLoot(level, damageSource, recentlyHit);
+        if (damageSource.getEntity() instanceof net.minecraft.world.entity.monster.Creeper creeper && creeper.isPowered() && !creeper.droppedSkulls) {
+            creeper.droppedSkulls = true;
+            this.spawnAtLocation(level, net.minecraft.world.item.Items.SKELETON_SKULL);
+        }
+    }
+
+    @Override
+    protected void dropFromLootTable(net.minecraft.server.level.ServerLevel level, DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(level, damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.getRootVehicle().getType() == EntityType.SPIDER && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
 }
diff --git a/net/minecraft/world/entity/monster/spider/Spider.java b/net/minecraft/world/entity/monster/spider/Spider.java
index 64a84f73d48deb254345a5c25f3fe650591e64a1..c4666ab51a1c308fbd632f8b598acb847f2b3ae9 100644
--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -141,6 +141,18 @@ public class Spider extends Monster {
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(net.minecraft.server.level.ServerLevel level, DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(level, damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Override
     public @Nullable SpawnGroupData finalizeSpawn(
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData spawnGroupData
