From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Mon, 29 Sep 2025 10:24:45 +0800
Subject: [PATCH] Do not check NaN values in entities


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d3b6291149af95a0a7256674288be5c5a49c896b..fd7ff66d5e1263e83b7ef5501523a6512a8e701d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2579,9 +2579,11 @@ public abstract class Entity
             }
             } // CraftBukkit
 
+            // Leaves start - Do not check NaN values
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
             // CraftBukkit start - Checking for NaN pitch/yaw and resetting to zero
             // TODO: make sure this is the best way to address this.
+            /*
             if (Float.isNaN(this.yRot)) {
                 this.yRot = 0;
             }
@@ -2589,7 +2591,9 @@ public abstract class Entity
             if (Float.isNaN(this.xRot)) {
                 this.xRot = 0;
             }
+            */
             // CraftBukkit end
+            // Leaves end - Do not check NaN values
             output.store("Rotation", Vec2.CODEC, new Vec2(this.getYRot(), this.getXRot()));
             output.putDouble("fall_distance", this.fallDistance);
             output.putShort("Fire", (short)this.remainingFireTicks);
