From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 5 Sep 2023 08:49:01 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/net/minecraft/world/entity/monster/Skeleton.java b/net/minecraft/world/entity/monster/Skeleton.java
index 4eb35a7a2dfcb4803c00e1e9179fea8186b15d41..b81ca006d33f14faf54ecce3ad4c7d198b9794d0 100644
--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -129,4 +129,26 @@ public class Skeleton extends AbstractSkeleton {
     SoundEvent getStepSound() {
         return SoundEvents.SKELETON_STEP;
     }
+
+
+    @Override
+    protected void dropCustomDeathLoot(ServerLevel level, DamageSource damageSource, boolean recentlyHit) {
+        super.dropCustomDeathLoot(level, damageSource, recentlyHit);
+        if (damageSource.getEntity() instanceof Creeper creeper && creeper.canDropMobsSkull()) {
+            creeper.increaseDroppedSkulls();
+            this.spawnAtLocation(level, Items.SKELETON_SKULL);
+        }
+    }
+
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(ServerLevel level, DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(level, damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.getRootVehicle().getType() == EntityType.SPIDER && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
 }
diff --git a/net/minecraft/world/entity/monster/Spider.java b/net/minecraft/world/entity/monster/Spider.java
index 13f49a2e43adbbbb4ec9cd5b479814fefcca1cce..44a40741128b719626a793ded9e314371c99ad3f 100644
--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -139,6 +139,18 @@ public class Spider extends Monster {
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(net.minecraft.server.level.ServerLevel level, DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(level, damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Nullable
     @Override
     public SpawnGroupData finalizeSpawn(
