From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 13:31:57 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/net/minecraft/world/entity/monster/skeleton/Skeleton.java b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
index f2759462a4423d2c5d31d2c7fbfe96dc61081c77..28142a6131426eb6aee85572c8b0574331e414e9 100644
--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -130,4 +130,25 @@ public class Skeleton extends AbstractSkeleton {
     SoundEvent getStepSound() {
         return SoundEvents.SKELETON_STEP;
     }
+
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropCustomDeathLoot(final net.minecraft.server.level.ServerLevel level, final DamageSource source, final boolean playerKilled) {
+        super.dropCustomDeathLoot(level, source, playerKilled);
+        if (source.getEntity() instanceof net.minecraft.world.entity.monster.Creeper creeper && creeper.isPowered() && !creeper.droppedSkulls) {
+            creeper.droppedSkulls = true;
+            this.spawnAtLocation(level, net.minecraft.world.item.Items.SKELETON_SKULL);
+        }
+    }
+
+    @Override
+    protected void dropFromLootTable(final net.minecraft.server.level.ServerLevel level, final DamageSource source, final boolean playerKilled) {
+        super.dropFromLootTable(level, source, playerKilled);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.getRootVehicle().getType() == EntityType.SPIDER && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
 }
diff --git a/net/minecraft/world/entity/monster/spider/Spider.java b/net/minecraft/world/entity/monster/spider/Spider.java
index 1507f340b16fd95985bfe49870e923c626833d18..975693c1aa4a6e441040ed8dc9cbb77ba33ddd68 100644
--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -141,6 +141,18 @@ public class Spider extends Monster {
         this.entityData.set(DATA_FLAGS_ID, flags);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(final net.minecraft.server.level.ServerLevel level, final net.minecraft.world.damagesource.DamageSource source, final boolean playerKilled) {
+        super.dropFromLootTable(level, source, playerKilled);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Override
     public @Nullable SpawnGroupData finalizeSpawn(
         final ServerLevelAccessor level, final DifficultyInstance difficulty, final EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
