From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 5 Sep 2023 08:49:01 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/net/minecraft/world/entity/monster/Spider.java b/net/minecraft/world/entity/monster/Spider.java
index af0305079a367899708ee2bbac82aefaa9129d2f..e9a4734cddc33785ca6811f8f566904ace78fd78 100644
--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -139,6 +139,18 @@ public class Spider extends Monster {
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(net.minecraft.server.level.ServerLevel wolrd, DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(wolrd, damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.modify.spiderJockeysDropGapples) {
+                this.spawnAtLocation(wolrd, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Nullable
     @Override
     public SpawnGroupData finalizeSpawn(
