From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 28 Sep 2023 17:07:02 +0800
Subject: [PATCH] CCE update suppression


diff --git a/net/minecraft/world/level/block/ShulkerBoxBlock.java b/net/minecraft/world/level/block/ShulkerBoxBlock.java
index 02f3e470ed70de87deeb8180f50da11724bf5198..1b6bddc15f39c381050eb20c5ad43581d471d876 100644
--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -183,7 +183,9 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
     protected int getAnalogOutputSignal(BlockState state, Level level, BlockPos pos, Direction direction) {
         // Leaves start - update suppression crash fix
         try {
-            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
+            return org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.cceUpdateSuppression ?
+                    AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) level.getBlockEntity(pos)) : // Leaves - make cce happy(?)
+                    AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
         } catch (ClassCastException ex) {
             if (org.leavesmc.leaves.LeavesConfig.modify.updateSuppressionCrashFix) {
                 throw new org.leavesmc.leaves.util.UpdateSuppressionException(pos, null, this, null, ex);
