From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 13:32:26 +0800
Subject: [PATCH] Copper Bulb 1 gt delay


diff --git a/net/minecraft/world/level/block/CopperBulbBlock.java b/net/minecraft/world/level/block/CopperBulbBlock.java
index cc07d62b43908ab65c5b739633faef27c134f706..313f9114288e2b144f2d560ad6b7a3684426d359 100644
--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -33,7 +33,13 @@ public class CopperBulbBlock extends Block {
     @Override
     protected void onPlace(final BlockState state, final Level level, final BlockPos pos, final BlockState oldState, final boolean movedByPiston) {
         if (oldState.getBlock() != state.getBlock() && level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt delay
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
@@ -42,9 +48,24 @@ public class CopperBulbBlock extends Block {
         final BlockState state, final Level level, final BlockPos pos, final Block block, final @Nullable Orientation orientation, final boolean movedByPiston
     ) {
         if (level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt delay
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
+        }
+    }
+
+    // Leaves start - copper bulb 1 gt delay
+    @Override
+    public void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final net.minecraft.util.RandomSource random) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+            this.checkAndFlip(state, level, pos);
         }
     }
+    // Leaves end - copper bulb 1 gt delay
 
     public void checkAndFlip(final BlockState state, final ServerLevel level, final BlockPos pos) {
         boolean signal = level.hasNeighborSignal(pos);
