From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bluemangoo <chenfy2006@qq.com>
Date: Fri, 19 Jul 2024 15:04:03 +0800
Subject: [PATCH] Disable crystal-portal proximity check


diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index 1eca60c5a3e1eeb6f44edbb81fe1af011edd8732..873361c11e013ecc580b769f201247f7f0016a3a 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -51,7 +51,7 @@ public class EndCrystalItem extends Item {
                         level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, above);
                         EnderDragonFight fight = ((ServerLevel)level).getDragonFight();
                         if (fight != null) {
-                            fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup
+                            fight.tryRespawn(); // fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup // Leaves - disable crystal-portal proximity check
                         }
                     }
 
diff --git a/net/minecraft/world/level/dimension/end/EnderDragonFight.java b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
index 7f0f5ce66dd0bc4a6e86d0534546c402c601fde2..fcb532263732232dbc225fa8dbf070cf41096191 100644
--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -554,12 +554,14 @@ public class EnderDragonFight extends SavedData {
     }
 
     public boolean tryRespawn() { // CraftBukkit - return boolean
+        /* Leaves - disable crystal-portal proximity check
         // Paper start - Perf: Do crystal-portal proximity check before entity lookup
         return this.tryRespawn(null);
     }
 
     public boolean tryRespawn(@Nullable final BlockPos placedEndCrystalPos) { // placedEndCrystalPos is null if the tryRespawn() call was not caused by a placed end crystal
         // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+         */
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos location = this.exitPortalLocation;
             if (location == null) {
@@ -574,6 +576,7 @@ public class EnderDragonFight extends SavedData {
 
                 location = this.exitPortalLocation;
             }
+            /* Leaves - disable crystal-portal proximity check
             // Paper start - Perf: Do crystal-portal proximity check before entity lookup
             if (placedEndCrystalPos != null && !this.level.paperConfig().misc.allowRemoteEnderDragonRespawning) {
                 int dy = placedEndCrystalPos.getY() - location.getY();
@@ -588,6 +591,7 @@ public class EnderDragonFight extends SavedData {
                 }
             }
             // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+             */
 
             List<EndCrystal> crystals = Lists.newArrayList();
             BlockPos center = location.above(1);
