From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 30 Oct 2023 10:43:44 +0800
Subject: [PATCH] Make Item tick vanilla


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 41050491bde33981448b3da0290f8edfaee82c30..80f9c003b125a3af29c1456b6f796a43954a9aad 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -126,6 +126,9 @@ public class ItemEntity extends Entity implements TraceableEntity {
     // Paper start - EAR 2
     @Override
     public void inactiveTick() {
+        // Leaves start - vanilla
+        this.tick();
+        /*
         super.inactiveTick();
         if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
             this.pickupDelay--;
@@ -143,6 +146,8 @@ public class ItemEntity extends Entity implements TraceableEntity {
             // CraftBukkit end
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
+         */
+        // Leaves end - vanilla
     }
     // Paper end - EAR 2
 
