From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 12:09:16 +0800
Subject: [PATCH] Copper Bulb 1 gt delay


diff --git a/net/minecraft/world/level/block/CopperBulbBlock.java b/net/minecraft/world/level/block/CopperBulbBlock.java
index 79ce45289d9ab81fb2cd7fdf1b32042e84a01a50..32b774ddee171387d79f7da34520996622daefa5 100644
--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -33,14 +33,26 @@ public class CopperBulbBlock extends Block {
     @Override
     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
         if (oldState.getBlock() != state.getBlock() && level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt delay
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston) {
         if (level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
@@ -57,6 +69,15 @@ public class CopperBulbBlock extends Block {
         }
     }
 
+    // Leaves start - copper bulb 1 gt delay
+    @Override
+    public void tick(BlockState state, ServerLevel world, BlockPos pos, net.minecraft.util.RandomSource random) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.copperBulb1gt) {
+            checkAndFlip(state, world, pos);
+        }
+    }
+    // Leaves end - copper bulb 1 gt delay
+
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(LIT, POWERED);
