From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 13:44:36 +0800
Subject: [PATCH 01/11] Config to disable method profiler


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 0b18251dfbfda8ee540f40713ec8ba38956241c5..97870987cd756606f86058ba89322daeec6ab8e6 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1368,7 +1368,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
                 // Paper - improve tick loop - done above
 
-                try (Profiler.Scope ignored = Profiler.use(this.createProfiler())) {
+                try (Profiler.Scope ignored = org.leavesmc.leaves.LeavesConfig.mics.disableMethodProfiler ? null : Profiler.use(this.createProfiler())) { // Leaves - disable profiler
                     this.processPacketsAndTick(sprinting);
                     ProfilerFiller profiler = Profiler.get();
                     profiler.push("nextTickWait");
diff --git a/net/minecraft/util/profiling/Profiler.java b/net/minecraft/util/profiling/Profiler.java
index 2a43ad5e4a16b6d90c1d5a791ef2330e7a7f2c79..325f71d4a5736294dd98c4195f521171bde6bda0 100644
--- a/net/minecraft/util/profiling/Profiler.java
+++ b/net/minecraft/util/profiling/Profiler.java
@@ -45,6 +45,7 @@ public final class Profiler {
     }
 
     public static ProfilerFiller get() {
+        if (org.leavesmc.leaves.LeavesConfig.mics.disableMethodProfiler) return InactiveProfiler.INSTANCE; // Leaves - disable profiler
         return ACTIVE_COUNT.get() == 0 ? getDefaultFiller() : Objects.requireNonNullElseGet(ACTIVE.get(), Profiler::getDefaultFiller);
     }
 
