From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 4 Feb 2025 20:27:46 +0800
Subject: [PATCH] Villager infinite discounts


diff --git a/net/minecraft/world/entity/ai/gossip/GossipType.java b/net/minecraft/world/entity/ai/gossip/GossipType.java
index 32548e375202452c5f5ec89fec8a82d62252e823..b8a05c113e8b228c904e952f6ab2b4c38694dc3e 100644
--- a/net/minecraft/world/entity/ai/gossip/GossipType.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipType.java
@@ -15,9 +15,9 @@ public enum GossipType implements StringRepresentable {
     public static final int REPUTATION_CHANGE_PER_TRADE = 2;
     public final String id;
     public final int weight;
-    public final int max;
+    public int max; // Leaves - not final
     public final int decayPerDay;
-    public final int decayPerTransfer;
+    public int decayPerTransfer; // Leaves - not final
     public static final Codec<GossipType> CODEC = StringRepresentable.fromEnum(GossipType::values);
 
     private GossipType(final String id, final int weight, final int max, final int decayPerDay, final int decayPerTransfer) {
