From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 14:15:54 +0800
Subject: [PATCH] Crafter 1 gt delay


diff --git a/net/minecraft/world/level/block/CrafterBlock.java b/net/minecraft/world/level/block/CrafterBlock.java
index 3faa541c4a4623e915269d6241e1a3a3c6e374ff..468a39219d82a665b99ac1768435557d6f3326c9 100644
--- a/net/minecraft/world/level/block/CrafterBlock.java
+++ b/net/minecraft/world/level/block/CrafterBlock.java
@@ -75,7 +75,7 @@ public class CrafterBlock extends BaseEntityBlock {
         boolean triggeredValue = state.getValue(TRIGGERED);
         BlockEntity blockEntity = level.getBlockEntity(pos);
         if (hasNeighborSignal && !triggeredValue) {
-            level.scheduleTick(pos, this, 4);
+            level.scheduleTick(pos, this, !org.leavesmc.leaves.LeavesConfig.modify.oldMC.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
             level.setBlock(pos, state.setValue(TRIGGERED, true), Block.UPDATE_CLIENTS);
             this.setBlockEntityTriggered(blockEntity, true);
         } else if (!hasNeighborSignal && triggeredValue) {
@@ -124,7 +124,7 @@ public class CrafterBlock extends BaseEntityBlock {
     @Override
     public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack stack) {
         if (state.getValue(TRIGGERED)) {
-            level.scheduleTick(pos, this, 4);
+            level.scheduleTick(pos, this, !org.leavesmc.leaves.LeavesConfig.modify.oldMC.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
         }
     }
 
