From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 5 Feb 2025 23:11:32 +0800
Subject: [PATCH] Servux Protocol


diff --git a/net/minecraft/server/ServerTickRateManager.java b/net/minecraft/server/ServerTickRateManager.java
index 4257c47d33e3773fac766817138cc21c55d9ab76..8638f3f7aa853802a60ddeb681d089e0c6143693 100644
--- a/net/minecraft/server/ServerTickRateManager.java
+++ b/net/minecraft/server/ServerTickRateManager.java
@@ -129,4 +129,10 @@ public class ServerTickRateManager extends TickRateManager {
         player.connection.send(ClientboundTickingStatePacket.from(this));
         player.connection.send(ClientboundTickingStepPacket.from(this));
     }
+
+    // Leaves start - servux
+    public long getRemainingSprintTicks() {
+        return remainingSprintTicks;
+    }
+    // Leaves end - servux
 }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 499714800b2160cf163fc7a37e287f655a9d80ee..81094f0ab63dfb48f7ae9a529d407dec5e356626 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2316,9 +2316,21 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         // Paper end
     }
 
+
     @Override
     public LevelData.RespawnData getRespawnData() {
         return this.getServer().getRespawnData();
+        // TODO: 1.21.9 removed spawn chunk, should we keep this?
+        // int i = this.getGameRules().getInt(GameRules.RULE_SPAWN_CHUNK_RADIUS) + 1;
+        // if (i > 1) {
+        //     // Paper start - allow disabling gamerule limits
+        //     for (ChunkPos chunkPos : io.papermc.paper.util.MCUtil.getSpiralOutChunks(pos, i - 2)) {
+        //         this.getChunkSource().addTicketAtLevel(TicketType.START, chunkPos, ChunkLevel.ENTITY_TICKING_LEVEL);
+        //     }
+        //     // Paper end - allow disabling gamerule limits
+        // }
+        // this.lastSpawnChunkRadius = i;
+        // org.leavesmc.leaves.protocol.servux.ServuxHudDataProtocol.refreshSpawnMetadata = true; // Leaves - servux
     }
 
     public LongSet getForceLoadedChunks() {
