From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 13:51:16 +0800
Subject: [PATCH] Chat command max length


diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
index 491af2413a4e793a121fec368259ff8211ed031e..0c1e4bba284d15cf9ea7178093460630c4458a22 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
@@ -12,11 +12,11 @@ public record ServerboundChatCommandPacket(String command) implements Packet<Ser
     );
 
     private ServerboundChatCommandPacket(final FriendlyByteBuf input) {
-        this(input.readUtf(MAX_CHAT_PACKET_INPUT_SIZE)); // Paper - limit chat command inputs
+        this(input.readUtf(org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength)); // Paper - limit chat command inputs // Leaves - chat command max length
     }
 
     private void write(final FriendlyByteBuf output) {
-        output.writeUtf(this.command);
+        output.writeUtf(this.command, org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength); // Leaves - chat command max length
     }
 
     @Override
diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
index 646e86b6a278a5cd3363e24ed9668359d3b250fd..0994cd31f894000450ba6dcb84586ba066b38bf3 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -16,11 +16,11 @@ public record ServerboundChatCommandSignedPacket(
     );
 
     private ServerboundChatCommandSignedPacket(final FriendlyByteBuf input) {
-        this(input.readUtf(net.minecraft.network.protocol.game.ServerboundChatCommandPacket.MAX_CHAT_PACKET_INPUT_SIZE), input.readInstant(), input.readLong(), new ArgumentSignatures(input), new LastSeenMessages.Update(input)); // Paper - limit chat command inputs
+        this(input.readUtf(org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength), input.readInstant(), input.readLong(), new ArgumentSignatures(input), new LastSeenMessages.Update(input)); // Paper - limit chat command inputs // Leaves - chat command max length
     }
 
     private void write(final FriendlyByteBuf output) {
-        output.writeUtf(this.command);
+        output.writeUtf(this.command, org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength); // Leaves - chat command max length
         output.writeInstant(this.timeStamp);
         output.writeLong(this.salt);
         this.argumentSignatures.write(output);
