From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 22 Jan 2024 11:15:53 +0800
Subject: [PATCH] Renewable deepslate


diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index 1f62dcf2709d430e996790750a691b12938eb09a..2d6f4579fa27b0f9117abf13671a47576fffb155 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -196,7 +196,7 @@ public class LiquidBlock extends Block implements BucketPickup {
             for (Direction direction : POSSIBLE_FLOW_DIRECTIONS) {
                 BlockPos blockPos = pos.relative(direction.getOpposite());
                 if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
-                    Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                    Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : (org.leavesmc.leaves.LeavesConfig.modify.renewableDeepslate && level.dimension() == Level.OVERWORLD && pos.getY() < 0 ? Blocks.COBBLED_DEEPSLATE : Blocks.COBBLESTONE); // Leaves - renewable deepslate
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState(), Block.UPDATE_ALL)) {
                         this.fizz(level, pos);
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index a42878c191445ba4b0d50dd2b789a132d5cf6a3e..69ea51ee111b9595a234c9f8177f35147be35b94 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -224,7 +224,7 @@ public abstract class LavaFluid extends FlowingFluid {
             if (this.is(FluidTags.LAVA) && fluidState1.is(FluidTags.WATER)) {
                 if (state.getBlock() instanceof LiquidBlock) {
                     // CraftBukkit start
-                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level.getMinecraftWorld(), pos, Blocks.STONE.defaultBlockState(), Block.UPDATE_ALL)) {
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level.getMinecraftWorld(), pos, org.leavesmc.leaves.LeavesConfig.modify.renewableDeepslate && level.getMinecraftWorld().dimension() == Level.OVERWORLD && pos.getY() < 0 ? Blocks.DEEPSLATE.defaultBlockState() : Blocks.STONE.defaultBlockState(), Block.UPDATE_ALL)) { // Leaves - renewable deepslate
                         return;
                     }
                     // CraftBukkit end
