From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 5 Feb 2025 23:42:22 +0800
Subject: [PATCH] Renewable coral


diff --git a/net/minecraft/world/level/block/CoralFanBlock.java b/net/minecraft/world/level/block/CoralFanBlock.java
index 2a48b215fde808fdc6b36a5d8fe88c7ae56c9567..4aa97ca825604478e2a86e655c6cf6cb4e0db336 100644
--- a/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/net/minecraft/world/level/block/CoralFanBlock.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 
-public class CoralFanBlock extends BaseCoralFanBlock {
+public class CoralFanBlock extends BaseCoralFanBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralFanBlock> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(coralFanBlock -> coralFanBlock.deadBlock), propertiesCodec())
             .apply(instance, CoralFanBlock::new)
@@ -69,4 +69,11 @@ public class CoralFanBlock extends BaseCoralFanBlock {
             return super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
         }
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/block/CoralPlantBlock.java b/net/minecraft/world/level/block/CoralPlantBlock.java
index b66a3e6e109cab913ff2593accaea04110d66094..128f768ec5c194b697796ac7f2f8b55229f5a4be 100644
--- a/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -16,7 +16,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
+public class CoralPlantBlock extends BaseCoralPlantTypeBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralPlantBlock> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(coralPlantBlock -> coralPlantBlock.deadBlock), propertiesCodec())
             .apply(instance, CoralPlantBlock::new)
@@ -78,4 +78,12 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
         return SHAPE;
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED
+            || org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.TRUE;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/block/CoralWallFanBlock.java b/net/minecraft/world/level/block/CoralWallFanBlock.java
index 2d34d3e7a39a27b1756573d16ec0f8878ab8f620..b702c69f881a56a8d9cd48cf017ea8c0392720fd 100644
--- a/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 
-public class CoralWallFanBlock extends BaseCoralWallFanBlock {
+public class CoralWallFanBlock extends BaseCoralWallFanBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralWallFanBlock> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(coralWallFanBlock -> coralWallFanBlock.deadBlock), propertiesCodec())
             .apply(instance, CoralWallFanBlock::new)
@@ -69,4 +69,11 @@ public class CoralWallFanBlock extends BaseCoralWallFanBlock {
             return super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
         }
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java b/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
index 19c810d820a559afdc8fac399e4021121063c2ff..d848f5e19347e69776b99d9feade002d66774513 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
@@ -17,7 +17,7 @@ public class CoralClawFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) {
+    public boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) { // Leaves - protected -> public
         if (!this.placeCoralBlock(level, random, pos, state)) {
             return false;
         } else {
diff --git a/net/minecraft/world/level/levelgen/feature/CoralFeature.java b/net/minecraft/world/level/levelgen/feature/CoralFeature.java
index 53e2f2f05368949365b9690fc16bc21efabb50bb..b8475b55b3d7aa65125ecf9460a5c9c4e1261225 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralFeature.java
@@ -31,7 +31,7 @@ public abstract class CoralFeature extends Feature<NoneFeatureConfiguration> {
         return !optional.isEmpty() && this.placeFeature(worldGenLevel, randomSource, blockPos, optional.get().defaultBlockState());
     }
 
-    protected abstract boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state);
+    public abstract boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state); // Leaves - protected -> public
 
     protected boolean placeCoralBlock(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) {
         BlockPos blockPos = pos.above();
diff --git a/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
index cd550c704cbe46727caf965d7910d2a1d55f5697..29c5c8ed793834dd5bbafb89e48670a30dc8184d 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
@@ -14,7 +14,7 @@ public class CoralMushroomFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) {
+    public boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) { // Leaves - protected -> public
         int i = random.nextInt(3) + 3;
         int i1 = random.nextInt(3) + 3;
         int i2 = random.nextInt(3) + 3;
diff --git a/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java b/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
index 521dbd30e9fb7a366d534fe6a952b0e3a8ec8449..13a0c047c86563d0f28866d513c52999bfce6fd1 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
@@ -15,7 +15,7 @@ public class CoralTreeFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) {
+    public boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) { // Leaves - protected -> public
         BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
         int i = random.nextInt(3) + 1;
 
