From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 14:08:24 +0800
Subject: [PATCH] Leaves Extra Yggdrasil Service


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index cf6152c5eaadf2586b00cad19b56401372f8bde0..98bb3d3a80d0c97b85dea2aa955a15ad4dc9e9c9 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -179,7 +179,7 @@ public class Main {
                 universePath = new File(bukkitConfiguration.getString("settings.world-container", "."));
             }
             // Paper end - fix SPIGOT-5824
-            Services services = Services.create(new com.destroystokyo.paper.profile.PaperAuthenticationService(Proxy.NO_PROXY), universePath, userCacheFile, options); // Paper - pass OptionSet to load paper config files; override authentication service; fix world-container
+            Services services = Services.create(new org.leavesmc.leaves.profile.LeavesAuthenticationService(Proxy.NO_PROXY), universePath, userCacheFile, options); // Paper - pass OptionSet to load paper config files; override authentication service; fix world-container // Leaves - extra-yggdrasil-service
             String levelName = Optional.ofNullable((String) options.valueOf("world")).orElse(settings.getProperties().levelName); // CraftBukkit
             LevelStorageSource levelStorageSource = LevelStorageSource.createDefault(universePath.toPath());
             LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(levelName);
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 1d5ef68c13ff9dda71e428210b905b38b531394f..bcaab44a2420fcfe5ec0852ec66ea0a4aaf74f71 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -256,7 +256,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     private boolean isDemo;
     private volatile boolean isReady;
     private long lastOverloadWarningNanos;
-    protected final Services services;
+    public final Services services; // Leaves - protected -> public
     private final NotificationManager notificationManager;
     private final ServerActivityMonitor serverActivityMonitor;
     private long lastServerStatus;
