From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 14:17:25 +0800
Subject: [PATCH] Redstone ignore upwards update


diff --git a/net/minecraft/world/level/block/ComparatorBlock.java b/net/minecraft/world/level/block/ComparatorBlock.java
index 0d8b5f723db5ae5b4e4ca88661ed2372c3888d06..36e91d5702dd5f6d49c563ee28eca150163f5c52 100644
--- a/net/minecraft/world/level/block/ComparatorBlock.java
+++ b/net/minecraft/world/level/block/ComparatorBlock.java
@@ -59,6 +59,7 @@ public class ComparatorBlock extends DiodeBlock implements EntityBlock {
         final BlockState neighbourState,
         final RandomSource random
     ) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneIgnoreUpwardsUpdate && directionToNeighbour == Direction.DOWN) return state; // Leaves - behavior to 1.20.1
         return directionToNeighbour == Direction.DOWN && !this.canSurviveOn(level, neighbourPos, neighbourState)
             ? Blocks.AIR.defaultBlockState()
             : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index ed471efa3ab705de62aec81c3940c1b9c5acdbc5..aba6646fc8eea944bcaa668c626279b96427e57f 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -179,7 +179,7 @@ public class RedStoneWireBlock extends Block {
         final RandomSource random
     ) {
         if (directionToNeighbour == Direction.DOWN) {
-            return !this.canSurviveOn(level, neighbourPos, neighbourState) ? Blocks.AIR.defaultBlockState() : state;
+            return org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneIgnoreUpwardsUpdate ? state : !this.canSurviveOn(level, neighbourPos, neighbourState) ? Blocks.AIR.defaultBlockState() : state; // Leaves - behavior to 1.19
         } else if (directionToNeighbour == Direction.UP) {
             return this.getConnectionState(level, state, pos);
         } else {
@@ -242,7 +242,7 @@ public class RedStoneWireBlock extends Block {
         BlockPos relativePos = pos.relative(direction);
         BlockState relativeState = level.getBlockState(relativePos);
         if (canConnectUp) {
-            boolean isPlaceableAbove = relativeState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, relativePos, relativeState);
+            boolean isPlaceableAbove = (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneIgnoreUpwardsUpdate && relativeState.getBlock() instanceof TrapDoorBlock) || this.canSurviveOn(level, relativePos, relativeState); // Leaves - behavior to 1.19
             if (isPlaceableAbove && shouldConnectTo(level.getBlockState(relativePos.above()))) {
                 if (relativeState.isFaceSturdy(level, relativePos, direction.getOpposite())) {
                     return RedstoneSide.UP;
diff --git a/net/minecraft/world/level/block/RepeaterBlock.java b/net/minecraft/world/level/block/RepeaterBlock.java
index d301ca4c91e75d715998aa7ed155fd44ccd64785..3398895b776f12090f9a0f7e5963339bd048dab1 100644
--- a/net/minecraft/world/level/block/RepeaterBlock.java
+++ b/net/minecraft/world/level/block/RepeaterBlock.java
@@ -70,6 +70,7 @@ public class RepeaterBlock extends DiodeBlock {
         final BlockState neighbourState,
         final RandomSource random
     ) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.redstoneIgnoreUpwardsUpdate && directionToNeighbour == Direction.DOWN) return state; // Leaves - behavior to 1.20.1
         if (directionToNeighbour == Direction.DOWN && !this.canSurviveOn(level, neighbourPos, neighbourState)) {
             return Blocks.AIR.defaultBlockState();
         } else {
