From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 5 Feb 2025 23:11:32 +0800
Subject: [PATCH] Servux Protocol


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 76f056351208432e78717ce41adf3a85ff6a6c57..a04da8e7b7619324a29f448a2dc32b02e9d00995 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2208,6 +2208,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         }
 
         this.lastSpawnChunkRadius = i;
+        org.leavesmc.leaves.protocol.servux.ServuxHudDataProtocol.refreshSpawnMetadata = true; // Leaves - servux
     }
 
     public LongSet getForceLoadedChunks() {
