From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 14:26:37 +0800
Subject: [PATCH 11/11] Old Throwable Projectile tick order


diff --git a/net/minecraft/world/entity/projectile/ThrowableProjectile.java b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
index cbcc8679cf4854c50f17c4a6fb87a34fceb7339d..f253054dcc8746cb2b71f8d908b06065b31483c7 100644
--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -45,6 +45,26 @@ public abstract class ThrowableProjectile extends Projectile {
     @Override
     public void tick() {
         this.handleFirstTickBubbleColumn();
+        // Leaves start - old-throwable-projectile-tick-order
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldThrowableProjectileTickOrder) {
+            super.tick();
+            HitResult hitResultOnMoveVector = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
+            if (hitResultOnMoveVector.getType() != HitResult.Type.MISS) {
+                this.preHitTargetOrDeflectSelf(hitResultOnMoveVector); // CraftBukkit - projectile hit event
+            }
+
+            Vec3 oldDelta = this.getDeltaMovement();
+            double nx = this.getX() + oldDelta.x;
+            double ny = this.getY() + oldDelta.y;
+            double nz = this.getZ() + oldDelta.z;
+            this.updateRotation();
+            this.applyInertia();
+            this.applyGravity();
+            this.setPos(nx, ny, nz);
+            this.applyEffectsFromBlocks();
+            return;
+        }
+        // Leaves end - old-throwable-projectile-tick-order
         this.applyGravity();
         this.applyInertia();
         HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
