From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Sat, 20 Jan 2024 02:50:56 +0800
Subject: [PATCH] Placing locked hopper no longer send NC updates


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index ea25b29a5f28a171151cd1ca185a8fc0ceaa248a..c5b7b97d54380084ca91a4e14310c855f000c63d 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1201,7 +1201,11 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
             }
 
             if ((flags & 1) != 0) {
-                this.updateNeighborsAt(pos, blockState.getBlock());
+                // Leaves start - Placing locked hopper doesn't send NC updates.
+                if (!(blockState.getBlock() instanceof net.minecraft.world.level.block.HopperBlock) || blockState.getValue(net.minecraft.world.level.block.HopperBlock.ENABLED)) {
+                    this.updateNeighborsAt(pos, blockState.getBlock());
+                }
+                // Leaves end - Placing locked hopper doesn't send NC updates.
                 if (!this.isClientSide && state.hasAnalogOutputSignal()) {
                     this.updateNeighbourForOutputSignal(pos, newState.getBlock());
                 }
