From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:22 +0800
Subject: [PATCH] Bytebuf API


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 7b5b9d0a5a99108ba5853891918565a22aaa8efd..88223de2f1089646b271d4e7f22b5bdc269e8748 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -294,6 +294,13 @@ public abstract class PlayerList {
         boolean flag = gameRules.get(GameRules.IMMEDIATE_RESPAWN);
         boolean flag1 = gameRules.get(GameRules.REDUCED_DEBUG_INFO);
         boolean flag2 = gameRules.get(GameRules.LIMITED_CRAFTING);
+
+        // Leaves start - Bytebuf API
+        if (!(player instanceof org.leavesmc.leaves.bot.ServerBot) && !(player instanceof org.leavesmc.leaves.replay.ServerPhotographer)) {
+            this.cserver.getBytebufHandler().injectPlayer(player);
+        }
+        // Leaves end - Bytebuf API
+
         serverGamePacketListenerImpl.send(
             new ClientboundLoginPacket(
                 player.getId(),
