From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:22 +0800
Subject: [PATCH] Bytebuf API


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 7b5b9d0a5a99108ba5853891918565a22aaa8efd..374bfdee6332bfd5d38fdb7345bcc3249b4fc5c2 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -290,6 +290,13 @@ public abstract class PlayerList {
             serverGamePacketListenerImpl
         );
         serverGamePacketListenerImpl.suspendFlushing();
+
+        // Leaves start - Bytebuf API
+        if (!(player instanceof org.leavesmc.leaves.bot.ServerBot) && !(player instanceof org.leavesmc.leaves.replay.ServerPhotographer)) {
+            org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.updatePlayer(player);
+        }
+        // Leaves end - Bytebuf API
+
         GameRules gameRules = serverLevel.getGameRules();
         boolean flag = gameRules.get(GameRules.IMMEDIATE_RESPAWN);
         boolean flag1 = gameRules.get(GameRules.REDUCED_DEBUG_INFO);
