From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:22 +0800
Subject: [PATCH] Bytebuf API


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 95cb24f44412e2aad5f15f8256e3506366c1412d..d4eb10d3734930a43b8270cdc7eb78bf40907c94 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -384,6 +384,13 @@ public abstract class PlayerList {
             boolean _boolean = gameRules.getBoolean(GameRules.RULE_DO_IMMEDIATE_RESPAWN);
             boolean _boolean1 = gameRules.getBoolean(GameRules.RULE_REDUCEDDEBUGINFO);
             boolean _boolean2 = gameRules.getBoolean(GameRules.RULE_LIMITED_CRAFTING);
+
+            // Leaves start - Bytebuf API
+            if (!(player instanceof org.leavesmc.leaves.bot.ServerBot) && !(player instanceof org.leavesmc.leaves.replay.ServerPhotographer)) {
+                this.cserver.getBytebufHandler().injectPlayer(player);
+            }
+            // Leaves end - Bytebuf API
+
             serverGamePacketListenerImpl.send(
                 new ClientboundLoginPacket(
                     player.getId(),
