From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 22 Jul 2024 18:30:42 +0800
Subject: [PATCH] Can disable LivingEntity aiStep alive check


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 67c29508c177449ce33fede89c7a9c3d9e8c6a6f..10676027d8ba2d55319e66f808f9a7854b122c4e 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3317,7 +3317,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
         }
 
-        if (!this.isRemoved()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableLivingEntityAiStepAliveCheck || !this.isRemoved()) { // Leaves - can disable check
             this.aiStep();
         }
 
