From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+lumine1909@users.noreply.github.com>
Date: Wed, 26 Jun 2024 17:59:56 +0800
Subject: [PATCH] Allow grindstone overstacking


diff --git a/net/minecraft/world/inventory/GrindstoneMenu.java b/net/minecraft/world/inventory/GrindstoneMenu.java
index f85bd2a90c2694d96f67cc3701a9bbf081fe8475..9855ccd18f325d7fa31c7f11532c03ed10776846 100644
--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -178,7 +178,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
             int i2 = i + i1 + max * 5 / 100;
             int i3 = 1;
             if (!inputItem.isDamageableItem()) {
-                if (inputItem.getMaxStackSize() < 2 || !ItemStack.matches(inputItem, additionalItem)) {
+                if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && inputItem.getMaxStackSize() < 2 || !ItemStack.matches(inputItem, additionalItem)) { // Leaves - allowGrindstoneOverstaking
                     return ItemStack.EMPTY;
                 }
 
@@ -247,7 +247,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
             ItemStack item1 = this.repairSlots.getItem(0);
             ItemStack item2 = this.repairSlots.getItem(1);
             if (index == 2) {
-                if (!this.moveItemStackTo(item, 3, 39, true)) {
+                if (!this.moveItemStackTo(item, 3, 39, true, org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking)) { // Leaves - allowGrindstoneOverstacking: Disable stack check
                     return ItemStack.EMPTY;
                 }
 
