From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 16:21:58 +0800
Subject: [PATCH] Bedrock break list


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index bcaab44a2420fcfe5ec0852ec66ea0a4aaf74f71..7723281c7b5605c520da8d86b8f7843ef4c2b504 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1867,6 +1867,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         profiler.popPush("server gui refresh");
 
+        org.leavesmc.leaves.util.BreakBedrockList.endTick(); // Leaves - break bedrock list
+
         for (Runnable tickable : this.tickables) {
             tickable.run();
         }
diff --git a/net/minecraft/server/ServerScoreboard.java b/net/minecraft/server/ServerScoreboard.java
index 9bf09b888425e491072864265c8de3885154a0ba..4e400a24c0a5968223c7cd4874bd4d1e04e9fa2b 100644
--- a/net/minecraft/server/ServerScoreboard.java
+++ b/net/minecraft/server/ServerScoreboard.java
@@ -164,6 +164,7 @@ public class ServerScoreboard extends Scoreboard {
     @Override
     public void onObjectiveAdded(final Objective objective) {
         super.onObjectiveAdded(objective);
+        org.leavesmc.leaves.util.BreakBedrockList.onScoreboardAdd(objective); // Leaves - break bedrock list
         this.setDirty();
     }
 
@@ -180,6 +181,7 @@ public class ServerScoreboard extends Scoreboard {
     @Override
     public void onObjectiveRemoved(final Objective objective) {
         super.onObjectiveRemoved(objective);
+        org.leavesmc.leaves.util.BreakBedrockList.onScoreboardRemove(objective); // Leaves - break bedrock list
         if (this.trackedObjectives.contains(objective)) {
             this.stopTrackingObjective(objective);
         }
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index a2bc07a8348c5d7f30aafa0344d66715c89994b0..c0847fcd32c90c9b4d6c268eb85506a90695ba69 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -75,6 +75,11 @@ public class PistonBaseBlock extends DirectionalBlock {
     @Override
     public void setPlacedBy(final Level level, final BlockPos pos, final BlockState state, final @Nullable LivingEntity by, final ItemStack itemStack) {
         if (!level.isClientSide()) {
+            // Leaves start - break bedrock list
+            if (by instanceof net.minecraft.world.entity.player.Player player) {
+                org.leavesmc.leaves.util.BreakBedrockList.onPlayerPlacePiston(level, player, pos);
+            }
+            // Leaves end - break bedrock list
             this.checkIfExtend(level, pos, state);
         }
     }
@@ -229,7 +234,13 @@ public class PistonBaseBlock extends DirectionalBlock {
                             }
                         }
                         // Paper end - Fix sticky pistons and BlockPistonRetractEvent
-                        level.removeBlock(pos.relative(direction), false);
+                        // Leaves start - break bedrock list
+                        BlockPos bedrockBreakPos = pos.relative(direction);
+                        if (level.getBlockState(bedrockBreakPos).getBlock() == Blocks.BEDROCK) {
+                            org.leavesmc.leaves.util.BreakBedrockList.onPistonBreakBedrock(level, bedrockBreakPos);
+                        }
+                        level.removeBlock(bedrockBreakPos, false);
+                        // Leaves end - break bedrock list
                     } else {
                         this.moveBlocks(level, pos, direction, false);
                     }
@@ -238,6 +249,11 @@ public class PistonBaseBlock extends DirectionalBlock {
                 // Paper start - Protect Bedrock and End Portal/Frames from being destroyed; fix headless pistons breaking blocks
                 BlockPos headPos = pos.relative(direction);
                 if (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits || level.getBlockState(headPos) == Blocks.PISTON_HEAD.defaultBlockState().setValue(FACING, direction)) { // double check to make sure we're not a headless piston
+                    // Leaves start - break bedrock list
+                    if (level.getBlockState(headPos).getBlock() == Blocks.BEDROCK) {
+                        org.leavesmc.leaves.util.BreakBedrockList.onPistonBreakBedrock(level, headPos);
+                    }
+                    // Leaves end - break bedrock list
                     level.removeBlock(headPos, false);
                 } else {
                     ((ServerLevel) level).getChunkSource().blockChanged(headPos); // ... fix client desync
