From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Sat, 26 Oct 2024 19:53:17 +0800
Subject: [PATCH] Allow anvil destroy item entities


diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index a5ce7877daa70cc5b3120d56fd50a57c5b7f2ea4..3faabf37f41688f51d17dc177e52078025879ca7 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -280,7 +280,7 @@ public class FallingBlockEntity extends Entity {
             if (ceil < 0) {
                 return false;
             } else {
-                Predicate<Entity> predicate = EntitySelector.NO_CREATIVE_OR_SPECTATOR.and(EntitySelector.LIVING_ENTITY_STILL_ALIVE);
+                Predicate<Entity> predicate = org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowAnvilDestroyItemEntities ? EntitySelector.NO_CREATIVE_OR_SPECTATOR : EntitySelector.NO_CREATIVE_OR_SPECTATOR.and(EntitySelector.LIVING_ENTITY_STILL_ALIVE); // Leaves - Allow anvil destroy item entities
                 DamageSource damageSource1 = this.blockState.getBlock() instanceof Fallable fallable
                     ? fallable.getFallDamageSource(this)
                     : this.damageSources().fallingBlock(this);
