From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 19:16:16 +0800
Subject: [PATCH] No block update command


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index f7390b91fe1866c6a11beb93711d0fcde4293164..b208d9a8af0ba0160651e23a2eed41f55a8f7d18 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -391,7 +391,7 @@ public class ServerPlayerGameMode {
                 org.bukkit.block.BlockState bState = bblock.getState();
                 this.level.captureDrops = new java.util.ArrayList<>();
                 // CraftBukkit end
-                BlockState adjustedState = block.playerWillDestroy(this.level, pos, state, this.player);
+                BlockState adjustedState = org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate() ? state : block.playerWillDestroy(this.level, pos, state, this.player); // Leaves - no block update
                 boolean changed = this.level.removeBlock(pos, false);
                 if (SharedConstants.DEBUG_BLOCK_BREAK) {
                     LOGGER.info("server broke {} {} -> {}", pos, adjustedState, this.level.getBlockState(pos));
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index e6a1600546efd739b7139718fb2042188d24662d..bf0c4e8e4828d82e91fbafc8f428aa14c00c9cc6 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1054,7 +1054,12 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
 
     @Override
-    public boolean setBlock(final BlockPos pos, final BlockState blockState, @Block.UpdateFlags final int updateFlags, final int updateLimit) {
+    public boolean setBlock(final BlockPos pos, final BlockState blockState, @Block.UpdateFlags int updateFlags, final int updateLimit) { // Leaves - remove final on updateFlags for no-block-update
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            updateFlags = updateFlags & ~Block.UPDATE_NEIGHBORS | Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
@@ -1156,6 +1161,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         BlockState state = newState;
         BlockState blockState = oldState;
         BlockState blockState1 = currentState;
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            flags = flags & ~Block.UPDATE_NEIGHBORS | Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         if (blockState1 == state) {
             if (blockState != blockState1) {
                 this.setBlocksDirty(pos, blockState, blockState1);
@@ -1178,7 +1188,12 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
             if ((flags & Block.UPDATE_KNOWN_SHAPE) == 0 && recursionLeft > 0) {
                 int i = flags & ~(Block.UPDATE_SUPPRESS_DROPS | Block.UPDATE_NEIGHBORS);
-
+                // Leaves start - no block update
+                if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+                    this.updatePOIOnBlockStateChange(pos, blockState, blockState1);
+                    return;
+                }
+                // Leaves end - no block update
                 // CraftBukkit start
                 blockState.updateIndirectNeighbourShapes(this, pos, i, recursionLeft - 1); // Don't call an event for the old block to limit event spam
                 boolean cancelledUpdates = false; // Paper - Fix block place logic
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index c0847fcd32c90c9b4d6c268eb85506a90695ba69..7d2b8864d21c350327e56114cac39ff1e1dac888 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -108,6 +108,7 @@ public class PistonBaseBlock extends DirectionalBlock {
     }
 
     private void checkIfExtend(final Level level, final BlockPos pos, final BlockState state) {
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) return; // Leaves - no block update
         Direction direction = state.getValue(FACING);
         boolean extend = this.getNeighborSignal(level, pos, direction);
         if (extend && !state.getValue(EXTENDED)) {
diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 353085151026b0d695be3c35d0638ecd39a962ce..e3487d2103cd7eb3f6d1a487ecd015fbcf76df8b 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -39,9 +39,14 @@ public interface NeighborUpdater {
         final BlockPos pos,
         final BlockPos neighborPos,
         final BlockState neighborState,
-        @Block.UpdateFlags final int updateFlags,
+        @Block.UpdateFlags int updateFlags, // Leaves - remove final for no-block-update
         final int updateLimit
     ) {
+        // Leaves start - no block update
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) {
+            updateFlags = updateFlags & ~net.minecraft.world.level.block.Block.UPDATE_NEIGHBORS | net.minecraft.world.level.block.Block.UPDATE_SKIP_ON_PLACE;
+        }
+        // Leaves end - no block update
         BlockState currentState = level.getBlockState(pos);
         if ((updateFlags & Block.UPDATE_SKIP_SHAPE_UPDATE_ON_WIRE) == 0 || !currentState.is(Blocks.REDSTONE_WIRE)) {
             try {
@@ -72,6 +77,7 @@ public interface NeighborUpdater {
 
     static void executeUpdate(Level level, BlockState state, BlockPos pos, Block changedBlock, @Nullable Orientation orientation, boolean movedByPiston, BlockPos sourcePos) {
         // Paper end - Add source block to BlockPhysicsEvent
+        if (org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate()) return; // Leaves - no block update
         try {
             // CraftBukkit start
             org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), state.asBlockData(), org.bukkit.craftbukkit.block.CraftBlock.at(level, sourcePos)); // Paper - Add source block to BlockPhysicsEvent
