From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Mon, 29 Jul 2024 08:20:31 +0800
Subject: [PATCH] Fix fortress mob spawn


diff --git a/net/minecraft/world/level/biome/MobSpawnSettings.java b/net/minecraft/world/level/biome/MobSpawnSettings.java
index 2d256b201df3afa1cf5879fdc931a1aa684110e4..6b8903bc4ec86302d6b5444742db22518fe91a8e 100644
--- a/net/minecraft/world/level/biome/MobSpawnSettings.java
+++ b/net/minecraft/world/level/biome/MobSpawnSettings.java
@@ -175,6 +175,14 @@ public class MobSpawnSettings {
             this.maxCount = maxCount;
         }
 
+        // Leaves start - fix fortress mob spawn
+        @Override
+        public boolean equals(Object obj) {
+            if(!org.leavesmc.leaves.LeavesConfig.modify.oldMC.fixFortressMobSpawn || !(obj instanceof SpawnerData other)) return super.equals(obj);
+            return this.type == other.type && this.minCount == other.minCount && this.maxCount == other.maxCount;
+        }
+        // Leaves end - fix fortress mob spawn
+
         @Override
         public String toString() {
             return EntityType.getKey(this.type) + "*(" + this.minCount + "-" + this.maxCount + "):" + this.getWeight();
