From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 2 Jan 2025 22:01:36 +0800
Subject: [PATCH] Fix CraftPortalEvent logic


diff --git a/net/minecraft/world/level/block/EndPortalBlock.java b/net/minecraft/world/level/block/EndPortalBlock.java
index 19590aa5af0eaf0f9bf76dc189ed67e67c692c77..a2b0f63c20a08e15b36fbcffd264e7004f4480c5 100644
--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -110,10 +110,10 @@ public class EndPortalBlock extends BaseEntityBlock implements Portal {
             }
 
             // CraftBukkit start
-            set.removeAll(Relative.ROTATION); // remove relative rotation flags to simplify event mutation
+            // set.removeAll(Relative.ROTATION); // remove relative rotation flags to simplify event mutation // Leaves - no
             float absoluteYaw = !flag ? f : entity.getYRot() + f;
             float absolutePitch = entity.getXRot() + f1;
-            org.bukkit.craftbukkit.event.PortalEventResult result = org.bukkit.craftbukkit.event.CraftEventFactory.handlePortalEvents(entity, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(bottomCenter, level1, absoluteYaw, absolutePitch), org.bukkit.PortalType.ENDER, 0, 0);
+            org.bukkit.craftbukkit.event.PortalEventResult result = org.bukkit.craftbukkit.event.CraftEventFactory.handlePortalEvents(entity, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(bottomCenter, level1.getWorld(), absoluteYaw, 0.0F), org.bukkit.PortalType.ENDER, 0, 0);
             if (result == null) {
                 return null;
             }
