From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Mar 2025 13:04:35 +0800
Subject: [PATCH] Support REI protocol


diff --git a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
index d02609198253b51f632db46c40afd737d686c520..7296e0b73f1c8e32374f39238d1dfac62bc41afb 100644
--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -82,6 +82,12 @@ public class SmithingTransformRecipe implements SmithingRecipe {
         );
     }
 
+    // Leaves start - REI
+    public SlotDisplay getResult() {
+        return this.result.display();
+    }
+    // Leaves end - REI
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
diff --git a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
index 6582370cfa3d58a7ac08d574e26e7a93d243bdaa..07501a27e6032e578b823d9b617166a9e78fcb20 100644
--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -84,6 +84,12 @@ public class SmithingTrimRecipe implements SmithingRecipe {
         return Optional.of(this.addition);
     }
 
+    // Leaves start
+    public Holder<TrimPattern> pattern() {
+        return pattern;
+    }
+    // Leaves end
+
     @Override
     public RecipeSerializer<SmithingTrimRecipe> getSerializer() {
         return RecipeSerializer.SMITHING_TRIM;
