From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 20:58:06 +0800
Subject: [PATCH 7/7] Renewable coral


diff --git a/net/minecraft/world/level/block/CoralFanBlock.java b/net/minecraft/world/level/block/CoralFanBlock.java
index 1179f10abd646f5475a82b65f5abd6e4e3a1b526..f236d3f4f30fafe2dc03b6556656576f0780c246 100644
--- a/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/net/minecraft/world/level/block/CoralFanBlock.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 
-public class CoralFanBlock extends BaseCoralFanBlock {
+public class CoralFanBlock extends BaseCoralFanBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralFanBlock> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(b -> b.deadBlock), propertiesCodec()).apply(i, CoralFanBlock::new)
     );
@@ -68,4 +68,11 @@ public class CoralFanBlock extends BaseCoralFanBlock {
             return super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
         }
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/block/CoralPlantBlock.java b/net/minecraft/world/level/block/CoralPlantBlock.java
index 51e824704a48d56404fbde838ba2e885b1e6acd0..cefe7b4be401c5c732697a5f1c74234dd3c886fe 100644
--- a/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -16,7 +16,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
+public class CoralPlantBlock extends BaseCoralPlantTypeBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralPlantBlock> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(b -> b.deadBlock), propertiesCodec()).apply(i, CoralPlantBlock::new)
     );
@@ -77,4 +77,12 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
     protected VoxelShape getShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
         return SHAPE;
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED
+            || org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.TRUE;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/block/CoralWallFanBlock.java b/net/minecraft/world/level/block/CoralWallFanBlock.java
index e3bc45c6de18c152853a7a3f220f9d0e33fc8954..f658142a82a7c49762b52b3393f674b64b59ba44 100644
--- a/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 
-public class CoralWallFanBlock extends BaseCoralWallFanBlock {
+public class CoralWallFanBlock extends BaseCoralWallFanBlock implements org.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
     public static final MapCodec<CoralWallFanBlock> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(CoralBlock.DEAD_CORAL_FIELD.forGetter(b -> b.deadBlock), propertiesCodec()).apply(i, CoralWallFanBlock::new)
     );
@@ -68,4 +68,11 @@ public class CoralWallFanBlock extends BaseCoralWallFanBlock {
             return super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
         }
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return org.leavesmc.leaves.LeavesConfig.modify.renewableCoral == org.leavesmc.leaves.LeavesConfig.ModifyConfig.RenewableCoralType.EXPANDED;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java b/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
index 8fe2c147bd9120d19e188814da63e79bc493adc3..be2485e1b1b5d6e0f2c20905a086f086d15a0702 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralClawFeature.java
@@ -17,7 +17,7 @@ public class CoralClawFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) {
+    public boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) { // Leaves - protected -> public
         if (!this.placeCoralBlock(level, random, origin, state)) {
             return false;
         } else {
diff --git a/net/minecraft/world/level/levelgen/feature/CoralFeature.java b/net/minecraft/world/level/levelgen/feature/CoralFeature.java
index 0ab324d8435826878861547fe3309ace9c37abe8..2f03682551798bf756d9b7d829f0945453d81345 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralFeature.java
@@ -31,7 +31,7 @@ public abstract class CoralFeature extends Feature<NoneFeatureConfiguration> {
         return !coral.isEmpty() && this.placeFeature(level, random, origin, coral.get().defaultBlockState());
     }
 
-    protected abstract boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state);
+    public abstract boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state); // Leaves - protected -> public
 
     protected boolean placeCoralBlock(final LevelAccessor level, final RandomSource random, final BlockPos pos, final BlockState state) {
         BlockPos above = pos.above();
diff --git a/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
index 15cdf75f5039334de38eb7a413453cdd7988938c..4bb0fdc30c2e425087d7e0ac8aadfe5162df43cb 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java
@@ -14,7 +14,7 @@ public class CoralMushroomFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) {
+    public boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) { // Leaves - protected -> public
         int height = random.nextInt(3) + 3;
         int width = random.nextInt(3) + 3;
         int length = random.nextInt(3) + 3;
diff --git a/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java b/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
index 789ddf552066a4adebab13c61ed895ed74d85741..68457021c66c19318ff39cc894535daa295066e2 100644
--- a/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralTreeFeature.java
@@ -15,7 +15,7 @@ public class CoralTreeFeature extends CoralFeature {
     }
 
     @Override
-    protected boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) {
+    public boolean placeFeature(final LevelAccessor level, final RandomSource random, final BlockPos origin, final BlockState state) { // Leaves - protected -> public
         BlockPos.MutableBlockPos mutPos = origin.mutable();
         int trunckHeight = random.nextInt(3) + 1;
 
