From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 30 Mar 2025 21:53:45 +0800
Subject: [PATCH] Vanilla player display name


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 5b0b3735caa2ddbcb83d3adc29126205fa4bf131..49dbb6ca89c8a8b7f999ed98bda635d443cd8f85 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -481,7 +481,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.object = null;
         // CraftBukkit start
         this.displayName = this.getScoreboardName();
-        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.adventure$displayName = org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName ? io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getDisplayName()) : net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper // Leaves - vanilla display name
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
         // CraftBukkit end
diff --git a/net/minecraft/world/scores/Scoreboard.java b/net/minecraft/world/scores/Scoreboard.java
index 8ac6138dbf9299a0253599a57f85138fc7c80bd2..4b425d768b0d24a4ca95b901a9c706b507395bf7 100644
--- a/net/minecraft/world/scores/Scoreboard.java
+++ b/net/minecraft/world/scores/Scoreboard.java
@@ -277,6 +277,16 @@ public class Scoreboard {
         }
 
         this.teamsByPlayer.put(playerName, team);
+        // Leaves start - vanilla display name
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName) {
+            boolean result = team.getPlayers().add(playerName);
+            net.minecraft.server.level.ServerPlayer player = net.minecraft.server.MinecraftServer.getServer().getPlayerList().getPlayerByName(playerName);
+            if (player != null) {
+                player.adventure$displayName = io.papermc.paper.adventure.PaperAdventure.asAdventure(player.getDisplayName());
+            }
+            return result;
+        }
+        // Leaves end - vanilla display name
         return team.getPlayers().add(playerName);
     }
 
@@ -296,6 +306,14 @@ public class Scoreboard {
         } else {
             this.teamsByPlayer.remove(playerName);
             team.getPlayers().remove(playerName);
+            // Leaves start - vanilla display name
+            if (org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName) {
+                net.minecraft.server.level.ServerPlayer player = net.minecraft.server.MinecraftServer.getServer().getPlayerList().getPlayerByName(playerName);
+                if (player != null) {
+                    player.adventure$displayName = io.papermc.paper.adventure.PaperAdventure.asAdventure(player.getDisplayName());
+                }
+            }
+            // Leaves end - vanilla display name
         }
     }
 
