From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Sun, 27 Oct 2024 15:57:37 +0800
Subject: [PATCH] Disable vault blacklist


diff --git a/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index 3a107ba4ca69b2dd948592441237c098abb8e5d6..6bc293aae8a8662de688ffa4a7a160761ea0497d 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -61,11 +61,12 @@ public class VaultServerData {
     }
 
     boolean hasRewardedPlayer(Player player) {
-        return this.rewardedPlayers.contains(player.getUUID());
+        return !org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist && this.rewardedPlayers.contains(player.getUUID()); // Leaves - disable vault blacklist
     }
 
     @VisibleForTesting
     public void addToRewardedPlayers(Player player) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist) return; // Leaves - disable vault blacklist
         this.rewardedPlayers.add(player.getUUID());
         if (this.rewardedPlayers.size() > 128) {
             Iterator<UUID> iterator = this.rewardedPlayers.iterator();
diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 4e4952e9fd95a54aefca3c3e06f98017b3119ec1..425ea46b05c2173815677b9ce87b5c7a692d88d9 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -68,8 +68,8 @@ public class VaultSharedData {
         Set<UUID> set = config.playerDetector()
             .detect(level, config.entitySelector(), pos, deactivationRange, false)
             .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+            .filter(uuid -> org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist || !serverData.getRewardedPlayers().contains(uuid))
+            .collect(Collectors.toSet()); // Leaves - disable vault blacklist
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
