From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Sun, 27 Oct 2024 15:57:37 +0800
Subject: [PATCH] Disable vault blacklist


diff --git a/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index c73ccb38eb4cd5995a9005f3319acad2605c6a2b..5500eeafb5c8d15049a5308c18323f549d7fde3e 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -57,7 +57,7 @@ public class VaultServerData {
     }
 
     boolean hasRewardedPlayer(Player player) {
-        return this.rewardedPlayers.contains(player.getUUID());
+        return !org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist && this.rewardedPlayers.contains(player.getUUID()); // Leaves - disable vault blacklist
     }
 
     @VisibleForTesting
@@ -66,6 +66,7 @@ public class VaultServerData {
         addToRewardedPlayers(player.getUUID());
     }
     public boolean addToRewardedPlayers(final java.util.UUID player) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist) return false; // Leaves - disable vault blacklist
         final boolean removed = this.rewardedPlayers.add(player);
     // Paper end - Vault API
         if (this.rewardedPlayers.size() > 128) {
diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 036087f8085f05b43703ceca3413e0b8ac0a412b..a55c773f37cf8a50954d9a0f28868e42d94ad90e 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -68,8 +68,8 @@ public class VaultSharedData {
         Set<UUID> set = config.playerDetector()
             .detect(level, config.entitySelector(), pos, deactivationRange, false)
             .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+            .filter(uuid -> org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist || !serverData.getRewardedPlayers().contains(uuid))
+            .collect(Collectors.toSet()); // Leaves - disable vault blacklist
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
