From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Thu, 24 Oct 2024 23:10:34 +0800
Subject: [PATCH] Fix Incorrect Collision Behavior for Block Shape


diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index a97a2a8492f3858e3b622d26768b4d819c9b47a7..8a85fa0dbfb6e273cc77ff896e5c6a7deb94467b 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -101,6 +101,14 @@ public final class CollisionUtil {
                (box1.minZ - box2.maxZ) < -COLLISION_EPSILON && (box1.maxZ - box2.minZ) > COLLISION_EPSILON;
     }
 
+    // Leaves start
+    public static boolean voxelShapeIntersectVanilla(final net.minecraft.world.phys.AABB box1, final net.minecraft.world.phys.AABB box2) {
+        return box1.minX < box2.maxX && box1.maxX > box2.minX &&
+               box1.minY < box2.maxY && box1.maxY > box2.minY &&
+               box1.minZ < box2.maxZ && box1.maxZ > box2.minZ;
+    }
+    // Leaves end
+
     // assume !isEmpty(target) && abs(source_move) >= COLLISION_EPSILON
     public static double collideX(final AABB target, final AABB source, final double source_move) {
         if ((source.minY - target.maxY) < -COLLISION_EPSILON && (source.maxY - target.minY) > COLLISION_EPSILON &&
@@ -2014,7 +2022,10 @@ public final class CollisionUtil {
                                     AABB singleAABB = ((CollisionVoxelShape)blockCollision).moonrise$getSingleAABBRepresentation();
                                     if (singleAABB != null) {
                                         singleAABB = singleAABB.move((double)blockX, (double)blockY, (double)blockZ);
-                                        if (!voxelShapeIntersect(aabb, singleAABB)) {
+                                        // Leaves start - Fix incorrect collision behavior for block shape
+                                        boolean isBlockShape = blockCollision == net.minecraft.world.phys.shapes.Shapes.block();
+                                        if (isBlockShape && !voxelShapeIntersectVanilla(aabb, singleAABB) || !isBlockShape && !voxelShapeIntersect(aabb, singleAABB)) {
+                                        // Leaves end - Fix incorrect collision behavior for block shape
                                             continue;
                                         }
 
