From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 15 May 2025 12:17:30 +0800
Subject: [PATCH] Vanilla portal handle


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index ff9e27a648e7fe8ff51fd1edc2e321b2213e4e6c..21c605ec1fc93366bc4c2d82d5ef9137a07b1d3b 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -846,6 +846,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     // CraftBukkit start
     public void postTick() {
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) return; // Leaves - vanilla
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
         if (!(this instanceof ServerPlayer) && this.isAlive()) { // Paper - don't attempt to teleport dead entities
             this.handlePortal();
@@ -867,7 +868,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             this.boardingCooldown--;
         }
 
-        if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle || this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick // Leaves - vanilla
         if (this.canSpawnSprintParticle()) {
             this.spawnSprintParticle();
         }
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 1123ec0038552e0b40774f4a433ff325695ea071..fdff726c89e25164437ec8ef370349df2e44c804 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -296,7 +296,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
         this.checkBelowWorld();
         this.computeSpeed();
-        // this.handlePortal(); // CraftBukkit - handled in postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) this.handlePortal(); // CraftBukkit - handled in postTick // Leaves - vanilla
         this.behavior.tick();
         // CraftBukkit start
         org.bukkit.World bworld = this.level().getWorld();
