From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 15 May 2025 12:17:30 +0800
Subject: [PATCH] Vanilla portal handle


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 0260ccee26a4dad7c7ea53504a45c645abe3c28f..03972434172ff8615f0f5defa1076d348e0b6b19 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -841,6 +841,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     // CraftBukkit start
     public void postTick() {
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) return; // Leaves - vanilla
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
         if (!(this instanceof ServerPlayer) && this.isAlive()) { // Paper - don't attempt to teleport dead entities
             this.handlePortal();
@@ -861,7 +862,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             this.boardingCooldown--;
         }
 
-        if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle || this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick // Leaves - vanilla
         if (this.canSpawnSprintParticle()) {
             this.spawnSprintParticle();
         }
diff --git a/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index 3454506b2d7ada401600e2adccd462a15bd86723..c8b053b01b06af6033b9a74452946333dbeb77fa 100644
--- a/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -293,7 +293,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
         }
 
         this.checkBelowWorld();
-        // this.handlePortal(); // CraftBukkit - handled in postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) this.handlePortal(); // CraftBukkit - handled in postTick // Leaves - vanilla
         this.behavior.tick();
         // CraftBukkit start
         org.bukkit.World bworld = this.level().getWorld();
