From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 21:24:48 +0800
Subject: [PATCH 5/7] Force Void Trade


diff --git a/net/minecraft/world/entity/npc/villager/AbstractVillager.java b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
index b5247d45159898cd6ede425a4a99db8439a105b6..e846948eced1a5a370e35f8ee5427be67f03a6ed 100644
--- a/net/minecraft/world/entity/npc/villager/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
@@ -57,6 +57,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     private @Nullable Player tradingPlayer;
     protected @Nullable MerchantOffers offers;
     private final SimpleContainer inventory = new SimpleContainer(8, (org.bukkit.craftbukkit.entity.CraftAbstractVillager) this.getBukkitEntity()); // CraftBukkit - add argument
+    protected boolean voidTrade = false; // Leaves - force void trade
 
     public AbstractVillager(final EntityType<? extends AbstractVillager> type, final Level level) {
         super(type, level);
@@ -148,7 +149,13 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void processTrade(MerchantOffer offer, io.papermc.paper.event.player.@Nullable PlayerPurchaseEvent event) { // The MerchantRecipe passed in here is the one set by the PlayerPurchaseEvent
         if (event == null || event.willIncreaseTradeUses()) {
-            offer.increaseUses();
+            // Leaves start - force void trade
+            if (this.voidTrade) {
+                offer.increaseVoidUses();
+            } else {
+                offer.increaseUses();
+            }
+            // Leaves end - force void trade
         }
         if (event == null || event.isRewardingExp()) {
             this.rewardTradeXp(offer);
@@ -160,7 +167,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void notifyTrade(final MerchantOffer offer) {
         // offer.increaseUses(); // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
-        this.ambientSoundTime = -this.getAmbientSoundInterval();
+        if (!this.voidTrade) this.ambientSoundTime = -this.getAmbientSoundInterval(); // Leaves - force void trade
         // this.rewardTradeXp(offer); // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
         if (this.tradingPlayer instanceof ServerPlayer) {
             CriteriaTriggers.TRADE.trigger((ServerPlayer)this.tradingPlayer, this, offer.getResult());
@@ -177,7 +184,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void notifyTradeUpdated(final ItemStack itemStack) {
         if (!this.level().isClientSide() && this.ambientSoundTime > -this.getAmbientSoundInterval() + 20) {
-            this.ambientSoundTime = -this.getAmbientSoundInterval();
+            if (!this.voidTrade) this.ambientSoundTime = -this.getAmbientSoundInterval(); // Leaves - force void trade
             this.makeSound(this.getTradeUpdatedSound(!itemStack.isEmpty()));
         }
     }
@@ -222,9 +229,21 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     }
 
     protected void stopTrading() {
+        // Leaves start - force void trade
+        this.voidTrade = false;
+        if (this.offers != null) {
+            this.offers.forEach(MerchantOffer::resetVoidUses);
+        }
+        // Leaves end - force void trade
         this.setTradingPlayer(null);
     }
 
+    // Leaves start - force void trade
+    public void setVoidTrade() {
+        this.voidTrade = true;
+    }
+    // Leaves end - force void trade
+
     @Override
     public void die(final DamageSource source) {
         super.die(source);
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index f481fd661440f77c98940cbde4a8b95818e4a22c..952f6bb64a417b83171d753f5924d8c7003f9156 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -359,6 +359,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     }
 
     private void startTrading(final Player player) {
+        this.voidTrade = false; // Leaves - force void trade
         this.updateSpecialPrices(player);
         this.setTradingPlayer(player);
         this.openTradingScreen(player, this.getDisplayName(), this.getVillagerData().level());
@@ -610,8 +611,12 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     @Override
     protected void rewardTradeXp(final MerchantOffer offer) {
         int popXp = 3 + this.random.nextInt(4);
-        this.villagerXp = this.villagerXp + offer.getXp();
-        this.lastTradedPlayer = this.getTradingPlayer();
+        // Leaves start - force void trade
+        if (!this.voidTrade) {
+            this.villagerXp += offer.getXp();
+            this.lastTradedPlayer = this.getTradingPlayer();
+        }
+        // Leaves end - force void trade
         if (this.shouldIncreaseLevel()) {
             this.updateMerchantTimer = 40;
             this.increaseProfessionLevelOnUpdate = true;
diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
index 4c7fad3444dd8718e31bda8a8ab2efdaf0a736a1..d95c6687476a854780ad0b0720940c2c68428583 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -121,6 +121,7 @@ public class WanderingTrader extends AbstractVillager implements Consumable.Over
                     return InteractionResult.CONSUME;
                 }
 
+                this.voidTrade = false; // Leaves - force void trade
                 this.setTradingPlayer(player);
                 this.openTradingScreen(player, this.getDisplayName(), 1);
             }
diff --git a/net/minecraft/world/inventory/MerchantMenu.java b/net/minecraft/world/inventory/MerchantMenu.java
index 7fcc2c2ad603699e176bb95559e51134174ecc20..8aa48fd5be819af44ce1ac4f5e101a9ae66a457d 100644
--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -25,7 +25,7 @@ public class MerchantMenu extends AbstractContainerMenu {
     private static final int SELLSLOT2_X = 162;
     private static final int BUYSLOT_X = 220;
     private static final int ROW_Y = 37;
-    private final Merchant trader;
+    public final Merchant trader; // Leaves - private -> public
     private final MerchantContainer tradeContainer;
     private int merchantLevel;
     private boolean showProgressBar;
diff --git a/net/minecraft/world/item/trading/MerchantOffer.java b/net/minecraft/world/item/trading/MerchantOffer.java
index 006e8400349131faadc946fe3f5bd9ea8180019e..7e1cee0a598b7ea5710b011c3d76c40567963997 100644
--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -39,6 +39,7 @@ public class MerchantOffer {
     public float priceMultiplier;
     public int xp;
     public boolean ignoreDiscounts; // Paper - Add ignore discounts API
+    public int voidTradeUses; // Leaves - force void trade
 
     // CraftBukkit start
     private org.bukkit.craftbukkit.inventory.@org.jspecify.annotations.Nullable CraftMerchantRecipe bukkitHandle;
@@ -78,6 +79,7 @@ public class MerchantOffer {
         this.priceMultiplier = priceMultiplier;
         this.xp = xp;
         this.ignoreDiscounts = ignoreDiscounts; // Paper
+        this.voidTradeUses = 0; // Leaves - force void trade
     }
 
     public MerchantOffer(final ItemCost buy, final ItemStack result, final int maxUses, final int xp, final float priceMultiplier) {
@@ -186,6 +188,16 @@ public class MerchantOffer {
         this.uses++;
     }
 
+    // Leaves start - force void trade
+    public void increaseVoidUses() {
+        ++this.voidTradeUses;
+    }
+
+    public void resetVoidUses() {
+        this.voidTradeUses = 0;
+    }
+    // Leaves end - force void trade
+
     public int getDemand() {
         return this.demand;
     }
@@ -215,7 +227,7 @@ public class MerchantOffer {
     }
 
     public boolean isOutOfStock() {
-        return this.uses >= this.maxUses;
+        return this.uses + this.voidTradeUses >= this.maxUses; // Leaves - force void trade
     }
 
     public void setToOutOfStock() {
diff --git a/net/minecraft/world/level/block/EndGatewayBlock.java b/net/minecraft/world/level/block/EndGatewayBlock.java
index eeef6b6991f13b9c682e763e70db9e736584b744..c269726570398a518e4010c6188a9f5adb96623b 100644
--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -112,6 +112,15 @@ public class EndGatewayBlock extends BaseEntityBlock implements Portal {
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
         if (currentLevel.getBlockEntity(portalEntryPos) instanceof TheEndGatewayBlockEntity endGatewayBlockEntity) {
             Vec3 teleportPosition = endGatewayBlockEntity.getPortalPosition(currentLevel, portalEntryPos);
+            // Leaves start - force void trade
+            if (org.leavesmc.leaves.LeavesConfig.modify.forceVoidTrade && teleportPosition != null && entity instanceof net.minecraft.server.level.ServerPlayer player) {
+                if (player.containerMenu instanceof net.minecraft.world.inventory.MerchantMenu merchantMenu) {
+                    if (merchantMenu.trader instanceof net.minecraft.world.entity.npc.villager.AbstractVillager villager) {
+                        villager.setVoidTrade();
+                    }
+                }
+            }
+            // Leaves end - force void trade
             if (teleportPosition == null) {
                 return null;
             } else {
