From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 10 May 2025 00:49:51 +0800
Subject: [PATCH] Vanilla creative pickup behavior


diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index 4de4e2ce8344db91a9c6644298dc75c2fc9ae237..94d7d74b19d1f3180709a50c381c239af3daa7f5 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -175,6 +175,13 @@ public class Inventory implements Container, Nameable {
     // CraftBukkit start - Watch method above! :D
     public int canHold(ItemStack itemStack) {
         int remains = itemStack.getCount();
+
+        // Leaves start - vanilla creative pickup behavior
+        if (player.hasInfiniteMaterials()) {
+            return remains;
+        }
+        // Leaves end - vanilla creative pickup behavior
+
         for (int slot = 0; slot < this.items.size(); ++slot) {
             ItemStack itemInSlot = this.getItem(slot);
             if (itemInSlot.isEmpty()) {
