From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Mar 2025 13:04:35 +0800
Subject: [PATCH] Support REI protocol


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index d89a1aa9355205883412eaaf535dad30f945a4dc..33e05636164144b3d2bdbd091c72583728cc294e 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1624,6 +1624,7 @@ public abstract class PlayerList {
             serverPlayer.getRecipeBook().sendInitialRecipeBook(serverPlayer);
         }
         org.leavesmc.leaves.protocol.BBORProtocol.onDataPackReload(); // Leaves - bbor
+        org.leavesmc.leaves.protocol.rei.REIServerProtocol.onRecipeReload(); // Leaves - rei
     }

     public boolean isAllowCommandsForAllPlayers() {
diff --git a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
index 143601053a6aeea4396f8e0ee0746ff7d5bbb323..0af83ec1fe3aa85c6bfc814a1339a54e9d3725d6 100644
--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -87,6 +87,12 @@ public class SmithingTransformRecipe implements SmithingRecipe {
         );
     }

+    // Leaves start
+    public ItemStack getResult() {
+        return this.result.copy();
+    }
+    // Leaves end
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
