From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Mar 2025 13:04:35 +0800
Subject: [PATCH] Recipe send all


diff --git a/net/minecraft/stats/ServerRecipeBook.java b/net/minecraft/stats/ServerRecipeBook.java
index e3985b70cee7f7d56f179aeef8c2a6a6b312d83a..a476f6b6554b4c7ba1625ab4b9da3bcf3d40955b 100644
--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -150,12 +150,23 @@ public class ServerRecipeBook extends RecipeBook {
 
     public void sendInitialRecipeBook(ServerPlayer player) {
         player.connection.send(new ClientboundRecipeBookSettingsPacket(this.getBookSettings()));
-        List<ClientboundRecipeBookAddPacket.Entry> list = new ArrayList<>(this.known.size());
+        // Leaves start - recipe-send-all
+        List<ClientboundRecipeBookAddPacket.Entry> list;
 
-        for (ResourceKey<Recipe<?>> resourceKey : this.known) {
-            this.displayResolver
-                .displaysForRecipe(resourceKey, entry -> list.add(new ClientboundRecipeBookAddPacket.Entry(entry, false, this.highlight.contains(resourceKey))));
+        if (org.leavesmc.leaves.LeavesConfig.protocol.recipeSendAll) {
+            list = new ArrayList<>(player.server.getRecipeManager().getRecipes().size());
+            player.server.getRecipeManager().getRecipes().stream().map(RecipeHolder::id).forEach( key -> {
+                this.displayResolver
+                    .displaysForRecipe(key, entry -> list.add(new ClientboundRecipeBookAddPacket.Entry(entry, false, this.highlight.contains(key))));
+            });
+        } else {
+            list = new ArrayList<>(this.known.size());
+            for (ResourceKey<Recipe<?>> resourceKey : this.known) {
+                this.displayResolver
+                    .displaysForRecipe(resourceKey, entry -> list.add(new ClientboundRecipeBookAddPacket.Entry(entry, false, this.highlight.contains(resourceKey))));
+            }
         }
+        // Leaves end - recipe-send-all
 
         player.connection.send(new ClientboundRecipeBookAddPacket(list, true));
     }
