From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 30 Mar 2025 21:53:45 +0800
Subject: [PATCH] Vanilla player display name


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 2b8eb0f0aae8c6534b124012e26bf683929131a8..b3cb4970db16ab12f6588336168a453a47d284fb 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -442,7 +442,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
         // CraftBukkit start
         this.displayName = this.getScoreboardName();
-        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.adventure$displayName = org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName ? io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getDisplayName()) : net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper // Leaves - Vanilla display name
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
