From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:06:03 +0800
Subject: [PATCH] Spawn invulnerable time


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 270d567b2059466b6e54d63241d6d9f86f6da437..3623f1e71633f5823efec4f56285d1d7efe8087a 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -240,6 +240,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private int lastSentFood = -99999999;
     private boolean lastFoodSaturationZero = true;
     public int lastSentExp = -99999999;
+    protected int spawnInvulnerableTime = 60; // Leaves - spawn invulnerable time
     private ChatVisiblity chatVisibility = ChatVisiblity.FULL;
     public ParticleStatus particleStatus = ParticleStatus.ALL;
     private boolean canChatColor = true;
@@ -727,6 +728,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.resetOperationCountPerTick(); // Leaves - player operation limiter
         this.gameMode.tick();
         this.wardenSpawnTracker.tick();
+        if (this.spawnInvulnerableTime > 0) --this.spawnInvulnerableTime; // Leaves - spawn invulnerable time
         if (this.invulnerableTime > 0) {
             this.invulnerableTime--;
         }
@@ -1310,6 +1312,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         if (this.isInvulnerableTo(level, damageSource)) {
             return false;
         } else {
+            // Leaves start - spawn invulnerable time
+            if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.spawnInvulnerableTime) {
+                if (this.spawnInvulnerableTime > 0 && !damageSource.is(net.minecraft.tags.DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                    return false;
+                }
+            }
+            // Leaves end - spawn invulnerable time
             Entity entity = damageSource.getEntity();
             if (!( // Paper - split the if statement. If below statement is false, hurtServer would not have been evaluated. Return false.
              !(entity instanceof Player player && !this.canHarmPlayer(player))
