From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Mar 2025 13:04:35 +0800
Subject: [PATCH] Support REI protocol


diff --git a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
index 9bc0a9c3577d63a0ad5489bfd4c07d5006245c5f..ddf603f94898e0383ee4974381da7b884f07fb66 100644
--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -83,6 +83,12 @@ public class SmithingTransformRecipe implements SmithingRecipe {
         );
     }
 
+    // Leaves start
+    public ItemStack getResult() {
+        return this.result.apply(net.minecraft.world.item.Items.OAK_LEAVES.getDefaultInstance()); // Awful usage due to Mojang's changing in 1.21.5
+    }
+    // Leaves end
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
diff --git a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
index c324896afc2ee28ebb4d426ce4a469ee847ce24d..d5b26ff7be916e07e1162536cb2ddf5674c82f46 100644
--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -85,6 +85,12 @@ public class SmithingTrimRecipe implements SmithingRecipe {
         return Optional.of(this.addition);
     }
 
+    // Leaves start
+    public Holder<TrimPattern> pattern() {
+        return pattern;
+    }
+    // Leaves end
+
     @Override
     public RecipeSerializer<SmithingTrimRecipe> getSerializer() {
         return RecipeSerializer.SMITHING_TRIM;
