From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:42:17 +0800
Subject: [PATCH] Old zombie reinforcement


diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index d9935d71393b28ee8052b1c05739bfd44d6afb47..19f892c01409af4d02a7771aca0ad098588ffd71 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -314,7 +314,7 @@ public class Zombie extends Monster {
                 int floor = Mth.floor(this.getX());
                 int floor1 = Mth.floor(this.getY());
                 int floor2 = Mth.floor(this.getZ());
-                EntityType<? extends Zombie> type = this.getType();
+                EntityType<? extends Zombie> type = org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldZombieReinforcement ? EntityType.ZOMBIE : this.getType(); // Leaves end - old zombie reinforcement
                 Zombie zombie = type.create(level, EntitySpawnReason.REINFORCEMENT);
                 if (zombie == null) {
                     return true;
