From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Mar 2025 13:04:35 +0800
Subject: [PATCH] Support REI protocol


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 5c68d9c2a8a129a5294dd76804848ebe8fd61a1d..a87bc493c11391833578c1faa5cf8fe246b7bf98 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1628,6 +1628,7 @@ public abstract class PlayerList {
             serverPlayer.getRecipeBook().sendInitialRecipeBook(serverPlayer);
         }
         org.leavesmc.leaves.protocol.BBORProtocol.onDataPackReload(); // Leaves - bbor
+        org.leavesmc.leaves.protocol.rei.REIServerProtocol.onRecipeReload(); // Leaves - rei
     }
 
     public boolean isAllowCommandsForAllPlayers() {
diff --git a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
index 9bc0a9c3577d63a0ad5489bfd4c07d5006245c5f..08d6569cf84155b868d3480ac7b4d23b6317a4f2 100644
--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -83,6 +83,12 @@ public class SmithingTransformRecipe implements SmithingRecipe {
         );
     }
 
+    // Leaves start
+    public ItemStack getResult() {
+        return this.result.copy();
+    }
+    // Leaves end
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
