From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <HyacinthHaru@users.noreply.github.com>
Date: Wed, 15 Apr 2026 23:05:36 +0800
Subject: [PATCH] Wool Hopper Counter


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index ffd54d51052ec228a56a2be26760b0934f341164..cdebec900b5a5f36aeb8e8e2d3b725ed885bd0f8 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -196,8 +196,30 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     changed |= action.getAsBoolean(); // Paper - note: this is not a validator, it's what adds/sucks in items
                 }
 
+                // Leaves start - Wool hopper counter
+                if (org.leavesmc.leaves.LeavesConfig.modify.hopperCounter.unlimitedSpeed && org.leavesmc.leaves.util.HopperCounter.isEnabled()) {
+                    net.minecraft.world.item.DyeColor woolColor = org.leavesmc.leaves.util.WoolUtils.getWoolColorAtPosition(level, entity.getBlockPos().relative(state.getValue(HopperBlock.FACING)));
+                    if (woolColor != null) {
+                        for (int unlimited = 0; unlimited < Short.MAX_VALUE; unlimited++) {
+                            changed |= suckInItems(level, entity);
+                            if (!changed) {
+                                break;
+                            } else {
+                                woolHopperCounter(level, pos, state, HopperBlockEntity.getContainerAt(level, pos));
+                            }
+                        }
+                    }
+                }
+                // Leaves end - Wool hopper counter
+
                 if (changed) {
                     entity.setCooldown(level.spigotConfig.hopperTransfer); // Spigot
+                    // Leaves start - Wool hopper counter
+                    if (org.leavesmc.leaves.LeavesConfig.modify.hopperCounter.unlimitedSpeed && org.leavesmc.leaves.util.HopperCounter.isEnabled() && woolHopperCounter(level, pos, state, HopperBlockEntity.getContainerAt(level, pos))) {
+                        entity.setCooldown(0);
+                        return true;
+                    }
+                    // Leaves end - Wool hopper counter
                     setChanged(level, pos, state);
                     return true;
                 }
@@ -437,6 +459,13 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(final Level level, final BlockPos blockPos, final HopperBlockEntity self) {
+        // Leaves start - hopper counter
+        if (org.leavesmc.leaves.util.HopperCounter.isEnabled()) {
+            if (woolHopperCounter(level, blockPos, level.getBlockState(blockPos), HopperBlockEntity.getContainerAt(level, blockPos))) {
+                return true;
+            }
+        }
+        // Leaves end - hopper counter
         Container container = getAttachedContainer(level, blockPos, self);
         if (container == null) {
             return false;
@@ -450,6 +479,26 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         }
     }
 
+    // Leaves start - hopper counter
+    private static boolean woolHopperCounter(final Level level, final BlockPos blockPos, final BlockState state, final @Nullable Container container) {
+        if (container == null) {
+            return false;
+        }
+        net.minecraft.world.item.DyeColor woolColor = org.leavesmc.leaves.util.WoolUtils.getWoolColorAtPosition(level, blockPos.relative(state.getValue(HopperBlock.FACING)));
+        if (woolColor != null) {
+            for (int i = 0; i < container.getContainerSize(); ++i) {
+                if (!container.getItem(i).isEmpty()) {
+                    ItemStack itemstack = container.getItem(i);
+                    org.leavesmc.leaves.util.HopperCounter.getCounter(woolColor).add(level.getServer(), itemstack);
+                    container.setItem(i, ItemStack.EMPTY);
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+    // Leaves end - hopper counter
+
     private static int[] getSlots(final Container container, final Direction direction) {
         if (container instanceof WorldlyContainer worldlyContainer) {
             return worldlyContainer.getSlotsForFace(direction);
@@ -535,6 +584,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     }
 
     public static boolean addItem(final Container container, final ItemEntity entity) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.hopperCounter.unlimitedSpeed && org.leavesmc.leaves.util.HopperCounter.isEnabled() && entity.isRemoved()) return false; // Leaves - Wool hopper counter
         boolean changed = false;
         // CraftBukkit start
         if (org.bukkit.event.inventory.InventoryPickupItemEvent.getHandlerList().getRegisteredListeners().length > 0) { // Paper - optimize hoppers
