From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:42:17 +0800
Subject: [PATCH] Old zombie reinforcement


diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index c0f7b625ca135772f3f979d64b1003699b4109f8..3f87bcaa806de4541d3f4f389e897d9875330009 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -340,7 +340,7 @@ public class Zombie extends Monster {
                 int floor = Mth.floor(this.getX());
                 int floor1 = Mth.floor(this.getY());
                 int floor2 = Mth.floor(this.getZ());
-                EntityType<? extends Zombie> type = this.getType();
+                EntityType<? extends Zombie> type = org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldZombieReinforcement ? EntityType.ZOMBIE : this.getType(); // Leaves end - old zombie reinforcement
                 Zombie zombie = type.create(level, EntitySpawnReason.REINFORCEMENT);
                 if (zombie == null) {
                     return true;
