From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 30 Mar 2025 21:53:45 +0800
Subject: [PATCH] Vanilla player display name


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 4d2576d4824769a1ef76a93e673da823bededc49..8764f6d497a5df9dc94ae4e041d3d07bc2a1e4e2 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -477,7 +477,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
         // CraftBukkit start
         this.displayName = this.getScoreboardName();
-        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.adventure$displayName = org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName ? io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getDisplayName()) : net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper // Leaves - Vanilla display name
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
