From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Fri, 23 May 2025 08:05:45 +0800
Subject: [PATCH] Old zombie piglin drop behavior


diff --git a/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java b/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
index 8fcb6cd1cefa6c9e1809944b4720cfdc35325729..0c5da9aa8e682ee37b04a5fc18d696f25fbb0ba2 100644
--- a/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
@@ -109,6 +109,9 @@ public class ZombifiedPiglin extends Zombie implements NeutralMob {
             this.maybeAlertOthers();
         }
 
+
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldZombiePiglinDrop && this.isAngry()) this.lastHurtByPlayerMemoryTime = this.tickCount; // Leaves - old zombie piglin drop
+
         super.customServerAiStep(level);
     }
 
