From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Fri, 9 May 2025 12:49:51 -0400
Subject: [PATCH] Vanilla creative pickup behavior


diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index 57bf2819271b3293a065b58d31b609f8463811b4..0a6c9754c6bc4e8200b2be68bede25806d69fc77 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -159,6 +159,11 @@ public class Inventory implements Container, Nameable {
     // CraftBukkit start - Watch method above! :D
     public int canHold(ItemStack itemStack) {
         int remains = itemStack.getCount();
+
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaCreativePickupBehavior && player.hasInfiniteMaterials()) {
+            return remains;
+        }
+
         for (int slot = 0; slot < this.items.size(); ++slot) {
             ItemStack itemInSlot = this.getItem(slot);
             if (itemInSlot.isEmpty()) {
