From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 1 Jul 2025 00:19:23 +0800
Subject: [PATCH] Configurable item damage check


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index c59dec418f94b468f0465d0375809031f10bac39..2d01258e377ad1d628b193d15d794a057c3d11b4 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -603,11 +603,11 @@ public final class ItemStack implements DataComponentHolder {
     }
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
+        return org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableItemDamageCheck ? this.getOrDefault(DataComponents.DAMAGE, 0) : Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage()); // Leaves - disable check
     }
 
     public void setDamageValue(int damage) {
-        this.set(DataComponents.DAMAGE, Mth.clamp(damage, 0, this.getMaxDamage()));
+        this.set(DataComponents.DAMAGE, org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableItemDamageCheck ? damage : Mth.clamp(damage, 0, this.getMaxDamage())); // Leaves - disable check
     }
 
     public int getMaxDamage() {
