From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Haru Hyacinth <haru@MacBook-Air.local>
Date: Thu, 16 Apr 2026 10:50:42 +0800
Subject: [PATCH] Fix latent compile errors in rebased patches

Co-Authored-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f883c4d515e31ab1f9f2c00f293d04942fc217f2..b7f3736745ab72a2d2da44aeb83192d815f18232 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4054,7 +4054,7 @@ public abstract class Entity
 
     public @Nullable Entity teleport(TeleportTransition transition) { // Paper - remove param final
         // Paper start - Fix item duplication and teleport issues
-        if (!(org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal && this instanceof net.minecraft.world.entity.boss.EnderDragonPart)) if ((!this.isAlive() || !this.valid) && (transition.newLevel() != this.level)) { // Leaves - endDragonPartCanUseEndPortal
+        if (!(org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal && this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragonPart)) if ((!this.isAlive() || !this.valid) && (transition.newLevel() != this.level)) { // Leaves - endDragonPartCanUseEndPortal
             LOGGER.warn("Illegal Entity Teleport {} to {}:{}", this, transition.newLevel(), transition.position(), new Throwable());
             return null;
         }
@@ -4170,7 +4170,7 @@ public abstract class Entity
             newEntity.restoreFrom(this);
             this.removeAfterChangingDimensions();
             newEntity.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
-            if (this.inWorld || (this instanceof net.minecraft.world.entity.boss.EnderDragonPart && org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal)) newLevel.addDuringTeleport(newEntity); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned // Leaves - ender dragon part can use end portal
+            if (this.inWorld || (this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragonPart && org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal)) newLevel.addDuringTeleport(newEntity); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned // Leaves - ender dragon part can use end portal
 
             for (Entity newPassenger : newPassengers) {
                 newPassenger.startRiding(newEntity, true, false);
@@ -4284,7 +4284,7 @@ public abstract class Entity
     }
 
     public boolean canTeleport(final Level from, final Level to) {
-        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal && this instanceof net.minecraft.world.entity.boss.EnderDragonPart) return true; // Leaves - enderDragonPartCanUseEndPortal
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal && this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragonPart) return true; // Leaves - enderDragonPartCanUseEndPortal
         if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowEntityPortalWithPassenger && (this.isPassenger() || this.isVehicle())) return false; // Leaves - allowEntityPortalWithPassenger
         if (!this.isAlive() || !this.valid) return false; // Paper - Fix item duplication and teleport issues
         if (from.dimension() == Level.END && to.dimension() == Level.OVERWORLD) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index d07ce96b2803f611006c8735651d15b3ff117141..c719a2ab757fe24f17afa29372a9895309ad4ff1 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -406,7 +406,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         // Leaves start - ender dragon part can use end portal
         if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.enderDragonPartCanUseEndPortal) return;
         if (!(this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon dragon)) return;
-        for (net.minecraft.world.entity.boss.EnderDragonPart part : dragon.getSubEntities()) {
+        for (net.minecraft.world.entity.boss.enderdragon.EnderDragonPart part : dragon.getSubEntities()) {
             PortalProcessor portalManager = part.portalProcess;
             if (portalManager == null) continue;
             if (!(portalManager.portal instanceof net.minecraft.world.level.block.EndPortalBlock)) continue;
@@ -597,7 +597,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
 
     // Leaves start - optimize sun burn tick
     private @Nullable BlockPos cachedSunBurnEyeBlockPos;
-    private @Nullable net.minecraft.world.phys.Vec3 cachedSunBurnPosition;
+    private net.minecraft.world.phys.@Nullable Vec3 cachedSunBurnPosition;
     // Leaves end - optimize sun burn tick
 
     public boolean isSunBurnTick() {
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 1fac92922e744b35e3027ddc0fe6c2a36439f88a..b44fea9dad4c27d8801d89e311ddf33af7cb1163 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -66,7 +66,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
         this.setItem(itemStack);
         // Leaves start - configurable vanilla random
         if (org.leavesmc.leaves.LeavesConfig.modify.useVanillaRandom) {
-            this.setDeltaMovement(level.random.nextDouble() * 0.2 - 0.1, 0.2, level.random.nextDouble() * 0.2 - 0.1);
+            this.setDeltaMovement(level.getRandom().nextDouble() * 0.2 - 0.1, 0.2, level.getRandom().nextDouble() * 0.2 - 0.1);
         } else {
             this.setDeltaMovement(this.random.nextDouble() * 0.2 - 0.1, 0.2, this.random.nextDouble() * 0.2 - 0.1);
         }
diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 4f539b127f2b33710fffe51b8bf93ac9116c5134..ba3f839e98ee1601a37169cbfe8b545acac63486 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -69,7 +69,7 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     public PrimedTnt(final Level level, final double x, final double y, final double z, final @Nullable LivingEntity owner) {
         this(EntityType.TNT, level);
         this.setPos(x, y, z);
-        double rot = (org.leavesmc.leaves.LeavesConfig.modify.useVanillaRandom ? level.random.nextDouble() : this.getRandom().nextDouble()) * (float) (Math.PI * 2); // Paper - Don't use level random in entity constructors // Leaves - configurable vanilla random
+        double rot = (org.leavesmc.leaves.LeavesConfig.modify.useVanillaRandom ? level.getRandom().nextDouble() : this.getRandom().nextDouble()) * (float) (Math.PI * 2); // Paper - Don't use level random in entity constructors // Leaves - configurable vanilla random
         this.setDeltaMovement(-Math.sin(rot) * 0.02, 0.2F, -Math.cos(rot) * 0.02);
         this.setFuse(80);
         this.xo = x;
diff --git a/net/minecraft/world/entity/monster/Guardian.java b/net/minecraft/world/entity/monster/Guardian.java
index a34852bf64623cf94f6e788853a44d8579204279..1e7235856e6a80cfcd73df41f4e8b5b1724d7f3e 100644
--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -446,9 +446,9 @@ public class Guardian extends Monster {
     @Override
     public void thunderHit(net.minecraft.server.level.ServerLevel level, net.minecraft.world.entity.LightningBolt lightning) {
         if (org.leavesmc.leaves.LeavesConfig.modify.renewableSponges && !this.isRemoved() && !(this instanceof ElderGuardian)) {
-            ElderGuardian elderGuardian = new ElderGuardian(EntityType.ELDER_GUARDIAN ,this.level());
+            ElderGuardian elderGuardian = new ElderGuardian(EntityType.ELDER_GUARDIAN, this.level());
             elderGuardian.snapTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
-            elderGuardian.finalizeSpawn(level ,this.level().getCurrentDifficultyAt(elderGuardian.blockPosition()), EntitySpawnReason.CONVERSION, null);
+            elderGuardian.finalizeSpawn(level, level.getCurrentDifficultyAt(elderGuardian.blockPosition()), EntitySpawnReason.CONVERSION, null);
             elderGuardian.setNoAi(this.isNoAi());
 
             if (this.hasCustomName()) {
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 880df960f1c7d7f4a16914a94281a805997a1704..e6ac9d48b2e5cd87017cbd154a8619cb61b6ed2f 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -2074,7 +2074,7 @@ public abstract class Player extends Avatar implements ContainerUser {
 
                 // Leaves start - bow infinity fix
                 if ((org.leavesmc.leaves.LeavesConfig.modify.bowInfinityFix || anyEventCancelled.booleanValue()) && !this.abilities.instabuild && this instanceof final ServerPlayer player) this.resyncUsingItem(player); // Paper - resync if no item matched the Predicate
-                return this.hasInfiniteMaterials() || (org.leavesmc.leaves.LeavesConfig.modify.bowInfinityFix && net.minecraft.world.item.enchantment.EnchantmentHelper.processAmmoUse((ServerLevel) this.level(), shootable, new ItemStack(Items.ARROW), 1) <= 0) ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return this.hasInfiniteMaterials() || (org.leavesmc.leaves.LeavesConfig.modify.bowInfinityFix && net.minecraft.world.item.enchantment.EnchantmentHelper.processAmmoUse((ServerLevel) this.level(), heldWeapon, new ItemStack(Items.ARROW), 1) <= 0) ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
                 // Leaves end - bow infinity fix
             }
         }
diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index b0aeb50e3221ac7710f52116d9bd936e3f72e8be..168b42c878ebe2cc4beb62fc34c7895fdf95b1da 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -899,9 +899,9 @@ public class Raid {
         BlockPos.MutableBlockPos mutable = new BlockPos.MutableBlockPos();
         net.minecraft.world.entity.SpawnPlacementType spawnLocation = net.minecraft.world.entity.SpawnPlacements.getPlacementType(EntityType.RAVAGER);
         for (int j = 0; j < tries; j++) {
-            float f = serverWorld.random.nextFloat() * (float) (Math.PI * 2);
-            int k = this.center.getX() + Mth.floor(Mth.cos(f) * 32.0F * (float) i + serverWorld.random.nextInt(5));
-            int l = this.center.getZ() + Mth.floor(Mth.sin(f) * 32.0F * (float) i + serverWorld.random.nextInt(5));
+            float f = serverWorld.getRandom().nextFloat() * (float) (Math.PI * 2);
+            int k = this.center.getX() + Mth.floor(Mth.cos(f) * 32.0F * (float) i + serverWorld.getRandom().nextInt(5));
+            int l = this.center.getZ() + Mth.floor(Mth.sin(f) * 32.0F * (float) i + serverWorld.getRandom().nextInt(5));
             int m = serverWorld.getHeight(Heightmap.Types.WORLD_SURFACE, k, l);
             mutable.set(k, m, l);
             if (serverWorld.isVillage(mutable) && proximity < 2) {
diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index f3f299b93105051aece7b5ac117d87ab17deaf43..2ff46782b8d25fc0a039f753f787b7da5ae18ff8 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -157,7 +157,7 @@ public abstract class Raider extends PatrollingMonster {
                     amp = net.minecraft.util.Mth.clamp(amp, 0, 4);
                     net.minecraft.world.effect.MobEffectInstance mobeffect1 = new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.BAD_OMEN, 120000, amp, false, false, true);
 
-                    if (!serverLevel.getGameRules().getBoolean(net.minecraft.world.level.gamerules.GameRules.RULE_DISABLE_RAIDS)) {
+                    if (serverLevel.getGameRules().get(net.minecraft.world.level.gamerules.GameRules.RAIDS)) {
                         entityhuman.addEffect(mobeffect1, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                     }
                     this.setPatrolLeader(false);
diff --git a/net/minecraft/world/item/BlockItem.java b/net/minecraft/world/item/BlockItem.java
index 49cda2255105da6b177c2429e1e215b3499af721..54ecf4c118ffd5d20d65d13dcc8c99abed87cd9c 100644
--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -58,13 +58,13 @@ public class BlockItem extends Item {
                 BlockState placementState = this.getPlacementState(updatedPlaceContext);
                 final org.bukkit.block.BlockState previousState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(updatedPlaceContext.getLevel(), updatedPlaceContext.getClickedPos()); // Paper - Reset placed block on exception
                 // Leaves start - player operation limiter
-                if (org.leavesmc.leaves.LeavesConfig.modify.playerOperationLimiter && !context.getLevel().isClientSide()) {
-                    ServerPlayer limiterPlayer = (ServerPlayer) context.getPlayer();
+                if (org.leavesmc.leaves.LeavesConfig.modify.playerOperationLimiter && !placeContext.getLevel().isClientSide()) {
+                    ServerPlayer limiterPlayer = (ServerPlayer) placeContext.getPlayer();
                     if (limiterPlayer != null && placementState != null) {
                         limiterPlayer.addPlaceBlockCountPerTick();
                         if (!limiterPlayer.allowOperation()) {
                             if (previousState != null) {
-                                context.getLevel().getCraftServer().getPluginManager().callEvent(new org.leavesmc.leaves.event.player.PlayerOperationLimitEvent(limiterPlayer.getBukkitEntity(), org.leavesmc.leaves.event.player.PlayerOperationLimitEvent.Operation.PLACE, previousState.getBlock()));
+                                placeContext.getLevel().getCraftServer().getPluginManager().callEvent(new org.leavesmc.leaves.event.player.PlayerOperationLimitEvent(limiterPlayer.getBukkitEntity(), org.leavesmc.leaves.event.player.PlayerOperationLimitEvent.Operation.PLACE, previousState.getBlock()));
                             }
                             return InteractionResult.FAIL;
                         }
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index 12d7b38003e1fcf046249eef937f9bd1ec095273..1098695e646c20f47bc4739034fcb69b96909667 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -81,7 +81,7 @@ public class ShearsItem extends Item {
             return InteractionResult.SUCCESS;
         } else {
             // Leaves start - shears wrench
-            InteractionResult result = org.leavesmc.leaves.util.ShearsWrenchUtil.tryApplyRotate(context, blockState);
+            InteractionResult result = org.leavesmc.leaves.util.ShearsWrenchUtil.tryApplyRotate(context, state);
             return result == null ? super.useOn(context) : result;
             // Leaves end - shears wrench
         }
diff --git a/net/minecraft/world/item/ShovelItem.java b/net/minecraft/world/item/ShovelItem.java
index 0388960821e4aa2ff17015c6d5367c3c1c4d821c..b1e6f4bcb3080523fa56f35cd8cb1fae865051bf 100644
--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -50,11 +50,11 @@ public class ShovelItem extends Item {
                 boolean hasSilkTouch = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(level.registryAccess().lookupOrThrow(net.minecraft.core.registries.Registries.ENCHANTMENT).getOrThrow(net.minecraft.world.item.enchantment.Enchantments.SILK_TOUCH), tool) > 0;
                 BlockState shavedBlockState = layers > 1 ? blockState.setValue(net.minecraft.world.level.block.SnowLayerBlock.LAYERS, layers - 1) : Blocks.AIR.defaultBlockState();
 
-                level.setBlock(clickedPos, shavedBlockState, Block.UPDATE_ALL_IMMEDIATE);
-                level.gameEvent(GameEvent.BLOCK_CHANGE, clickedPos, GameEvent.Context.of(player, shavedBlockState));
+                level.setBlock(pos, shavedBlockState, Block.UPDATE_ALL_IMMEDIATE);
+                level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(player, shavedBlockState));
 
-                Block.popResource(level, clickedPos, new ItemStack(hasSilkTouch ? Items.SNOW : Items.SNOWBALL));
-                level.playSound(player, clickedPos, SoundEvents.SNOW_BREAK, SoundSource.BLOCKS, 1.0F, 1.0F);
+                Block.popResource(level, pos, new ItemStack(hasSilkTouch ? Items.SNOW : Items.SNOWBALL));
+                level.playSound(player, pos, SoundEvents.SNOW_BREAK, SoundSource.BLOCKS, 1.0F, 1.0F);
 
                 if (player != null) {
                     tool.hurtAndBreak(1, player, context.getHand());
diff --git a/net/minecraft/world/level/block/BuddingAmethystBlock.java b/net/minecraft/world/level/block/BuddingAmethystBlock.java
index bb94004a9e9798961dabf27eb74e82c568a30e33..d6ce752822d81da77aadf74a8bc1f4c30438f6d4 100644
--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -60,7 +60,6 @@ public class BuddingAmethystBlock extends AmethystBlock {
     }
 
     // Leaves start - budding amethyst can push by piston
-    @Override
     @org.jetbrains.annotations.Nullable
     public net.minecraft.world.level.material.PushReaction getResetPushReaction() {
         return org.leavesmc.leaves.LeavesConfig.modify.movableBuddingAmethyst ? net.minecraft.world.level.material.PushReaction.NORMAL : null;
