From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 18 May 2025 23:48:59 +0800
Subject: [PATCH] Old nether portal collision

It will be removed in 1.21.6 mojang revert it again.

diff --git a/net/minecraft/world/level/block/NetherPortalBlock.java b/net/minecraft/world/level/block/NetherPortalBlock.java
index 3aabae6e27e692aa65cc931e57306426e0f4d645..76dc4eb719244f002e327919c6b8aac8b359b5f0 100644
--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -60,7 +60,7 @@ public class NetherPortalBlock extends Block implements Portal {
 
     @Override
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
-        return SHAPES.get(state.getValue(AXIS));
+        return org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldNetherPortalCollision ? Shapes.block() : SHAPES.get(state.getValue(AXIS)); // Leaves - Old nether portal collision
     }
 
     @Override
