From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 18 May 2025 23:48:59 +0800
Subject: [PATCH] Old nether portal collision

It will be removed in 1.21.6 mojang revert it again.

diff --git a/net/minecraft/world/level/block/NetherPortalBlock.java b/net/minecraft/world/level/block/NetherPortalBlock.java
index 171b383efabbbe849aff28832c47076f85a46307..173e97c27e9373a41d33cda6a877b17a38f516a1 100644
--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -65,7 +65,7 @@ public class NetherPortalBlock extends Block implements Portal {
 
     @Override
     protected VoxelShape getEntityInsideCollisionShape(BlockState state, BlockGetter level, BlockPos pos, Entity entity) {
-        return state.getShape(level, pos);
+        return org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldNetherPortalCollision ? Shapes.block() : state.getShape(level, pos); // Leaves - Old nether portal collision
     }
 
     @Override
