From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leaves <leaves@leaves>
Date: Thu, 16 Apr 2026 11:15:32 +0800
Subject: [PATCH 03/11] CCE update suppression


diff --git a/net/minecraft/world/level/block/ShulkerBoxBlock.java b/net/minecraft/world/level/block/ShulkerBoxBlock.java
index ad6a14cd898dd837e45248f608b1ae71f7a89bab..552edae124e744f661da01f677301c69e4a78fad 100644
--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -183,7 +183,9 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
     protected int getAnalogOutputSignal(final BlockState state, final Level level, final BlockPos pos, final Direction direction) {
         // Leaves start - update suppression crash fix
         try {
-            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
+            return org.leavesmc.leaves.LeavesConfig.modify.oldMC.updater.cceUpdateSuppression ?
+                    AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) level.getBlockEntity(pos)) : // Leaves - make cce happy(?)
+                    AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
         } catch (ClassCastException ex) {
             if (org.leavesmc.leaves.LeavesConfig.modify.updateSuppressionCrashFix) {
                 throw new org.leavesmc.leaves.util.UpdateSuppressionException(pos, null, this, null, ex);
