From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Sun, 18 May 2025 03:21:56 +0000
Subject: [PATCH] Do not prevent block entity and entity crash at LevelChunk


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 2643e1f8c330f7e7e682d2304a2c97f5f14d8bff..42f32d4a76e2c854db401e06b51f786efd8ae3ba 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -944,12 +944,14 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
                         profilerFiller.pop();
                     } catch (Throwable var5) {
+                        // Leaves start - do not prevent here
                         // Paper start - Prevent block entity and entity crashes
-                        final String msg = String.format("BlockEntity threw exception at %s:%s,%s,%s", LevelChunk.this.getLevel().getWorld().getName(), this.getPos().getX(), this.getPos().getY(), this.getPos().getZ());
-                        net.minecraft.server.MinecraftServer.LOGGER.error(msg, var5);
-                        net.minecraft.world.level.chunk.LevelChunk.this.level.getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, var5))); // Paper - ServerExceptionEvent
-                        LevelChunk.this.removeBlockEntity(this.getPos());
+                        // final String msg = String.format("BlockEntity threw exception at %s:%s,%s,%s", LevelChunk.this.getLevel().getWorld().getName(), this.getPos().getX(), this.getPos().getY(), this.getPos().getZ());
+                        // net.minecraft.server.MinecraftServer.LOGGER.error(msg, var5);
+                        // net.minecraft.world.level.chunk.LevelChunk.this.level.getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, var5))); // Paper - ServerExceptionEvent
+                        // LevelChunk.this.removeBlockEntity(this.getPos());
                         // Paper end - Prevent block entity and entity crashes
+                        // Leaves end - do not prevent here
                     }
                 }
             }
