From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:42:17 +0800
Subject: [PATCH] Old zombie reinforcement


diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index a23607874a72723914cbfeea0ad1c51236c044d8..cb7b248276bfb6a9721e0bcc42ef00bf7029f263 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -341,7 +341,7 @@ public class Zombie extends Monster {
                 int floor = Mth.floor(this.getX());
                 int floor1 = Mth.floor(this.getY());
                 int floor2 = Mth.floor(this.getZ());
-                EntityType<? extends Zombie> type = this.getType();
+                EntityType<? extends Zombie> type = org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldZombieReinforcement ? EntityType.ZOMBIE : this.getType(); // Leaves end - old zombie reinforcement
                 Zombie zombie = type.create(level, EntitySpawnReason.REINFORCEMENT);
                 if (zombie == null) {
                     return true;
