From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:06:03 +0800
Subject: [PATCH] Spawn invulnerable time


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index d4984c919a88fa930dfec823cd9b471fa47f3565..8a622da26cff56f404b7d8b6131c27c65a31cdfe 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -474,6 +474,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.adventure$displayName = org.leavesmc.leaves.LeavesConfig.fix.vanillaDisplayName ? io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getDisplayName()) : net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper // Leaves - Vanilla display name
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.spawnInvulnerableTime) this.invulnerableTime = 60; // Leaves - spawn invulnerable time
     }
 
     @Override
