From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Fri, 4 Jul 2025 08:55:31 +0800
Subject: [PATCH] Old wet tnt explode behavior


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 1e94326047622d8975fd1cc5f44e75104fd56946..b6c289e9615753b2da66c995f2f950e874514ad9 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -717,6 +717,7 @@ public class ServerExplosion implements Explosion {
     public boolean shouldAffectBlocklikeEntities() {
         boolean flag = this.level.getGameRules().get(GameRules.MOB_GRIEFING);
         boolean flag1 = this.source == null || this.source.getType() != EntityType.BREEZE_WIND_CHARGE && this.source.getType() != EntityType.WIND_CHARGE;
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.tntWetExplosionNoItemDamage) flag1 = flag1 && (this.source == null || !this.source.isInWater()); // Leaves - Old MC TNT wet explosion no item damage
         return flag ? flag1 : this.blockInteraction.shouldAffectBlocklikeEntities() && flag1;
     }
 
