From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Fri, 4 Jul 2025 17:13:15 +0800
Subject: [PATCH] Prevent loss of item drops due to update suppression when
 breaking blocks


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index 24632205303bbcb2240a851b9d8baf8bbdfe4622..caf753c700ae2542627ff60e420bc40ab04ae013 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -399,11 +399,21 @@ public class ServerPlayerGameMode {
                 this.level.captureDrops = new java.util.ArrayList<>();
                 // CraftBukkit end
                 BlockState blockState1 = org.leavesmc.leaves.command.leaves.subcommands.BlockUpdateCommand.isNoBlockUpdate() ? blockState : block.playerWillDestroy(this.level, pos, blockState, this.player); // Leaves - no block update
-                boolean flag = this.level.removeBlock(pos, false);
+                // Leaves start - Prevent loss of item drops due to update suppression when breaking blocks
+                boolean flag;
                 if (SharedConstants.DEBUG_BLOCK_BREAK) {
                     LOGGER.info("server broke {} {} -> {}", pos, blockState1, this.level.getBlockState(pos));
                 }
-
+                org.leavesmc.leaves.util.UpdateSuppressionException ex = null;
+                try {
+                    flag = this.level.removeBlock(pos, false);
+                } catch (org.leavesmc.leaves.util.UpdateSuppressionException e) {
+                    ex = e;
+                    ex.provideBlock(level, pos, block);
+                    ex.providePlayer(this.player);
+                    flag = false;
+                }
+                // Leaves end - Prevent loss of item drops due to update suppression when breaking blocks
                 if (flag) {
                     block.destroy(this.level, pos, blockState1);
                 }
@@ -431,6 +441,7 @@ public class ServerPlayerGameMode {
                 if (event.isDropItems()) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, this.player, itemsToDrop); // Paper - capture all item additions to the world
                 }
+                if (ex != null) throw ex; // Leaves - Prevent loss of item drops due to update suppression when breaking blocks
 
                 // Drop event experience
                 if (flag) {
