From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Fri, 4 Jul 2025 08:55:31 +0800
Subject: [PATCH] Old wet tnt explode behavior


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 1c521f9f32340cf75310686c90777e521ac3ae5c..8b0341c060d9f61698f32dd05c360c303a443c9b 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -719,6 +719,7 @@ public class ServerExplosion implements Explosion {
     public boolean shouldAffectBlocklikeEntities() {
         boolean _boolean = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
         boolean flag = this.source == null || this.source.getType() != EntityType.BREEZE_WIND_CHARGE && this.source.getType() != EntityType.WIND_CHARGE;
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.tntWetExplosionNoItemDamage) flag = flag && (this.source == null || !this.source.isInWater()); // Leaves - Old MC TNT wet explosion no item damage
         return _boolean ? flag : this.blockInteraction.shouldAffectBlocklikeEntities() && flag;
     }
 
