From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Fri, 4 Jul 2025 09:11:46 +0800
Subject: [PATCH] Old projectile explosion behavior


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 8b0341c060d9f61698f32dd05c360c303a443c9b..8447b824ad5b68c5af5507c3594cd74ffdc40e73 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -539,7 +539,7 @@ public class ServerExplosion implements Explosion {
                     }
                     // CraftBukkit end
-                    entity.push(vec32);
+                    if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldProjectileExplosionBehavior) entity.setDeltaMovement(entity.getDeltaMovement().add(vec32)); else entity.push(vec32); // Leaves - old projectile explosion behavior
                     if (entity.getType().is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
                         projectile.setOwner(this.damageSource.getEntity());
                     } else if (entity instanceof Player player && !player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Option to disable explosion knockback
