From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Sat, 5 Jul 2025 09:48:47 +0800
Subject: [PATCH] Fix block place desync due to update suppression


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 2d01258e377ad1d628b193d15d794a057c3d11b4..f0a27175e044f49f47ac012e0d81b61832ff5af5 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -395,8 +395,18 @@ public final class ItemStack implements DataComponentHolder {
                 }
             }
             InteractionResult interactionResult;
+            // Leaves start - fix block place desync due to update suppression
+            org.leavesmc.leaves.util.UpdateSuppressionException ue = null;
             try {
                 interactionResult = item.useOn(context);
+            } catch (org.leavesmc.leaves.util.UpdateSuppressionException ex) {
+                interactionResult = net.minecraft.world.InteractionResult.SUCCESS.configurePaper(e -> e.placedBlockAt(clickedPos.immutable()));
+                ue = ex;
+                ue.provideBlock(context.getLevel(), context.getClickedPos(), serverLevel.getBlockState(context.getClickedPos()).getBlock());
+                if (player != null) {
+                    ue.providePlayer((net.minecraft.server.level.ServerPlayer) player);
+                }
+            // Leaves end - fix block place desync due to update suppression
             } finally {
                 serverLevel.captureBlockStates = false;
             }
@@ -543,6 +553,7 @@ public final class ItemStack implements DataComponentHolder {
             serverLevel.capturedBlockStates.clear();
             // CraftBukkit end
 
+            if (ue != null) throw ue; // Leaves - fix block place desync due to update suppression
             return interactionResult;
         }
     }
