From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Thu, 24 Oct 2024 23:10:34 +0800
Subject: [PATCH] Configurable collision behavior


diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index de764da80793d272ff597b4a0742ce785838799a..c818a3a64e5f7deadd4a1ca702dd396a7c3fedde 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -2022,7 +2022,7 @@ public final class CollisionUtil {
                                                 continue;
                                             }
                                         } else {
-                                            if (!voxelShapeIntersect(aabb, singleAABB)) {
+                                            if (shouldSkip(aabb, singleAABB)) { // Leaves - Configurable collision behavior
                                                 continue;
                                             }
                                         }
@@ -2076,6 +2076,16 @@ public final class CollisionUtil {
         return ret;
     }
 
+    // Leaves start - Configurable collision behavior
+    private static boolean shouldSkip(net.minecraft.world.phys.AABB aabb, net.minecraft.world.phys.AABB singleAABB) {
+        return switch (org.leavesmc.leaves.LeavesConfig.fix.collisionBehavior) {
+            // Paper has changed to block shape vanilla, so we only check fully vanilla here
+            case PAPER -> !voxelShapeIntersect(aabb, singleAABB);
+            case VANILLA -> !aabb.intersects(singleAABB);
+        };
+    }
+    // Leaves end - Configurable collision behavior
+
     public static boolean getEntityHardCollisions(final Level world, final Entity entity, AABB aabb,
                                                   final List<AABB> into, final int collisionFlags, final Predicate<Entity> predicate) {
         final boolean checkOnly = (collisionFlags & COLLISION_FLAG_CHECK_ONLY) != 0;
