From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Thu, 19 Jun 2025 00:49:24 +0800
Subject: [PATCH] Vanilla Fluid Pushing


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index b42b0788b7df9f39429d18ce12b525beb7609bd5..aba79f3848ab826869d4050ce02b3483b91f34e2 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4823,8 +4823,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
                                 final Vec3 flow = fluidState.getFlow(world, mutablePos);
 
-                                if (diff < 0.4) {
-                                    pushVector = pushVector.add(flow.scale(diff));
+                                // Leaves start - Vanilla Fluid Pushing
+                                if (maxHeightDiff < 0.4) {
+                                    pushVector = pushVector.add(flow.scale(maxHeightDiff));
+                                // Leaves end - Vanilla Fluid Pushing
                                 } else {
                                     pushVector = pushVector.add(flow);
                                 }
