From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 13:51:26 +0800
Subject: [PATCH] Alternative block placement Protocol


diff --git a/net/minecraft/world/item/BlockItem.java b/net/minecraft/world/item/BlockItem.java
index 54ecf4c118ffd5d20d65d13dcc8c99abed87cd9c..8dfe237a01ec026fdee388f3e8b8387830bfafd4 100644
--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -152,7 +152,7 @@ public class BlockItem extends Item {
     }
 
     protected @Nullable BlockState getPlacementState(final BlockPlaceContext context) {
-        BlockState stateForPlacement = this.getBlock().getStateForPlacement(context);
+        BlockState stateForPlacement = this.getBlock().getRealStateForPlacement(context);
         return stateForPlacement != null && this.canPlace(context, stateForPlacement) ? stateForPlacement : null;
     }
 
diff --git a/net/minecraft/world/item/StandingAndWallBlockItem.java b/net/minecraft/world/item/StandingAndWallBlockItem.java
index 0d6f5941e2e9d8007cf3b30a51994a03d40b2449..0e9a48c5cacbe2dd97af1b4e994eb69509160043 100644
--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -26,14 +26,14 @@ public class StandingAndWallBlockItem extends BlockItem {
 
     @Override
     protected @Nullable BlockState getPlacementState(final BlockPlaceContext context) {
-        BlockState wallState = this.wallBlock.getStateForPlacement(context);
+        BlockState wallState = this.wallBlock.getRealStateForPlacement(context);
         BlockState stateForPlacement = null;
         LevelReader level = context.getLevel();
         BlockPos pos = context.getClickedPos();
 
         for (Direction direction : context.getNearestLookingDirections()) {
             if (direction != this.attachmentDirection.getOpposite()) {
-                BlockState possibleState = direction == this.attachmentDirection ? this.getBlock().getStateForPlacement(context) : wallState;
+                BlockState possibleState = direction == this.attachmentDirection ? this.getBlock().getRealStateForPlacement(context) : wallState;
                 if (possibleState != null && this.canPlace(level, possibleState, pos)) {
                     stateForPlacement = possibleState;
                     break;
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index 070dd284188936ec791f13371757700d164104bf..23a3d0f23375e2aa59c9b709971a4018c0de012c 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -516,6 +516,33 @@ public class Block extends BlockBehaviour implements ItemLike {
     public void stepOn(final Level level, final BlockPos pos, final BlockState onState, final Entity entity) {
     }
 
+    // Leaves start - alternativeBlockPlacement
+    @Nullable
+    public BlockState getRealStateForPlacement(BlockPlaceContext ctx) {
+        BlockState vanillaState = this.getStateForPlacement(ctx);
+        switch (org.leavesmc.leaves.LeavesConfig.protocol.alternativeBlockPlacement) {
+            case CARPET -> {
+                BlockState tryState = org.leavesmc.leaves.protocol.CarpetAlternativeBlockPlacement.alternativeBlockPlacement(this, ctx);
+                if (tryState != null) {
+                    return tryState;
+                }
+            }
+            case CARPET_FIX -> {
+                BlockState tryState = org.leavesmc.leaves.protocol.CarpetAlternativeBlockPlacement.alternativeBlockPlacementFix(this, ctx);
+                if (tryState != null) {
+                    return tryState;
+                }
+            }
+            case LITEMATICA -> {
+                if (vanillaState != null) {
+                    return org.leavesmc.leaves.protocol.LitematicaEasyPlaceProtocol.applyPlacementProtocol(vanillaState, ctx);
+                }
+            }
+        }
+        return vanillaState;
+    }
+    // Leaves end - alternativeBlockPlacement
+
     public @Nullable BlockState getStateForPlacement(final BlockPlaceContext context) {
         return this.defaultBlockState();
     }
