From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 30 Jul 2025 17:58:24 +0800
Subject: [PATCH] Chat command max length


diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
index 073e369fc21d1e098cb680638859a37b154895a4..4365970ec12a2f52250834d4a0c4ef464ea80db3 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
@@ -12,11 +12,11 @@ public record ServerboundChatCommandPacket(String command) implements Packet<Ser
     );
 
     private ServerboundChatCommandPacket(FriendlyByteBuf buffer) {
-        this(buffer.readUtf(MAX_CHAT_PACKET_INPUT_SIZE)); // Paper - limit chat command inputs
+        this(buffer.readUtf(org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength)); // Leaves - chat command max length
     }
 
     private void write(FriendlyByteBuf buffer) {
-        buffer.writeUtf(this.command);
+        buffer.writeUtf(this.command, org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength); // Leaves - chat command max length
     }
 
     @Override
diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
index 7f24fd1f5ff5100aeb404dc049af1893be824047..d1976d89ddc29e77db498c50f961634e76868521 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -16,11 +16,11 @@ public record ServerboundChatCommandSignedPacket(
     );
 
     private ServerboundChatCommandSignedPacket(FriendlyByteBuf buffer) {
-        this(buffer.readUtf(net.minecraft.network.protocol.game.ServerboundChatCommandPacket.MAX_CHAT_PACKET_INPUT_SIZE), buffer.readInstant(), buffer.readLong(), new ArgumentSignatures(buffer), new LastSeenMessages.Update(buffer)); // Paper - limit chat command inputs
+        this(buffer.readUtf(org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength), buffer.readInstant(), buffer.readLong(), new ArgumentSignatures(buffer), new LastSeenMessages.Update(buffer)); // Leaves - chat command max length
     }
 
     private void write(FriendlyByteBuf buffer) {
-        buffer.writeUtf(this.command);
+        buffer.writeUtf(this.command, org.leavesmc.leaves.LeavesConfig.mics.chatCommandMaxLength); // Leaves - chat command max length
         buffer.writeInstant(this.timeStamp);
         buffer.writeLong(this.salt);
         this.argumentSignatures.write(buffer);
