From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Sat, 5 Jul 2025 11:31:44 +0800
Subject: [PATCH] TEMP Merge Paper#11831

Dec 27, 2024 ~ Today

diff --git a/net/minecraft/server/commands/GiveCommand.java b/net/minecraft/server/commands/GiveCommand.java
index f6e18d62db18616a422ae74d6f329d1b3a96300b..eb34f3f9821f50f0a31fb22d9314c26b3e829d47 100644
--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -54,7 +54,6 @@ public class GiveCommand {
 
     private static int giveItem(CommandSourceStack source, ItemInput item, Collection<ServerPlayer> targets, int count) throws CommandSyntaxException {
         ItemStack itemStack = item.createItemStack(1, false);
-        final Component displayName = itemStack.getDisplayName(); // Paper - get display name early
         int maxStackSize = org.leavesmc.leaves.util.ItemOverstackUtils.getItemStackMaxCount(itemStack); // Leaves - item over-stack util
         int i = maxStackSize * 100;
         if (count > i) {
@@ -99,11 +98,11 @@ public class GiveCommand {
 
             if (targets.size() == 1) {
                 source.sendSuccess(
-                    () -> Component.translatable("commands.give.success.single", count, displayName, targets.iterator().next().getDisplayName()), // Paper - use cached display name
+                    () -> Component.translatable("commands.give.success.single", count, itemStack.getDisplayName(), targets.iterator().next().getDisplayName()), // Paper - use cached display name
                     true
                 );
             } else {
-                source.sendSuccess(() -> Component.translatable("commands.give.success.single", count, displayName, targets.size()), true); // Paper - use cached display name
+                source.sendSuccess(() -> Component.translatable("commands.give.success.single", count, itemStack.getDisplayName(), targets.size()), true); // Paper - use cached display name
             }
 
             return targets.size();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 5c63b9603f7b6f6c4ceca0d836df4a4972568631..01b4bcf6f975a9d7d083909f23c9bfb02fc074b2 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2888,10 +2888,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 return null;
             }
             // CraftBukkit end
-            ItemEntity itemEntity = new ItemEntity(level, this.getX() + offset.x, this.getY() + offset.y, this.getZ() + offset.z, stack.copy()); // Paper - copy so we can destroy original
-            stack.setCount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe
-
-            itemEntity.setDefaultPickUpDelay(); // Paper - diff on change (in dropConsumer)
+            ItemEntity itemEntity = new ItemEntity(level, this.getX() + offset.x, this.getY() + offset.y, this.getZ() + offset.z, stack.copy()); // Paper - copy
             // Paper start - Call EntityDropItemEvent
             return this.spawnAtLocation(level, itemEntity);
         }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 0610df1c343f1c5cf724a79fbc68ed1372de5da5..7a849ba0e22574f00f3580a432e2fe8343c883c6 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3996,11 +3996,6 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             return null;
         } else {
             double d = this.getEyeY() - 0.3F;
-            // Paper start
-            final ItemStack tmp = stack.copy();
-            stack.setCount(0);
-            stack = tmp;
-            // Paper end
             ItemEntity itemEntity = new ItemEntity(this.level(), this.getX(), d, this.getZ(), stack);
             itemEntity.setPickUpDelay(40);
             if (includeThrower) {
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index 509d368f7fa9e6330da02615d7bf78debb932623..5ac49ee4743f7c4847e40dae6575f559df25abeb 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -652,7 +652,11 @@ public abstract class AbstractContainerMenu {
                     }
 
                     carried = slot2.safeTake(i1, Integer.MAX_VALUE, player);
-                    player.drop(carried, true);
+                    // CraftBukkit start - SPIGOT-8010: break loop
+                    if (player.drop(carried, true) == null) {
+                        break;
+                    }
+                    // CraftBukkit end - SPIGOT-8010: break loop
                     player.handleCreativeModeItemDrop(carried);
                 }
             }
