From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leaves <leaves@leaves>
Date: Thu, 16 Apr 2026 11:30:00 +0800
Subject: [PATCH 10/11] Xaero Map Protocol


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index e52e02388a0e8db070b04452ee757530541b4bd8..e2c47f4e0e66d8d2a832d925a586956df11664f9 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1023,6 +1023,7 @@ public abstract class PlayerList {
         //player.connection.send(new ClientboundSetChunkCacheRadiusPacket(io.papermc.paper.FeatureHooks.getViewDistance(level))); // Paper - rewrite chunk system
         //player.connection.send(new ClientboundSetSimulationDistancePacket(io.papermc.paper.FeatureHooks.getSimulationDistance(level))); // Paper - rewrite chunk system
         // Paper end - view distances
+        org.leavesmc.leaves.protocol.XaeroMapProtocol.onSendWorldInfo(player); // Leaves - xaero map protocol
         if (level.isRaining()) {
             // CraftBukkit start - handle player weather
             // player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
