From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leaves <leaves@leaves>
Date: Thu, 16 Apr 2026 11:30:00 +0800
Subject: [PATCH] Support REI protocol


diff --git a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
index 33f9870a4f966a887023b1debd4ab9ef9d24378f..432bfa6991919a2803846838a1dfe3415ae668c2 100644
--- a/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTransformRecipe.java
@@ -115,6 +115,12 @@ public class SmithingTransformRecipe extends SimpleSmithingRecipe {
         );
     }
 
+    // Leaves start - REI
+    public SlotDisplay getResult() {
+        return this.result.display();
+    }
+    // Leaves end - REI
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
diff --git a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
index f9fa7d0eaf9f32959a4bd12b82e5a49ea808a44c..2f3241563ad7e5ed5663b913671d2af056df97bb 100644
--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -109,6 +109,12 @@ public class SmithingTrimRecipe extends SimpleSmithingRecipe {
         return Optional.of(this.addition);
     }
 
+    // Leaves start
+    public Holder<TrimPattern> pattern() {
+        return pattern;
+    }
+    // Leaves end
+
     @Override
     public RecipeSerializer<SmithingTrimRecipe> getSerializer() {
         return SERIALIZER;
