From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Thu, 14 Aug 2025 15:47:58 +0800
Subject: [PATCH] Do not tick Arrow life regardless


diff --git a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
index 51d3560c5d8cba89469c07fe6b13f0016adc6db7..cae4bb2d630131ba5eda5ee7bffc39b8ef1f30e2 100644
--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -218,10 +218,14 @@ public abstract class AbstractArrow extends Projectile {
                 this.setSharedFlagOnFire(this.getRemainingFireTicks() > 0);
             }
         } else {
+            // Leaves start - do not tick arrow life regardless
+            /*
             // Paper start - tick life regardless after X seconds
             final io.papermc.paper.configuration.type.number.IntOr.Disabled maxArrowDespawnInvulnerability = this.level().paperConfig().entities.spawning.maxArrowDespawnInvulnerability;
             if (maxArrowDespawnInvulnerability.enabled() && this.tickCount > maxArrowDespawnInvulnerability.intValue()) this.tickDespawn();
             // Paper end - tick life regardless after X seconds
+             */
+            // Leaves end - do not tick arrow life regardless
             this.inGroundTime = 0;
             Vec3 vec31 = this.position();
             if (this.isInWater()) {
