From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leaves <leaves@leaves>
Date: Thu, 16 Apr 2026 13:02:40 +0800
Subject: [PATCH] Optimize noise generation


diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
index ef8f2a00945a3d240a23ca73f138960ee0fe8742..d31fd89d8127b898486a8a455115498756900c64 100644
--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -11,6 +11,27 @@ public final class ImprovedNoise {
     public final double yo;
     public final double zo;
 
+    // Leaves start - optimize noise generation
+    private static final double[] FLAT_SIMPLEX_GRAD = new double[]{
+        1, 1, 0, 0,
+        -1, 1, 0, 0,
+        1, -1, 0, 0,
+        -1, -1, 0, 0,
+        1, 0, 1, 0,
+        -1, 0, 1, 0,
+        1, 0, -1, 0,
+        -1, 0, -1, 0,
+        0, 1, 1, 0,
+        0, -1, 1, 0,
+        0, 1, -1, 0,
+        0, -1, -1, 0,
+        1, 1, 0, 0,
+        0, -1, 1, 0,
+        -1, 1, 0, 0,
+        0, -1, -1, 0,
+    };
+    // Leaves end - optimize noise generation
+
     public ImprovedNoise(final RandomSource random) {
         this.xo = random.nextDouble() * 256.0;
         this.yo = random.nextDouble() * 256.0;
@@ -38,40 +59,78 @@ public final class ImprovedNoise {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
-        double xr = x - xf;
-        double yr = y - yf;
-        double zr = z - zf;
-        double yrFudge;
-        if (yScale != 0.0) {
-            double fudgeLimit;
-            if (yFudge >= 0.0 && yFudge < yr) {
-                fudgeLimit = yFudge;
+        // Leaves start - optimize noise generation
+        if (!org.leavesmc.leaves.LeavesConfig.performance.optimizeNoiseGeneration) {
+            int xf = Mth.floor(x);
+            int yf = Mth.floor(y);
+            int zf = Mth.floor(z);
+            double xr = x - xf;
+            double yr = y - yf;
+            double zr = z - zf;
+            double yrFudge;
+            if (yScale != 0.0) {
+                double fudgeLimit;
+                if (yFudge >= 0.0 && yFudge < yr) {
+                    fudgeLimit = yFudge;
+                } else {
+                    fudgeLimit = yr;
+                }
+
+                yrFudge = Mth.floor(fudgeLimit / yScale + 1.0E-7F) * yScale;
             } else {
-                fudgeLimit = yr;
+                yrFudge = 0.0;
             }
 
-            yrFudge = Mth.floor(fudgeLimit / yScale + 1.0E-7F) * yScale;
+            return this.sampleAndLerp(xf, yf, zf, xr, yr - yrFudge, zr, yr);
         } else {
-            yrFudge = 0.0;
-        }
+            double xf = Math.floor(x);
+            double yf = Math.floor(y);
+            double zf = Math.floor(z);
+            double xr = x - xf;
+            double yr = y - yf;
+            double zr = z - zf;
+            double yrFudge;
+            if (yScale != 0.0) {
+                double fudgeLimit;
+                if (yFudge >= 0.0 && yFudge < yr) {
+                    fudgeLimit = yFudge;
+                } else {
+                    fudgeLimit = yr;
+                }
+
+                yrFudge = Math.floor(fudgeLimit / yScale + 1.0E-7F) * yScale;
+            } else {
+                yrFudge = 0.0;
+            }
 
-        return this.sampleAndLerp(xf, yf, zf, xr, yr - yrFudge, zr, yr);
+            return this.sampleAndLerp((int) xf, (int) yf, (int) zf, xr, yr - yrFudge, zr, yr);
+        }
+        // Leaves end - optimize noise generation
     }
 
     public double noiseWithDerivative(final double _x, final double _y, final double _z, final double[] derivativeOut) {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
-        double xr = x - xf;
-        double yr = y - yf;
-        double zr = z - zf;
-        return this.sampleWithDerivative(xf, yf, zf, xr, yr, zr, derivativeOut);
+        // Leaves start - optimize noise generation
+        if (!org.leavesmc.leaves.LeavesConfig.performance.optimizeNoiseGeneration) {
+            int xf = Mth.floor(x);
+            int yf = Mth.floor(y);
+            int zf = Mth.floor(z);
+            double xr = x - xf;
+            double yr = y - yf;
+            double zr = z - zf;
+            return this.sampleWithDerivative(xf, yf, zf, xr, yr, zr, derivativeOut);
+        } else {
+            double xf = Math.floor(x);
+            double yf = Math.floor(y);
+            double zf = Math.floor(z);
+            double xr = x - xf;
+            double yr = y - yf;
+            double zr = z - zf;
+            return this.sampleWithDerivative((int) xf, (int) yf, (int) zf, xr, yr, zr, derivativeOut);
+        }
+        // Leaves end - optimize noise generation
     }
 
     private static double gradDot(final int hash, final double x, final double y, final double z) {
@@ -83,24 +142,90 @@ public final class ImprovedNoise {
     }
 
     private double sampleAndLerp(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double yrOriginal) {
-        int x0 = this.p(x);
-        int x1 = this.p(x + 1);
-        int xy00 = this.p(x0 + y);
-        int xy01 = this.p(x0 + y + 1);
-        int xy10 = this.p(x1 + y);
-        int xy11 = this.p(x1 + y + 1);
-        double d000 = gradDot(this.p(xy00 + z), xr, yr, zr);
-        double d100 = gradDot(this.p(xy10 + z), xr - 1.0, yr, zr);
-        double d010 = gradDot(this.p(xy01 + z), xr, yr - 1.0, zr);
-        double d110 = gradDot(this.p(xy11 + z), xr - 1.0, yr - 1.0, zr);
-        double d001 = gradDot(this.p(xy00 + z + 1), xr, yr, zr - 1.0);
-        double d101 = gradDot(this.p(xy10 + z + 1), xr - 1.0, yr, zr - 1.0);
-        double d011 = gradDot(this.p(xy01 + z + 1), xr, yr - 1.0, zr - 1.0);
-        double d111 = gradDot(this.p(xy11 + z + 1), xr - 1.0, yr - 1.0, zr - 1.0);
-        double xAlpha = Mth.smoothstep(xr);
-        double yAlpha = Mth.smoothstep(yrOriginal);
-        double zAlpha = Mth.smoothstep(zr);
-        return Mth.lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
+        // Leaves start - optimize noise generation
+        if (!org.leavesmc.leaves.LeavesConfig.performance.optimizeNoiseGeneration) {
+            int x0 = this.p(x);
+            int x1 = this.p(x + 1);
+            int xy00 = this.p(x0 + y);
+            int xy01 = this.p(x0 + y + 1);
+            int xy10 = this.p(x1 + y);
+            int xy11 = this.p(x1 + y + 1);
+            double d000 = gradDot(this.p(xy00 + z), xr, yr, zr);
+            double d100 = gradDot(this.p(xy10 + z), xr - 1.0, yr, zr);
+            double d010 = gradDot(this.p(xy01 + z), xr, yr - 1.0, zr);
+            double d110 = gradDot(this.p(xy11 + z), xr - 1.0, yr - 1.0, zr);
+            double d001 = gradDot(this.p(xy00 + z + 1), xr, yr, zr - 1.0);
+            double d101 = gradDot(this.p(xy10 + z + 1), xr - 1.0, yr, zr - 1.0);
+            double d011 = gradDot(this.p(xy01 + z + 1), xr, yr - 1.0, zr - 1.0);
+            double d111 = gradDot(this.p(xy11 + z + 1), xr - 1.0, yr - 1.0, zr - 1.0);
+            double xAlpha = Mth.smoothstep(xr);
+            double yAlpha = Mth.smoothstep(yrOriginal);
+            double zAlpha = Mth.smoothstep(zr);
+            return Mth.lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
+        } else {
+            final int var0 = x & 0xFF;
+            final int var1 = (x + 1) & 0xFF;
+            final int var2 = this.p[var0] & 0xFF;
+            final int var3 = this.p[var1] & 0xFF;
+            final int var4 = (var2 + y) & 0xFF;
+            final int var5 = (var3 + y) & 0xFF;
+            final int var6 = (var2 + y + 1) & 0xFF;
+            final int var7 = (var3 + y + 1) & 0xFF;
+            final int var8 = this.p[var4] & 0xFF;
+            final int var9 = this.p[var5] & 0xFF;
+            final int var10 = this.p[var6] & 0xFF;
+            final int var11 = this.p[var7] & 0xFF;
+
+            final int var12 = (var8 + z) & 0xFF;
+            final int var13 = (var9 + z) & 0xFF;
+            final int var14 = (var10 + z) & 0xFF;
+            final int var15 = (var11 + z) & 0xFF;
+            final int var16 = (var8 + z + 1) & 0xFF;
+            final int var17 = (var9 + z + 1) & 0xFF;
+            final int var18 = (var10 + z + 1) & 0xFF;
+            final int var19 = (var11 + z + 1) & 0xFF;
+            final int var20 = (this.p[var12] & 15) << 2;
+            final int var21 = (this.p[var13] & 15) << 2;
+            final int var22 = (this.p[var14] & 15) << 2;
+            final int var23 = (this.p[var15] & 15) << 2;
+            final int var24 = (this.p[var16] & 15) << 2;
+            final int var25 = (this.p[var17] & 15) << 2;
+            final int var26 = (this.p[var18] & 15) << 2;
+            final int var27 = (this.p[var19] & 15) << 2;
+            final double var60 = xr - 1.0;
+            final double var61 = yr - 1.0;
+            final double var62 = zr - 1.0;
+            final double var87 = FLAT_SIMPLEX_GRAD[(var20) | 0] * xr + FLAT_SIMPLEX_GRAD[(var20) | 1] * yr + FLAT_SIMPLEX_GRAD[(var20) | 2] * zr;
+            final double var88 = FLAT_SIMPLEX_GRAD[(var21) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var21) | 1] * yr + FLAT_SIMPLEX_GRAD[(var21) | 2] * zr;
+            final double var89 = FLAT_SIMPLEX_GRAD[(var22) | 0] * xr + FLAT_SIMPLEX_GRAD[(var22) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var22) | 2] * zr;
+            final double var90 = FLAT_SIMPLEX_GRAD[(var23) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var23) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var23) | 2] * zr;
+            final double var91 = FLAT_SIMPLEX_GRAD[(var24) | 0] * xr + FLAT_SIMPLEX_GRAD[(var24) | 1] * yr + FLAT_SIMPLEX_GRAD[(var24) | 2] * var62;
+            final double var92 = FLAT_SIMPLEX_GRAD[(var25) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var25) | 1] * yr + FLAT_SIMPLEX_GRAD[(var25) | 2] * var62;
+            final double var93 = FLAT_SIMPLEX_GRAD[(var26) | 0] * xr + FLAT_SIMPLEX_GRAD[(var26) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var26) | 2] * var62;
+            final double var94 = FLAT_SIMPLEX_GRAD[(var27) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var27) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var27) | 2] * var62;
+
+            final double var95 = xr * 6.0 - 15.0;
+            final double var96 = yrOriginal * 6.0 - 15.0;
+            final double var97 = zr * 6.0 - 15.0;
+            final double var98 = xr * var95 + 10.0;
+            final double var99 = yrOriginal * var96 + 10.0;
+            final double var100 = zr * var97 + 10.0;
+            final double var101 = xr * xr * xr * var98;
+            final double var102 = yrOriginal * yrOriginal * yrOriginal * var99;
+            final double var103 = zr * zr * zr * var100;
+
+            final double var113 = var87 + var101 * (var88 - var87);
+            final double var114 = var93 + var101 * (var94 - var93);
+            final double var115 = var91 + var101 * (var92 - var91);
+            final double var116 = var89 + var101 * (var90 - var89);
+            final double var117 = var114 - var115;
+            final double var118 = var102 * (var116 - var113);
+            final double var119 = var102 * var117;
+            final double var120 = var113 + var118;
+            final double var121 = var115 + var119;
+            return var120 + (var103 * (var121 - var120));
+        }
+        // Leaves end - optimize noise generation
     }
 
     private double sampleWithDerivative(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double[] derivativeOut) {
diff --git a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
index dc632666201028e680443eb27f235583506349ad..49badaa7bf362dc0d3ef4933179e90a56bf8034a 100644
--- a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
@@ -26,6 +26,10 @@ public class PerlinNoise {
     private final double lowestFreqValueFactor;
     private final double lowestFreqInputFactor;
     private final double maxValue;
+    // Leaves start - optimize noise generation
+    private final int octaveSamplersCount;
+    private final double[] amplitudesArray;
+    // Leaves end - optimize noise generation
 
     @Deprecated
     public static PerlinNoise createLegacyForBlendedNoise(final RandomSource random, final IntStream octaves) {
@@ -127,6 +131,10 @@ public class PerlinNoise {
         this.lowestFreqInputFactor = Math.pow(2.0, -zeroOctaveIndex);
         this.lowestFreqValueFactor = Math.pow(2.0, octaves - 1) / (Math.pow(2.0, octaves) - 1.0);
         this.maxValue = this.edgeValue(2.0);
+        // Leaves start - optimize noise generation
+        this.octaveSamplersCount = this.noiseLevels.length;
+        this.amplitudesArray = this.amplitudes.toDoubleArray();
+        // Leaves end - optimize noise generation
     }
 
     protected double maxValue() {
@@ -138,7 +146,30 @@ public class PerlinNoise {
     }
 
     public double getValue(final double x, final double y, final double z) {
-        return this.getValue(x, y, z, 0.0, 0.0);
+        // Leaves start - optimize noise generation
+        if (!org.leavesmc.leaves.LeavesConfig.performance.optimizeNoiseGeneration) {
+            return this.getValue(x, y, z, 0.0, 0.0);
+        } else {
+            double d = 0.0;
+            double e = this.lowestFreqInputFactor;
+            double f = this.lowestFreqValueFactor;
+
+            for (int i = 0; i < this.octaveSamplersCount; ++i) {
+                ImprovedNoise perlinNoiseSampler = this.noiseLevels[i];
+                if (perlinNoiseSampler != null) {
+                    @SuppressWarnings("deprecation")
+                    double g = perlinNoiseSampler.noise(
+                        wrap(x * e), wrap(y * e), wrap(z * e), 0.0, 0.0
+                    );
+                    d += this.amplitudesArray[i] * g * f;
+                }
+
+                e *= 2.0;
+                f /= 2.0;
+            }
+            return d;
+        }
+        // Leaves end - optimize noise generation
     }
 
     @Deprecated
@@ -186,7 +217,13 @@ public class PerlinNoise {
     }
 
     public static double wrap(final double x) {
-        return x - Mth.lfloor(x / 3.3554432E7 + 0.5) * 3.3554432E7;
+        // Leaves start - optimize noise generation
+        if (!org.leavesmc.leaves.LeavesConfig.performance.optimizeNoiseGeneration) {
+            return x - Mth.lfloor(x / 3.3554432E7 + 0.5) * 3.3554432E7;
+        } else {
+            return x - Math.floor(x / 3.3554432E7 + 0.5) * 3.3554432E7;
+        }
+        // Leaves end - optimize noise generation
     }
 
     protected int firstOctave() {
