From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Mon, 29 Sep 2025 10:24:45 +0800
Subject: [PATCH] Do not check NaN values in entities


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f150ccb26cb191f4db68f1f4596eacd62343d117..9ea93411034d035af0d7ff1cb656be2dec040c67 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2588,6 +2588,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
             // CraftBukkit start - Checking for NaN pitch/yaw and resetting to zero
             // TODO: make sure this is the best way to address this.
+            /*
             if (Float.isNaN(this.yRot)) {
                 this.yRot = 0;
             }
@@ -2595,7 +2596,9 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             if (Float.isNaN(this.xRot)) {
                 this.xRot = 0;
             }
+            */
             // CraftBukkit end
+            // Leaves end - Do not check NaN values
             output.store("Rotation", Vec2.CODEC, new Vec2(this.getYRot(), this.getXRot()));
             output.putDouble("fall_distance", this.fallDistance);
             output.putShort("Fire", (short)this.remainingFireTicks);
