From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Mon, 29 Sep 2025 10:24:45 +0800
Subject: [PATCH] Allow Inf NaN values in entity motions


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f150ccb26cb191f4db68f1f4596eacd62343d117..c6d0043c99c24d4883f17815dc54b2b7d283c950 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2370,7 +2370,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void push(Vec3 vector) {
-        if (vector.isFinite()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues || vector.isFinite()) { // Leaves - allow Inf NaN motion values
             this.push(vector.x, vector.y, vector.z);
         }
     }
@@ -2382,7 +2382,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     public void push(double x, double y, double z, @Nullable Entity pushingEntity) {
         // Paper end - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
-        if (Double.isFinite(x) && Double.isFinite(y) && Double.isFinite(z)) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues || (Double.isFinite(x) && Double.isFinite(y) && Double.isFinite(z))) { // Leaves - allow Inf NaN motion values
             // Paper start - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
             org.bukkit.util.Vector delta = new org.bukkit.util.Vector(x, y, z);
             if (pushingEntity != null) {
@@ -2588,14 +2588,18 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
             // CraftBukkit start - Checking for NaN pitch/yaw and resetting to zero
             // TODO: make sure this is the best way to address this.
-            if (Float.isNaN(this.yRot)) {
-                this.yRot = 0;
-            }
+            // Leaves start - allow Inf NaN motion values
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues) {
+                if (Float.isNaN(this.yRot)) {
+                    this.yRot = 0;
+                }
 
-            if (Float.isNaN(this.xRot)) {
-                this.xRot = 0;
+                if (Float.isNaN(this.xRot)) {
+                    this.xRot = 0;
+                }
             }
             // CraftBukkit end
+            // Leaves end - allow Inf NaN motion values
             output.store("Rotation", Vec2.CODEC, new Vec2(this.getYRot(), this.getXRot()));
             output.putDouble("fall_distance", this.fallDistance);
             output.putShort("Fire", (short)this.remainingFireTicks);
@@ -4984,15 +4988,15 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void setDeltaMovement(Vec3 deltaMovement) {
-        if (deltaMovement.isFinite()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues || deltaMovement.isFinite()) { // Leaves - allow Inf NaN motion values
             synchronized (this.posLock) { // Paper - detailed watchdog information
-            this.deltaMovement = deltaMovement;
+                this.deltaMovement = deltaMovement;
             } // Paper - detailed watchdog information
         }
     }
 
     public void addDeltaMovement(Vec3 addend) {
-        if (addend.isFinite()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues || addend.isFinite()) { // Leaves - allow Inf NaN motion values
             this.setDeltaMovement(this.getDeltaMovement().add(addend));
         }
     }
@@ -5213,7 +5217,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void setYRot(float yRot) {
-        if (!Float.isFinite(yRot)) {
+        if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues && !Float.isFinite(yRot)) { // Leaves - allow Inf NaN motion values
             Util.logAndPauseIfInIde("Invalid entity rotation: " + yRot + ", discarding.");
         } else {
             this.yRot = yRot;
@@ -5225,7 +5229,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void setXRot(float xRot) {
-        if (!Float.isFinite(xRot)) {
+        if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowInfNanMotionValues && !Float.isFinite(xRot)) { // Leaves - allow Inf NaN motion values
             Util.logAndPauseIfInIde("Invalid entity rotation: " + xRot + ", discarding.");
         } else {
             this.xRot = Math.clamp(xRot % 360.0F, -90.0F, 90.0F);
