From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Leaves <leaves@leaves>
Date: Thu, 16 Apr 2026 13:41:30 +0800
Subject: [PATCH] Leaves Fakeplayer


diff --git a/net/minecraft/advancements/criterion/SimpleCriterionTrigger.java b/net/minecraft/advancements/criterion/SimpleCriterionTrigger.java
index dfce5ced082de0caaff8da071fb632cd6ae787c9..d5c1862ebf6066859da74f47fa0be0256ba0f653 100644
--- a/net/minecraft/advancements/criterion/SimpleCriterionTrigger.java
+++ b/net/minecraft/advancements/criterion/SimpleCriterionTrigger.java
@@ -41,6 +41,7 @@ public abstract class SimpleCriterionTrigger<T extends SimpleCriterionTrigger.Si
     }
 
     protected void trigger(final ServerPlayer player, final Predicate<T> matcher) {
+        if (player instanceof org.leavesmc.leaves.bot.ServerBot) return; // Leaves - bot skip
         PlayerAdvancements advancements = player.getAdvancements();
         Set<CriterionTrigger.Listener<T>> allListeners = (Set) advancements.criterionData.get(this); // Paper - fix PlayerAdvancements leak
         if (allListeners != null && !allListeners.isEmpty()) {
diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 4e0d4b18f693fdffd2e431f836776754e11d07c2..8f5e5ad5a97956f59b45f86233c43b3a86ebc3af 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -74,7 +74,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public boolean preparing = true;
     // Spigot end
     private volatile @Nullable PacketListener disconnectListener;
-    private volatile @Nullable PacketListener packetListener;
+    protected volatile @Nullable PacketListener packetListener; // Leaves - private -> protected
     private @Nullable DisconnectionDetails disconnectionDetails;
     private boolean encrypted;
     private boolean disconnectionHandled;
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 4af7cd900ebe6bac0fbbda98154cd4feeec8cb87..f43053662c1ab215ae03789e944865e85662ae4c 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -298,6 +298,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // Paper - per-level scheduledEvents
     private final ServerClockManager clockManager;
 
+    private org.leavesmc.leaves.bot.BotList botList; // Leaves - fakeplayer
+
     public static <S extends MinecraftServer> S spin(final Function<Thread, S> factory) {
         ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.init(); // Paper - rewrite data converter system
         AtomicReference<S> serverReference = new AtomicReference<>();
@@ -1046,6 +1048,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
+        this.getBotList().removeAll(); // Leaves - save or remove bot
         // CraftBukkit start
         if (this.server != null) {
             this.server.spark.disable(); // Paper - spark
@@ -1606,7 +1609,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         int emptyTickThreshold = this.pauseWhenEmptySeconds() * 20;
         this.removeDisabledPluginsBlockingSleep(); // Paper - API to allow/disallow tick sleeping
         if (emptyTickThreshold > 0) {
-            if (this.playerList.getPlayerCount() == 0 && !this.tickRateManager.isSprinting() && this.pluginsBlockingSleep.isEmpty()) { // Paper - API to allow/disallow tick sleeping
+            if (this.playerList.getPlayerCount() == 0 && this.botList.bots.isEmpty() && !this.tickRateManager.isSprinting() && this.pluginsBlockingSleep.isEmpty()) { // Paper - API to allow/disallow tick sleeping // Leaves - fakeplayer
                 this.emptyTicks++;
             } else {
                 this.emptyTicks = 0;
@@ -1659,6 +1662,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.isSaving = true;
             if (playerSaveInterval > 0) {
                 this.playerList.saveAll(playerSaveInterval);
+                org.leavesmc.leaves.bot.BotList.INSTANCE.saveAllResumeBots(playerSaveInterval); // Leaves - resident fakeplayer
             }
             if (fullSave) {
                 this.saveGlobalData(false);
@@ -1905,6 +1909,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     protected void tickConnection() {
         this.getConnection().tick();
+        this.botList.networkTick(); // Leaves - fakeplayer
     }
 
     public void forceGameTimeSynchronization() {
@@ -3082,6 +3087,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return this.debugSubscribers;
     }
 
+    // Leaves start - fakeplayer
+    protected void setBotList(org.leavesmc.leaves.bot.BotList botList) {
+        this.botList = botList;
+    }
+
+    public org.leavesmc.leaves.bot.BotList getBotList() {
+        return botList;
+    }
+    // Leaves end - fakeplayer
+
     public record ReloadableResources(CloseableResourceManager resourceManager, ReloadableServerResources managers) implements AutoCloseable {
         @Override
         public void close() {
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index c5f99ca1167fd912c7fb46c9835f4e4766c152aa..2320b796887ac38150d70c4470841ad841f42a2e 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -168,6 +168,11 @@ public class PlayerAdvancements {
     }
 
     public boolean award(final AdvancementHolder holder, final String criterion) {
+        // Leaves start - bot can't get advancement
+        if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+            return false;
+        }
+        // Leaves end - bot can't get advancement
         // Leaves start - spectator don't get advancement
         if (org.leavesmc.leaves.LeavesConfig.modify.spectatorDontGetAdvancement && player.gameMode.getGameModeForPlayer() == net.minecraft.world.level.GameType.SPECTATOR) {
             return false;
diff --git a/net/minecraft/server/commands/BanIpCommands.java b/net/minecraft/server/commands/BanIpCommands.java
index 36b858bfe123edc2f1c5c7861dd0e4475ef7a5f1..e52a2ad8408d9691701dc51fa702efdb51f0acbb 100644
--- a/net/minecraft/server/commands/BanIpCommands.java
+++ b/net/minecraft/server/commands/BanIpCommands.java
@@ -40,6 +40,12 @@ public class BanIpCommands {
             return banIp(source, target, reason);
         } else {
             ServerPlayer player = source.getServer().getPlayerList().getPlayerByName(target);
+            // Leaves start - disable ban
+            if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+                source.sendFailure(Component.literal("Permission denied"));
+                return 0;
+            }
+            // Leaves end - disable ban
             if (player != null) {
                 return banIp(source, player.getIpAddress(), reason);
             } else {
diff --git a/net/minecraft/server/commands/BanPlayerCommands.java b/net/minecraft/server/commands/BanPlayerCommands.java
index c87bc86c57fde9d92e3927bdead5cdaf2e935744..950f8ed955bc09713348936ae7aa1a8b14006c2a 100644
--- a/net/minecraft/server/commands/BanPlayerCommands.java
+++ b/net/minecraft/server/commands/BanPlayerCommands.java
@@ -38,8 +38,15 @@ public class BanPlayerCommands {
     private static int banPlayers(final CommandSourceStack source, final Collection<NameAndId> players, final @Nullable Component reason) throws CommandSyntaxException {
         UserBanList list = source.getServer().getPlayerList().getBans();
         int count = 0;
-
+        boolean hasBot = false; // Leaves - disable kick
         for (NameAndId player : players) {
+            // Leaves start - disable ban
+            if (player.isBot()) {
+                source.sendFailure(Component.literal("Permission denied"));
+                hasBot = true;
+                continue;
+            }
+            // Leaves end - disable ban
             if (!list.isBanned(player)) {
                 UserBanListEntry entry = new UserBanListEntry(player, null, source.getTextName(), null, reason == null ? null : reason.getString());
                 list.add(entry);
@@ -53,7 +60,13 @@ public class BanPlayerCommands {
         }
 
         if (count == 0) {
-            throw ERROR_ALREADY_BANNED.create();
+            // Leaves start - disable kick
+            if (hasBot) {
+                return count;
+            } else {
+                throw ERROR_ALREADY_BANNED.create();
+            }
+            // Leaves end - disable kick
         } else {
             return count;
         }
diff --git a/net/minecraft/server/commands/KickCommand.java b/net/minecraft/server/commands/KickCommand.java
index 0227dd63f67537d769494e122616873c18d35350..18d49018b1330cfcaa91d9b02d0d23e4ba941809 100644
--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -38,9 +38,17 @@ public class KickCommand {
         if (!source.getServer().isPublished()) {
             throw ERROR_SINGLEPLAYER.create();
         } else {
+            boolean hasBot = false; // Leaves - disable kick
             int count = 0;
 
             for (ServerPlayer player : players) {
+                // Leaves start - disable kick
+                if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+                    source.sendFailure(Component.literal("Permission denied"));
+                    hasBot = true;
+                    continue;
+                }
+                // Leaves end - disable kick
                 if (!source.getServer().isSingleplayerOwner(player.nameAndId())) {
                     player.connection.disconnect(reason, org.bukkit.event.player.PlayerKickEvent.Cause.KICK_COMMAND); // Paper - kick event cause
                     source.sendSuccess(() -> Component.translatable("commands.kick.success", player.getDisplayName(), reason), true);
@@ -49,7 +57,13 @@ public class KickCommand {
             }
 
             if (count == 0) {
-                throw ERROR_KICKING_OWNER.create();
+                // Leaves start - disable kick
+                if (hasBot) {
+                    return count;
+                } else {
+                    throw ERROR_KICKING_OWNER.create();
+                }
+                // Leaves end - disable kick
             } else {
                 return count;
             }
diff --git a/net/minecraft/server/commands/OpCommand.java b/net/minecraft/server/commands/OpCommand.java
index 636f56bfc0e9075b257e2eb14e858975bdedd801..20cbc0e65fea3fc2d0033259c2f35e158ea81926 100644
--- a/net/minecraft/server/commands/OpCommand.java
+++ b/net/minecraft/server/commands/OpCommand.java
@@ -39,6 +39,7 @@ public class OpCommand {
         int count = 0;
 
         for (NameAndId player : players) {
+            if (player.isBot()) continue; // Leaves - disable op
             if (!list.isOp(player)) {
                 list.op(player);
                 count++;
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 2c3aa2c89b6bb990f0e55ff7bccc7fb9a8116bed..de11777a37c915cd1b1b2103b61f6535cf71b7ed 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -274,6 +274,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
 
         // Spigot start
+        this.setBotList(new org.leavesmc.leaves.bot.BotList(this)); // Leaves - fakeplayer
         this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
         org.spigotmc.SpigotConfig.init((java.io.File) this.options.valueOf("spigot-settings"));
         org.spigotmc.SpigotConfig.registerCommands();
@@ -295,6 +296,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.server.spark.registerCommandBeforePlugins(this.server); // Paper - spark
 
         org.leavesmc.leaves.LeavesConfig.init((java.io.File) options.valueOf("leaves-settings")); // Leaves - Server Config
+        this.getBotList().loadResumeBotInfo(); // Leaves - load resident bot info
         org.leavesmc.leaves.util.ServerI18nUtil.init(); // Leaves I18n
 
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 18808f2bf537930802b171920c73413b42d9c5d7..95709714f9e461dde795c2db77bf6fded850dff2 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1402,6 +1402,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                     }
                 } else {
                     this.removePlayer(player);
+                    // Leaves start - render bot
+                    if (this.entity instanceof org.leavesmc.leaves.bot.ServerBot bot) {
+                        if (bot.needSendFakeData(player)) {
+                            bot.sendFakeData(player.connection, false);
+                        }
+                    }
+                    // Leaves end - render bot
                 }
             }
         }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index cd766676ff70e82de507bec2022d2dac41fd35a6..535ae2094476d1d3858674facb37d4433c02f5ec 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -233,6 +233,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public boolean hasPhysicsEvent = true; // Paper - BlockPhysicsEvent
     public boolean hasEntityMoveEvent; // Paper - Add EntityMoveEvent
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
+    final List<ServerPlayer> realPlayers; // Leaves - skip
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -731,6 +732,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        this.realPlayers = Lists.newArrayList(); // Leaves - skip
     }
 
     // Paper start
@@ -2404,6 +2406,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return this.players;
     }
 
+    // Leaves start - fakeplayer skip
+    public List<ServerPlayer> realPlayers() {
+        return this.realPlayers;
+    }
+    // Leaves end - fakeplayer skip
+
     @Override
     public void updatePOIOnBlockStateChange(final BlockPos pos, final BlockState oldState, final BlockState newState) {
         Optional<Holder<PoiType>> oldType = PoiTypes.forState(oldState);
@@ -2883,6 +2891,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.add(player);
+                // Leaves start - skip
+                if (!(player instanceof org.leavesmc.leaves.bot.ServerBot)) {
+                    ServerLevel.this.realPlayers.add(player);
+                }
+                // Leaves end - skip
                 if (player.isReceivingWaypoints()) {
                     ServerLevel.this.getWaypointManager().addPlayer(player);
                 }
@@ -2961,6 +2974,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             ServerLevel.this.getChunkSource().removeEntity(entity);
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.remove(player);
+                // Leaves start - skip
+                if (!(player instanceof org.leavesmc.leaves.bot.ServerBot)) {
+                    ServerLevel.this.realPlayers.remove(player);
+                }
+                // Leaves end - skip
                 ServerLevel.this.getWaypointManager().removePlayer(player);
                 ServerLevel.this.updateSleepingPlayerList();
             }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index bf056cc8b531592a2ebed1adba957419626c1fba..9f88d73221832b6ba5071f5f85ff7a472486d912 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -229,7 +229,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private static final boolean DEFAULT_SPAWN_EXTRA_PARTICLES_ON_FALL = false;
     public ServerGamePacketListenerImpl connection;
     private final MinecraftServer server;
-    public final ServerPlayerGameMode gameMode;
+    public ServerPlayerGameMode gameMode; // Leaves - not final
     private final PlayerAdvancements advancements;
     private final ServerStatsCounter stats;
     private float lastRecordedHealthAndAbsorption = Float.MIN_VALUE;
@@ -250,7 +250,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private @Nullable Entity camera;
     public boolean isChangingDimension;
     public boolean seenCredits = false;
-    private final ServerRecipeBook recipeBook;
+    protected ServerRecipeBook recipeBook; // Leaves - not final and private -> protected
     private @Nullable Vec3 levitationStartPos;
     private int levitationStartTime;
     private boolean disconnected;
@@ -1289,7 +1289,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     // Leaves end - exp fix
 
-    private void tellNeutralMobsThatIDied() {
+    protected void tellNeutralMobsThatIDied() { // Leaves private -> protected
         AABB aabb = new AABB(this.blockPosition()).inflate(32.0, 10.0, 32.0);
         this.level()
             .getEntitiesOfClass(Mob.class, aabb, EntitySelector.NO_SPECTATORS)
@@ -1615,6 +1615,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 this.lastSentHealth = -1.0F;
                 this.lastSentFood = -1;
                 this.teleportSpectators(transition, oldLevel);
+                // Leaves start - bot support
+                if (org.leavesmc.leaves.LeavesConfig.modify.fakeplayer.enable) {
+                    this.server.getBotList().bots.forEach(bot -> bot.sendFakeDataIfNeed(this, true)); // Leaves - render bot
+                }
+                // Leaves end - bot support
                 // CraftBukkit start
                 org.bukkit.event.player.PlayerChangedWorldEvent changeEvent = new org.bukkit.event.player.PlayerChangedWorldEvent(this.getBukkitEntity(), oldLevel.getWorld());
                 this.level().getCraftServer().getPluginManager().callEvent(changeEvent);
diff --git a/net/minecraft/server/players/CachedUserNameToIdResolver.java b/net/minecraft/server/players/CachedUserNameToIdResolver.java
index aecbdddf4cf3ab627477156cd314f2872e7a5dbb..60c9c520247effa5512d89cb0d7c63633cbfef2f 100644
--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -122,6 +122,12 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
 
     @Override
     public Optional<NameAndId> get(final String name) {
+        // Leaves start - fix bot
+        org.leavesmc.leaves.bot.ServerBot bot = org.leavesmc.leaves.bot.BotList.INSTANCE.getBotByName(name);
+        if (bot != null) {
+            return Optional.of(bot.nameAndId());
+        }
+        // Leaves end - fix bot
         String userName = name.toLowerCase(Locale.ROOT);
         boolean stateLocked = true; try { this.stateLock.lock(); // Paper - Fix GameProfileCache concurrency
         CachedUserNameToIdResolver.GameProfileInfo profileInfo = this.profilesByName.get(userName);
diff --git a/net/minecraft/server/players/NameAndId.java b/net/minecraft/server/players/NameAndId.java
index 484714beb5f7cdf8178834c5edd4f847fb06ebb8..ba7a280aad07f36366c15cd158b36ca9768cef19 100644
--- a/net/minecraft/server/players/NameAndId.java
+++ b/net/minecraft/server/players/NameAndId.java
@@ -8,18 +8,22 @@ import java.util.UUID;
 import net.minecraft.core.UUIDUtil;
 import org.jspecify.annotations.Nullable;
 
-public record NameAndId(UUID id, String name) {
+public record NameAndId(UUID id, String name, boolean isBot) { // Leaves - fakeplayer
     public static final Codec<NameAndId> CODEC = RecordCodecBuilder.create(
         i -> i.group(UUIDUtil.STRING_CODEC.fieldOf("id").forGetter(NameAndId::id), Codec.STRING.fieldOf("name").forGetter(NameAndId::name))
             .apply(i, NameAndId::new)
     );
 
     public NameAndId(final GameProfile profile) {
-        this(profile.id(), profile.name());
+        this(profile.id(), profile.name(), com.google.common.collect.Iterables.getFirst(profile.properties().get("is_bot"), "false").equals("true")); // Leaves - fakeplayer
     }
 
     public NameAndId(final com.mojang.authlib.yggdrasil.response.NameAndId profile) {
-        this(profile.id(), profile.name());
+        this(profile.id(), profile.name(), false); // Leaves - fakeplayer
+    }
+
+    public NameAndId(UUID uuid, String name) {
+        this(uuid, name, false); // Leaves - fakeplayer
     }
 
     public static @Nullable NameAndId fromJson(final JsonObject object) {
@@ -33,7 +37,7 @@ public record NameAndId(UUID id, String name) {
                 return null;
             }
 
-            return new NameAndId(uuid, object.get("name").getAsString());
+            return new NameAndId(uuid, object.get("name").getAsString(), false); // Leaves - fakeplayer
         } else {
             return null;
         }
@@ -46,7 +50,7 @@ public record NameAndId(UUID id, String name) {
 
     public static NameAndId createOffline(final String name) {
         UUID id = UUIDUtil.createOfflinePlayerUUID(name);
-        return new NameAndId(id, name);
+        return new NameAndId(id, name, false); // Leaves - fakeplayer
     }
 
     // Paper start - utility method for common conversion back to the game profile
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index e2c47f4e0e66d8d2a832d925a586956df11664f9..7cbcecb5f761d1a728a2bde003ffe6d451ba0cbf 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -255,6 +255,19 @@ public abstract class PlayerList {
 
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handlePlayerJoin(player); // Leaves - protocol
 
+        // Leaves start - bot support
+        if (org.leavesmc.leaves.LeavesConfig.modify.fakeplayer.enable) {
+            org.leavesmc.leaves.bot.ServerBot bot = this.server.getBotList().getBotByName(player.getScoreboardName());
+            if (bot != null) {
+                this.server.getBotList().removeBot(bot, org.leavesmc.leaves.event.bot.BotRemoveEvent.RemoveReason.INTERNAL, player.getBukkitEntity(), false, false);
+            }
+            this.server.getBotList().bots.forEach(bot1 -> {
+                bot1.sendPlayerInfo(player);
+                bot1.sendFakeDataIfNeed(player, true);
+            }); // Leaves - render bot
+        }
+        // Leaves end - bot support
+
         final net.kyori.adventure.text.Component jm = playerJoinEvent.joinMessage();
 
         if (jm != null && !jm.equals(net.kyori.adventure.text.Component.empty())) { // Paper - Adventure
@@ -727,6 +740,11 @@ public abstract class PlayerList {
             respawnReason
         ).callEvent();
         // Paper end
+        // Leaves start - bot support
+        if (org.leavesmc.leaves.LeavesConfig.modify.fakeplayer.enable) {
+            this.server.getBotList().bots.forEach(bot -> bot.sendFakeDataIfNeed(player, true)); // Leaves - render bot
+        }
+        // Leaves end - bot support
 
         return player;
     }
@@ -829,11 +847,16 @@ public abstract class PlayerList {
     }
 
     public String[] getPlayerNamesArray() {
-        String[] names = new String[this.players.size()];
+        String[] names = new String[this.players.size() + this.server.getBotList().bots.size()]; // Leaves - fakeplayer support
 
         for (int i = 0; i < this.players.size(); i++) {
             names[i] = this.players.get(i).getGameProfile().name();
         }
+        // Leaves start - fakeplayer support
+        for (int i = this.players.size(); i < names.length; ++i) {
+            names[i] = this.server.getBotList().bots.get(i - this.players.size()).getGameProfile().getName();
+        }
+        // Leaves end - fakeplayer support
 
         return names;
     }
@@ -942,8 +965,15 @@ public abstract class PlayerList {
             || this.allowCommandsForAllPlayers;
     }
 
-    public @Nullable ServerPlayer getPlayerByName(final String name) {
-        return this.playersByName.get(name.toLowerCase(java.util.Locale.ROOT)); // Spigot
+    public @Nullable ServerPlayer getPlayerByName(String name) {
+        // Leaves start - fakeplayer support
+        name = name.toLowerCase(java.util.Locale.ROOT);
+        ServerPlayer player = this.playersByName.get(name);
+        if (player == null) {
+            player = this.server.getBotList().getBotByName(name);
+        }
+        return player; // Spigot
+        // Leaves end - fakeplayer support
     }
 
     public void broadcast(
@@ -1260,7 +1290,13 @@ public abstract class PlayerList {
     }
 
     public @Nullable ServerPlayer getPlayer(final UUID uuid) {
-        return this.playersByUUID.get(uuid);
+        // Leaves start - fakeplayer support
+        ServerPlayer player = this.playersByUUID.get(uuid);
+        if (player == null) {
+            player = this.server.getBotList().getBot(uuid);
+        }
+        return player;
+        // Leaves start - fakeplayer support
     }
 
     public @Nullable ServerPlayer getPlayer(final String playerName) {
diff --git a/net/minecraft/server/waypoints/ServerWaypointManager.java b/net/minecraft/server/waypoints/ServerWaypointManager.java
index 434f2870d64b1f5ba85338cbd9aff62feaea09f7..3a0d2e88bc10fcf35d288df667e8f5f083a87fb3 100644
--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -29,6 +29,11 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
 
     @Override
     public void trackWaypoint(final WaypointTransmitter waypoint) {
+        // Leaves start - fakeplayer
+        if (waypoint instanceof org.leavesmc.leaves.bot.ServerBot bot && !bot.getConfigValue(org.leavesmc.leaves.bot.agent.Configs.ENABLE_LOCATOR_BAR)) {
+            return;
+        }
+        // Leaves end - fakeplayer
         this.waypoints.add(waypoint);
 
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
@@ -62,6 +67,11 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     public void addPlayer(final ServerPlayer player) {
+        // Leaves start - fakeplayer
+        if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+            return;
+        }
+        // Leaves end - fakeplayer
         this.players.add(player);
 
         if (this.locatorBarEnabled) { // Paper - optimize ServerWaypointManager with locator bar disabled
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 14fc232d45d3f8a3ceb001dc5a4357f848d4f69e..82ec42bb0ec2f04518a7c3f71543288793256f5d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1233,7 +1233,7 @@ public abstract class Entity
 
             BlockPos effectPos = this.getOnPosLegacy();
             BlockState effectState = this.level().getBlockState(effectPos);
-            if (this.isLocalInstanceAuthoritative()) {
+            if (this.isLocalInstanceAuthoritative() || this instanceof org.leavesmc.leaves.bot.ServerBot) {  // Leaves - ServerBot needs check fall damage
                 this.checkFallDamage(movement.y, this.onGround(), effectState, effectPos);
             }
 
@@ -1588,7 +1588,7 @@ public abstract class Entity
         return colliders.isEmpty() ? maxDistance : -Shapes.collide(Direction.Axis.Y, aabb, colliders, -maxDistance);
     }
 
-    private Vec3 collide(final Vec3 movement) {
+    public Vec3 collide(final Vec3 movement) { // Leaves - private -> public
         // Paper start - optimise collisions
         final boolean xZero = movement.x == 0.0;
         final boolean yZero = movement.y == 0.0;
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index df38eb2678a77682b24afd0fd363104ff6fb822a..af8419ca97ef9f4df349e48590f38cb1104530b6 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3287,7 +3287,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     private void travelRidden(final Player controller, final Vec3 selfInput) {
         Vec3 riddenInput = this.getRiddenInput(controller, selfInput);
         this.tickRidden(controller, riddenInput);
-        if (this.canSimulateMovement()) {
+        if (this.canSimulateMovement() || this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot) { // Leaves - Fakeplayer
             this.setSpeed(this.getRiddenSpeed(controller));
             this.travel(riddenInput);
         } else {
@@ -4210,7 +4210,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Paper start - lag compensate eating
         // we add 1 to the expected time to avoid lag compensating when we should not
         final boolean shouldLagCompensate = this.useItem.has(DataComponents.FOOD) && this.eatStartTime != -1 && (System.nanoTime() - this.eatStartTime) > ((1L + this.totalEatTimeTicks) * 50L * (1000L * 1000L));
-        if ((--this.useItemRemaining == 0 || shouldLagCompensate) && !this.level().isClientSide() && !useItem.useOnRelease()) {
+        if ((--this.useItemRemaining == 0 || shouldLagCompensate) && !(this instanceof org.leavesmc.leaves.bot.ServerBot) && !this.level().isClientSide() && !useItem.useOnRelease()) {
             this.useItemRemaining = 0;
             // Paper end - lag compensate eating
             this.completeUsingItem();
@@ -4399,6 +4399,23 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         this.stopUsingItem();
     }
 
+    // Leaves start - Fakeplayer
+    public boolean releaseUsingItemWithResult() {
+        ItemStack itemInHand = this.getItemInHand(this.getUsedItemHand());
+        boolean result = false;
+        if (!this.useItem.isEmpty() && ItemStack.isSameItem(itemInHand, this.useItem)) {
+            this.useItem = itemInHand;
+            result = this.useItem.releaseUsingWithResult(this.level(), this, this.getUseItemRemainingTicks());
+            if (this.useItem.useOnRelease()) {
+                this.updatingUsingItem();
+            }
+        }
+
+        this.stopUsingItem();
+        return result;
+    }
+    // Leaves end - Fakeplayer
+
     public void stopUsingItem() {
         if (!this.level().isClientSide()) {
             boolean wasUsingItem = this.isUsingItem();
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index e6ac9d48b2e5cd87017cbd154a8619cb61b6ed2f..bcd9ddc9663dabf56ee20d85ff350e173b9063fe 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -167,7 +167,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     private int lastLevelUpTime;
     public GameProfile gameProfile;
     private boolean reducedDebugInfo;
-    private ItemStack lastItemInMainHand = ItemStack.EMPTY;
+    protected ItemStack lastItemInMainHand = ItemStack.EMPTY; // Leaves - fakeplayer
     private final ItemCooldowns cooldowns = this.createItemCooldowns();
     private Optional<GlobalPos> lastDeathLocation = Optional.empty();
     public @Nullable FishingHook fishing;
@@ -301,6 +301,12 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.updatePlayerPose();
     }
 
+    // Leaves start - fakeplayer
+    protected void livingEntityTick() {
+        super.tick();
+    }
+    // Leaves end - fakeplayer
+
     @Override
     protected float getMaxHeadRotationRelativeToBody() {
         return this.isBlocking() ? 15.0F : super.getMaxHeadRotationRelativeToBody();
@@ -526,7 +532,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     public void removeEntitiesOnShoulder() {
     }
 
-    private void touch(final Entity entity) {
+    public void touch(final Entity entity) { // Leaves - private -> public
         entity.playerTouch(this);
     }
 
@@ -1208,7 +1214,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             // Paper end - Configurable sprint interruption on attack
         }
 
-        if (entity instanceof ServerPlayer && entity.hurtMarked) {
+        if ((entity instanceof ServerPlayer && !(entity instanceof org.leavesmc.leaves.bot.ServerBot)) && entity.hurtMarked) { // Leaves - bot knockback
             // CraftBukkit start - Add Velocity Event
             boolean cancelled = false;
             org.bukkit.entity.Player player = (org.bukkit.entity.Player) entity.getBukkitEntity();
diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index 5f8711b3d6ffcbd9ecfedec90ff09e6e0ddfa228..d078df637e4ffc484d70b4038fbc4df6f7332570 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -58,7 +58,7 @@ public class FishingHook extends Projectile {
     public static final EntityDataAccessor<Integer> DATA_HOOKED_ENTITY = SynchedEntityData.defineId(FishingHook.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<Boolean> DATA_BITING = SynchedEntityData.defineId(FishingHook.class, EntityDataSerializers.BOOLEAN);
     private int life;
-    private int nibble;
+    public int nibble; // Leaves - private -> public
     public int timeUntilLured;
     public int timeUntilHooked;
     public float fishAngle;
diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index de254df4f13dd1544481fc8a3762ced649930da8..dc6581cbb9fff0613570608b719283165ee2f4e6 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -270,6 +270,11 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
             }
 
             this.move(MoverType.SELF, this.getDeltaMovement());
+        } else if (this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot) { // Leaves start - Fakeplayer
+            this.floatBoat();
+            this.controlBoat();
+            this.move(MoverType.SELF, this.getDeltaMovement());
+            // Leaves end - Fakeplayer
         } else {
             this.setDeltaMovement(Vec3.ZERO);
         }
@@ -378,6 +383,13 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
         }
     }
 
+    // Leaves start - Fakeplayer
+    @Override
+    public boolean canSimulateMovement() {
+        return super.canSimulateMovement() || this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot;
+    }
+    // Leaves end - Fakeplayer
+
     protected @Nullable SoundEvent getPaddleSound() {
         return switch (this.getStatus()) {
             case IN_WATER, UNDER_WATER, UNDER_FLOWING_WATER -> SoundEvents.BOAT_PADDLE_WATER;
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index d2439de25b83db20f22acba4a9170551e758326a..9d8f56f4cfd83a6c0f319af0358eb576fabf5169 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -407,6 +407,7 @@ public abstract class AbstractContainerMenu {
 
     private void doClick(final int slotIndex, final int buttonNum, final ContainerInput containerInput, final Player player) {
         Inventory inventory = player.getInventory();
+        if (!doClickCheck(slotIndex, buttonNum, containerInput, player)) return; // Leaves - doClick check
         if (containerInput == ContainerInput.QUICK_CRAFT) {
             int expectedStatus = this.quickcraftStatus;
             this.quickcraftStatus = getQuickcraftHeader(buttonNum);
@@ -681,6 +682,22 @@ public abstract class AbstractContainerMenu {
         }
     }
 
+    // Leaves start - doClick check
+    private boolean doClickCheck(int slotIndex, int buttonNum, ContainerInput containerInput, Player player) {
+        if (slotIndex < 0) {
+            return true;
+        }
+
+        Slot slot = getSlot(slotIndex);
+        ItemStack itemStack = slot.getItem();
+        net.minecraft.world.item.component.CustomData customData = itemStack.get(net.minecraft.core.component.DataComponents.CUSTOM_DATA);
+        if (customData != null && customData.contains("Leaves.Gui.Placeholder")) {
+            return !customData.copyTag().getBoolean("Leaves.Gui.Placeholder").orElse(false);
+        }
+        return true;
+    }
+    // Leaves end - doClick check
+
     private boolean tryItemClickBehaviourOverride(
         final Player player, final ClickAction clickAction, final Slot slot, final ItemStack clicked, final ItemStack carried
     ) {
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index cd7b18263adcdb8ac25d5f5adcaa8cec1ce7af14..21b3da5b5fb194434a3f5fb00e8c96ed54adc5ee 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -458,7 +458,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
                     placeEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceEvent(level, player, hand, blocks.getFirst(), pos);
                 }
 
-                if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
+                if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild()) && (!(player instanceof org.leavesmc.leaves.bot.ServerBot))) { // Leaves - Fakeplayer skip this check
                     result = InteractionResult.FAIL; // cancel placement
                     // PAIL: Remove this when MC-99075 fixed
                     player.containerMenu.forceHeldSlot(hand);
@@ -988,6 +988,20 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
         }
     }
 
+    // Leaves start - Fakeplayer
+    public boolean releaseUsingWithResult(Level level, LivingEntity livingEntity, int timeLeft) {
+        ItemStack itemStack = this.copy();
+        if (this.getItem().releaseUsing(this, level, livingEntity, timeLeft)) {
+            ItemStack itemStack1 = this.applyAfterUseComponentSideEffects(livingEntity, itemStack);
+            if (itemStack1 != this) {
+                livingEntity.setItemInHand(livingEntity.getUsedItemHand(), itemStack1);
+            }
+            return true;
+        }
+        return false;
+    }
+    // Leaves end - Fakeplayer
+
     public boolean useOnRelease() {
         return this.getItem().useOnRelease(this);
     }
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index d3978f67328c4e58ac896ea83596fa915ef9ffbd..049a8b31c78f76caa5e978f34c72794d6b11fa3a 100644
--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -142,7 +142,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
                                 break;
                             }
 
-                            if (!(entity instanceof ServerPlayer)) {
+                            if (!(entity instanceof ServerPlayer) || (entity instanceof org.leavesmc.leaves.bot.ServerBot)) { // Leaves - bot slime block
                                 Vec3 deltaMovement = entity.getDeltaMovement();
                                 double dx = deltaMovement.x;
                                 double dy = deltaMovement.y;
diff --git a/net/minecraft/world/level/levelgen/PhantomSpawner.java b/net/minecraft/world/level/levelgen/PhantomSpawner.java
index 1d691dd4c94ddb47f22fd8f9db3df531a10c0919..7c3f6a1889da66d71ad8f3db883cf836bb3fcdf7 100644
--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -48,6 +48,11 @@ public class PhantomSpawner implements CustomSpawner {
                                         ServerStatsCounter stats = player.getStats();
                                         int value = Mth.clamp(stats.getValue(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE);
                                         int dayLength = 24000;
+                                        // Leaves start - fakeplayer spawn
+                                        if (player instanceof org.leavesmc.leaves.bot.ServerBot bot && bot.getConfigValue(org.leavesmc.leaves.bot.agent.Configs.SPAWN_PHANTOM)) {
+                                            dayLength = Math.max(bot.notSleepTicks, 1);
+                                        }
+                                        // Leaves end - fakeplayer spawn
                                         if (random.nextInt(value) >= 72000) {
                                             BlockPos spawnPos = playerPos.above(20 + random.nextInt(15))
                                                 .east(-10 + random.nextInt(21))
