From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Mon, 29 Sep 2025 10:24:45 +0800
Subject: [PATCH] Allow Inf NaN values in entity motions


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f150ccb26cb191f4db68f1f4596eacd62343d117..df0e1c54163391a76cbf88100d547b533b21b7a9 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2588,14 +2588,18 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
             // CraftBukkit start - Checking for NaN pitch/yaw and resetting to zero
             // TODO: make sure this is the best way to address this.
-            if (Float.isNaN(this.yRot)) {
-                this.yRot = 0;
-            }
+            // Leaves start - allow Inf NaN motion values
+            if (!org.leavesmc.leaves.LeavesConfig.modify.allowInfNanMotionValues) {
+                if (Float.isNaN(this.yRot)) {
+                    this.yRot = 0;
+                }
 
-            if (Float.isNaN(this.xRot)) {
-                this.xRot = 0;
+                if (Float.isNaN(this.xRot)) {
+                    this.xRot = 0;
+                }
             }
             // CraftBukkit end
+            // Leaves end - allow Inf NaN motion values
             output.store("Rotation", Vec2.CODEC, new Vec2(this.getYRot(), this.getXRot()));
             output.putDouble("fall_distance", this.fallDistance);
             output.putShort("Fire", (short)this.remainingFireTicks);
@@ -4984,9 +4988,9 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void setDeltaMovement(Vec3 deltaMovement) {
-        if (deltaMovement.isFinite()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.allowInfNanMotionValues || deltaMovement.isFinite()) { // Leaves - allow Inf NaN motion values
             synchronized (this.posLock) { // Paper - detailed watchdog information
-            this.deltaMovement = deltaMovement;
+                this.deltaMovement = deltaMovement;
             } // Paper - detailed watchdog information
         }
     }
