From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <122684177+HyacinthHaru@users.noreply.github.com>
Date: Thu, 16 Apr 2026 13:45:41 +0800
Subject: [PATCH] Fix Fakeplayer getGameProfile API


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 7cbcecb5f761d1a728a2bde003ffe6d451ba0cbf..8b83f4aa5712290cbf19291c0bc2740518016180 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -854,7 +854,7 @@ public abstract class PlayerList {
         }
         // Leaves start - fakeplayer support
         for (int i = this.players.size(); i < names.length; ++i) {
-            names[i] = this.server.getBotList().bots.get(i - this.players.size()).getGameProfile().getName();
+            names[i] = this.server.getBotList().bots.get(i - this.players.size()).gameProfile.name();
         }
         // Leaves end - fakeplayer support
 
