From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HyacinthHaru <122684177+HyacinthHaru@users.noreply.github.com>
Date: Thu, 16 Apr 2026 15:11:00 +0800
Subject: [PATCH] Fix @Nullable annotation position on IRegionFile


diff --git a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index 52ec6ef13e5c6193bd6e64ccfd0a021d3c02fca2..3f2bda6b84796cfafde66a023be2043dfe589198 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -283,7 +283,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         this.isChunkData = info.dfuType()[0] == net.minecraft.util.datafix.DataFixTypes.CHUNK; // Paper - recalculate region file headers
     }
 
-    @org.jetbrains.annotations.Contract("_, false -> !null") private @Nullable org.leavesmc.leaves.region.IRegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit // Leaves - more region format
+    @org.jetbrains.annotations.Contract("_, false -> !null") private org.leavesmc.leaves.region.@Nullable IRegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit // Leaves - more region format
         // Paper start - rewrite chunk system
         if (existingOnly) {
             return this.moonrise$getRegionFileIfExists(pos.x(), pos.z());
