From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix Paper config preventMovingIntoUnloadedChunks

Co-authored-by: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>

diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1881e388d24da249a1a59f695f2945ec33f2fe2f..6e08ceaef972ee9c0bc30ae6b728ce02843fd3f4 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1780,11 +1780,15 @@ public class ServerGamePacketListenerImpl
     private boolean hasNewCollision(final ServerLevel level, final Entity entity, final AABB oldBox, final AABB newBox) {
         final List<AABB> collisionsBB = new java.util.ArrayList<>();
         final List<VoxelShape> collisionsVoxel = new java.util.ArrayList<>();
+        // Leaves start - Fix Paper config preventMovingIntoUnloadedChunks
+        int flags = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER;
+        if (level.paperConfig().chunks.preventMovingIntoUnloadedChunks) flags |= ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_COLLIDE_WITH_UNLOADED_CHUNKS;
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisions(
             level, entity, newBox, collisionsVoxel, collisionsBB,
-            ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_COLLIDE_WITH_UNLOADED_CHUNKS | ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER,
+            flags,
             null, null
         );
+        // Leaves end - Fix Paper config preventMovingIntoUnloadedChunks
 
         for (int i = 0, len = collisionsBB.size(); i < len; ++i) {
             final AABB box = collisionsBB.get(i);
