From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 19 Jul 2023 19:59:58 +0800
Subject: [PATCH] Disable packet limit


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 42777adb028fe282c1619aeb5431c442ad5df0d0..90d448f9b183d0dd31db8eaf43bd039eeace5c65 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -282,7 +282,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
 
             public boolean isEnabled() {
-                return this.interval > 0.0 && this.maxPacketRate > 0.0;
+                return !org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit && (this.interval > 0.0 && this.maxPacketRate > 0.0); // Leaves - disable packet limit
             }
 
             public enum ViolateAction {
