From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:23 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3c130ecb6c2898596d5e4749b04b4aa14c34f289..164830c0d9bae7849a4953e489ea144919d9911a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -317,6 +317,7 @@ public final class CraftServer implements Server {
     private final ServerConfiguration serverConfig = new PaperServerConfiguration();
     private final org.leavesmc.leaves.entity.bot.CraftBotManager botManager; // Leaves
     private final org.leavesmc.leaves.entity.photographer.CraftPhotographerManager photographerManager = new org.leavesmc.leaves.entity.photographer.CraftPhotographerManager(); // Leaves
+    private final org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler internalBytebufHandler = new org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler(); // Leaves
 
     // Paper start - Folia region threading API
     private final io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler regionizedScheduler = new io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler();
@@ -3203,4 +3204,15 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Leaves start - Bytebuf API
+    @Override
+    public org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager() {
+        return internalBytebufHandler.getManager();
+    }
+
+    public org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler getBytebufHandler() {
+        return internalBytebufHandler;
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 231007b5f7a97e0fa4cc2f259ed8ac9740895298..25b9ff06411d1675672a7ee630dcfab5baf8d974 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3503,4 +3503,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     public PlayerGameConnection getConnection() {
         return this.getHandle().connection.playerGameConnection;
     }
+
+    // Leaves start - Bytebuf API
+    @Override
+    public void sendPacket(org.leavesmc.leaves.bytebuf.packet.Packet packet) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), packet);
+    }
+
+    @Override
+    public void sendPacket(org.leavesmc.leaves.bytebuf.Bytebuf bytebuf, org.leavesmc.leaves.bytebuf.packet.PacketType type) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), new org.leavesmc.leaves.bytebuf.packet.Packet(type, bytebuf));
+    }
+    // Leaves end - Bytebuf API
 }
