From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:58:42 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 8233fc1702f9063f5b94929e9771dc9cca5899c9..913c8173d960882b6a9bb82f36de8ab0c394c972 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2992,4 +2992,11 @@ public final class CraftServer implements Server {
         return internalBytebufHandler;
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    @Override
+    public org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return org.leavesmc.leaves.config.InternalConfigProvider.INSTANCE;
+    }
+    // Leaves end - Config API
 }
