From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:23 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index af648b9301a7a964af88a55e181b68780a9f4ca5..09a9c062a0ad15df3896fba161bdfb17d34181ba 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -317,6 +317,7 @@ public final class CraftServer implements Server {
     public final io.papermc.paper.SparksFly spark; // Paper - spark
     private final org.leavesmc.leaves.entity.CraftBotManager botManager; // Leaves
     private final org.leavesmc.leaves.entity.CraftPhotographerManager photographerManager = new org.leavesmc.leaves.entity.CraftPhotographerManager(); // Leaves
+    private final org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler internalBytebufHandler = new org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler(); // Leaves
 
     // Paper start - Folia region threading API
     private final io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler regionizedScheduler = new io.papermc.paper.threadedregions.scheduler.FallbackRegionScheduler();
@@ -3311,4 +3312,15 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Leaves start - Bytebuf API
+    @Override
+    public org.leavesmc.leaves.bytebuf.BytebufManager getBytebufManager() {
+        return internalBytebufHandler.getManager();
+    }
+
+    public org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler getBytebufHandler() {
+        return internalBytebufHandler;
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ed687326dfda04f006815f66c48e978221ce653c..f6fb0b7e7ee67c5a1dfe1ff113c774f402af3d32 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3619,4 +3619,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void setSimplifyContainerDesyncCheck(final boolean simplifyContainerDesyncCheck) {
         this.simplifyContainerDesyncCheck = simplifyContainerDesyncCheck;
     }
+
+    // Leaves start - Bytebuf API
+    @Override
+    public void sendPacket(org.leavesmc.leaves.bytebuf.packet.Packet packet) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), packet);
+    }
+
+    @Override
+    public void sendPacket(org.leavesmc.leaves.bytebuf.Bytebuf bytebuf, org.leavesmc.leaves.bytebuf.packet.PacketType type) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), new org.leavesmc.leaves.bytebuf.packet.Packet(type, bytebuf));
+    }
+    // Leaves end - Bytebuf API
 }
