From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:58:42 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 8481618e09ed02afa04393a36c384237d0cca59d..9504f84d4a75419bc5551534ce7914f9a33c09f9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3001,4 +3001,11 @@ public final class CraftServer implements Server {
         return internalBytebufHandler;
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    @Override
+    public org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return org.leavesmc.leaves.config.InternalConfigProvider.INSTANCE;
+    }
+    // Leaves end - Config API
 }
