From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Sun, 6 Jul 2025 13:10:13 +0800
Subject: [PATCH] Old ender dragon part can use end portal


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftComplexPart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftComplexPart.java
index e6d884596d20f3e970cc6f089f178e642773290d..efd18ffd9014692b3d481a12c4062d04e004a609 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftComplexPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftComplexPart.java
@@ -12,14 +12,19 @@ public class CraftComplexPart extends CraftEntity implements ComplexEntityPart {
         super(server, entity);
     }
 
-    @Override
-    public EnderDragonPart getHandle() {
-        return (EnderDragonPart) this.entity;
-    }
+    // @Override // Leaves start - endDragonPartCanUseEndPortal
+    // public EnderDragonPart getHandle() {
+    //     return (EnderDragonPart) this.entity;
+    // }
 
     @Override
     public ComplexLivingEntity getParent() {
-        return (ComplexLivingEntity) this.getHandle().parentMob.getBukkitEntity();
+        net.minecraft.world.entity.Entity entity = this.getHandle();
+        if (entity instanceof EnderDragonPart part) return (ComplexLivingEntity) part.parentMob.getBukkitEntity();
+        CraftEnderDragon dragon = new org.bukkit.craftbukkit.entity.CraftEnderDragon((org.bukkit.craftbukkit.CraftServer) org.bukkit.Bukkit.getServer(), (net.minecraft.world.entity.boss.enderdragon.EnderDragon) entity);
+        entity.bukkitEntity = dragon;
+        return dragon;
+        // Leaves end - endDragonPartCanUseEndPortal
     }
 
     @Override
