From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Fri, 26 Sep 2025 20:23:59 +0800
Subject: [PATCH] TEMP remove mappings for build


diff --git a/src/main/java/io/papermc/paper/util/ObfHelper.java b/src/main/java/io/papermc/paper/util/ObfHelper.java
index 6067be951c4c52c4b1da51efc01436b2c90ea3bf..0e75169592debca8f579da95fd0b0c6a79abf7d7 100644
--- a/src/main/java/io/papermc/paper/util/ObfHelper.java
+++ b/src/main/java/io/papermc/paper/util/ObfHelper.java
@@ -21,13 +21,13 @@ public enum ObfHelper {
     private final @Nullable Map<String, ClassMapping> mappingsByMojangName;
 
     ObfHelper() {
-        final @Nullable Set<ClassMapping> maps = loadMappingsIfPresent();
+        final @Nullable Set<ClassMapping> maps = null;
         if (maps != null) {
             this.mappingsByObfName = maps.stream().collect(Collectors.toUnmodifiableMap(ClassMapping::obfName, map -> map));
             this.mappingsByMojangName = maps.stream().collect(Collectors.toUnmodifiableMap(ClassMapping::mojangName, map -> map));
         } else {
-            this.mappingsByObfName = null;
-            this.mappingsByMojangName = null;
+            this.mappingsByObfName = new java.util.HashMap<>();
+            this.mappingsByMojangName = new java.util.HashMap<>();
         }
     }
 
