--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -217,10 +_,16 @@
         }
     }
 
-    // Contributed by Techcable <Techcable@outlook.com> in GH-65
     private static int fetchDistanceFromGitHub(final String branch, final String hash) {
+        // Leaves start - build changes
+        return fetchDistanceFromGitHub(PaperVersionFetcher.REPOSITORY, branch, hash);
+    }
+
+    // Contributed by Techcable <Techcable@outlook.com> in GH-65
+    protected static int fetchDistanceFromGitHub(final String repo, final String branch, final String hash) {
         try {
-            final HttpURLConnection connection = (HttpURLConnection) URI.create("https://api.github.com/repos/%s/compare/%s...%s".formatted(PaperVersionFetcher.REPOSITORY, branch, hash)).toURL().openConnection();
+            final HttpURLConnection connection = (HttpURLConnection) URI.create("https://api.github.com/repos/%s/compare/%s...%s".formatted(repo, branch, hash)).toURL().openConnection();
+            // Leaves end - build changes
             connection.setConnectTimeout(5000);
             connection.setReadTimeout(5000);
             connection.setRequestProperty("User-Agent", PaperVersionFetcher.USER_AGENT);
@@ -244,7 +_,7 @@
         }
     }
 
-    private @Nullable Component getHistory() {
+    protected @Nullable Component getHistory() { // Leaves - private -> protected
         final VersionHistoryManager.@Nullable VersionData data = VersionHistoryManager.INSTANCE.getVersionData();
         if (data == null) {
             return null;
