From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <admin@s-yh-china.com>
Date: Fri, 29 Oct 2021 16:59:07 +0800
Subject: [PATCH] Leaves Server Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2204336d8800311b65e894739ab1b27273e7c6f2..da7d83aee6dff653f131718233a21290e76242de 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1970,6 +1970,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Paper end
+		
+		// Leaves start - Not supported yet too
+		@NotNull
+		public org.bukkit.configuration.file.YamlConfiguration getLeavesConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+		// Leaves end - Not supported yet too
 
         /**
          * Sends the component to the player
