From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <admin@s-yh-china.com>
Date: Fri, 29 Oct 2021 16:59:07 +0800
Subject: [PATCH] Leaves Server Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9f7029d327b55c8b1b9cfef8ea2678b87dad8ddb..2d14421051ffa6951aef42f3d2742e7207f63859 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1966,6 +1966,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Paper end
+		
+		// Leaves start - Not supported yet too
+		@NotNull
+		public org.bukkit.configuration.file.YamlConfiguration getLeavesConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+		// Leaves end - Not supported yet too
 
         /**
          * Sends the component to the player
