From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <admin@s-yh-china.com>
Date: Fri, 29 Oct 2021 16:59:07 +0800
Subject: [PATCH] Leaves Server Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b4f8281d3797ec825a7671f38077cd65d5a1d76e..127a59fdf5318f6cb22b159fc5a5089d3ed2ac96 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2213,6 +2213,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Paper end
+		
+		// Leaves start - Not supported yet too
+		@NotNull
+		public org.bukkit.configuration.file.YamlConfiguration getLeavesConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+		// Leaves end - Not supported yet too
 
         /**
          * Sends the component to the player
