From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 30 Mar 2022 08:58:32 +0000
Subject: [PATCH] Add isShrink to EntityResurrectEvent


diff --git a/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java b/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
index fa6b6c3b047edd1653f3c81027ec752821bcdb09..4e2e345bafa812c37a46b86e20fb2ce9c8660425 100644
--- a/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityResurrectEvent.java
@@ -13,11 +13,36 @@ import org.jetbrains.annotations.NotNull;
 public class EntityResurrectEvent extends EntityEvent implements Cancellable {
 
     private static final HandlerList handlers = new HandlerList();
-    //
+    
     private boolean cancelled;
+    private boolean shrink;
 
     public EntityResurrectEvent(@NotNull LivingEntity what) {
         super(what);
+        this.shrink = true;
+    }
+
+    public EntityResurrectEvent(@NotNull LivingEntity what, boolean shrink) {
+        super(what);
+        this.shrink = shrink;
+    }
+
+    /**
+     * Get is shrink item
+     *
+     * @return is shrink item
+     */
+    public boolean isShrink() {
+        return shrink;
+    }
+
+    /**
+     * Set is shrink item
+     *
+     * @param shrink is shrink item
+     */
+    public void setShrink(boolean shrink) {
+        this.shrink = shrink;
     }
 
     @NotNull
