From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 11 Dec 2022 18:43:36 +0800
Subject: [PATCH] Player operation limiter


diff --git a/src/main/java/org/leavesmc/leaves/event/player/PlayerOperationLimitEvent.java b/src/main/java/org/leavesmc/leaves/event/player/PlayerOperationLimitEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f5852eb6d53db5774db0ab4eba50bcb8733cfd6
--- /dev/null
+++ b/src/main/java/org/leavesmc/leaves/event/player/PlayerOperationLimitEvent.java
@@ -0,0 +1,56 @@
+package org.leavesmc.leaves.event.player;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when a player operation is limited
+ */
+public class PlayerOperationLimitEvent extends PlayerEvent {
+    private static final HandlerList handlers = new HandlerList();
+
+    private final Block block;
+    private final Operation operation;
+
+    public PlayerOperationLimitEvent(@NotNull Player who, Operation operation, Block block) {
+        super(who);
+        this.block = block;
+        this.operation = operation;
+    }
+
+    /**
+     * Gets the operated block
+     *
+     * @return block
+     */
+    public Block getBlock() {
+        return block;
+    }
+
+    /**
+     * Gets the type of operation
+     *
+     * @return operation type
+     */
+    public Operation getOperation() {
+        return operation;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    public enum Operation {
+        MINE, PLACE
+    }
+}
