From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 19 Jul 2023 20:08:16 +0800
Subject: [PATCH] Hide irrelevant compilation warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index afe9e9b08499ac504f4198ffb8da34cd286a6b44..c3f808e08ceaa48c982cf42c669aaee7451bfcbb 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -89,6 +89,15 @@ val generateApiVersioningFile by tasks.registering {
     }
 }
 
+// Leaves start - hide irrelevant compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Leaves end - hide irrelevant compilation warnings
+
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
         into("META-INF/maven/${project.group}/${project.name}")
@@ -146,6 +155,8 @@ tasks.withType<Javadoc> {
             into("build/docs/javadoc")
         }
     }
+
+    options.addStringOption("Xdoclint:none", "-quiet") // Leaves - hide irrelevant compilation warnings
 }
 
 // Paper start
