From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 19 Jul 2023 20:09:38 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index c070e3d308b14cc020475c2644955e7546ef6cef..286349a28b06eddebbf103def9bd33bcc224f090 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -136,6 +136,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -199,6 +200,7 @@ tasks.withType<Javadoc> {
     }
 
     options.addStringOption("Xdoclint:none", "-quiet") // Leaves - hide irrelevant compilation warnings
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Leaves - SIMD support
 }
 
 tasks.test {
