From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 19 Jul 2023 20:09:38 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index 580a37740f04a4aad435fc50b521a59a772f16af..45c2e0b9417cc5e0d54dc007f1259c0568966aab 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -95,6 +95,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -157,6 +158,7 @@ tasks.withType<Javadoc> {
     }
 
     options.addStringOption("Xdoclint:none", "-quiet") // Leaves - hide irrelevant compilation warnings
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Leaves - SIMD support
 }
 
 tasks.test {
