From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 02:33:52 +0800
Subject: [PATCH] Leaves API publish


diff --git a/build.gradle.kts b/build.gradle.kts
index 311a3134cf4b60d48c3096aebe500106dd4bd267..c124ac034906a065514f5a1af53b88e550ba37eb 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -212,3 +212,14 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+
+// Leaves start - publish api
+publishing {
+    repositories {
+        maven("https://repo.leavesmc.top/snapshots") {
+            name = "leaves"
+            credentials(PasswordCredentials::class)
+        }
+    }
+}
+// Leaves end - publish api
\ No newline at end of file
