From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 02:33:52 +0800
Subject: [PATCH] Leaves API publish


diff --git a/build.gradle.kts b/build.gradle.kts
index e0f2438944bda900f3a6391797e7842647b844be..bb020dc0ad0cbc1d342966dbf5fce9d11547f56e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -212,3 +212,14 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+
+// Leaves start - publish api
+publishing {
+    repositories {
+        maven("https://repo.leavesmc.top/snapshots") {
+            name = "leaves"
+            credentials(PasswordCredentials::class)
+        }
+    }
+}
+// Leaves end - publish api
\ No newline at end of file
