From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 02:33:52 +0800
Subject: [PATCH] Leaves API publish


diff --git a/build.gradle.kts b/build.gradle.kts
index 45c2e0b9417cc5e0d54dc007f1259c0568966aab..7c50ed670facceac89df004416ce417237d88d11 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -181,3 +181,14 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+
+// Leaves start - publish api
+publishing {
+    repositories {
+        maven("https://repo.leavesmc.top/snapshots") {
+            name = "leaves"
+            credentials(PasswordCredentials::class)
+        }
+    }
+}
+// Leaves end - publish api
\ No newline at end of file
