From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 02:33:52 +0800
Subject: [PATCH] Leaves API publish


diff --git a/build.gradle.kts b/build.gradle.kts
index 03c8e56ba97badce267b3e882dfba5d97708baae..37195381168ddf116a722924323f694479dfbbdf 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -209,3 +209,14 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+
+// Leaves start - publish api
+publishing {
+    repositories {
+        maven("https://repo.leavesmc.top/snapshots") {
+            name = "leaves"
+            credentials(PasswordCredentials::class)
+        }
+    }
+}
+// Leaves end - publish api
\ No newline at end of file
