From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 3 Aug 2023 14:21:55 +0800
Subject: [PATCH] Force peaceful mode switch


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index fdb87adfb8d6eff2bfabe7a41398c53d15d4cd98..531a6cae97fe2a415cf906108a0899bf04e613d8 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4344,6 +4344,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     void setSendViewDistance(int viewDistance);
     // Paper end - view distance api
 
+    // Leaves start - unsupported settings
+    public void setPeacefulModeSwitchTick(int tick);
+
+    public int getPeacefulModeSwitchTick();
+    // Leaves end - unsupported settings
+
     /**
      * Gets all generated structures that intersect the chunk at the given
      * coordinates. <br>
