From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 3 Aug 2023 14:21:55 +0800
Subject: [PATCH] Force peaceful mode switch


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 8d861f5522a33669f67b3e41dfbc5234637114b4..54fb966467cf2aa13bbc39b6360517ff6541980d 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4018,6 +4018,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public Set<FeatureFlag> getFeatureFlags();
 
+    // Leaves start - unsupported settings
+    public void setPeacefulModeSwitchTick(int tick);
+
+    public int getPeacefulModeSwitchTick();
+    // Leaves end - unsupported settings
+
     /**
      * Represents various map environment types that a world may be
      */
