From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7studio@foxmail.com>
Date: Fri, 10 May 2024 23:59:49 +0800
Subject: [PATCH] Placeholder for Replay-Mod-API


diff --git a/.gitignore b/.gitignore
index 9764fa643039f215627c20a33ca70c9e36b2d599..97e78e27ee0eea2c8b24886eeb19164d552323fe 100644
--- a/.gitignore
+++ b/.gitignore
@@ -38,4 +38,3 @@
 # vs code
 /.vscode
 /.factorypath
-
