From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 3 Aug 2023 14:21:55 +0800
Subject: [PATCH] Force peaceful mode switch


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index da524a71af74b02515b037f7fe09ba6988e2c8bf..649a80fc7cb12c4d2fe2c472178b6ca3b16372f7 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4016,6 +4016,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public Set<FeatureFlag> getFeatureFlags();
 
+    // Leaves start - unsupported settings
+    public void setPeacefulModeSwitchTick(int tick);
+
+    public int getPeacefulModeSwitchTick();
+    // Leaves end - unsupported settings
+
     /**
      * Represents various map environment types that a world may be
      */
