From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 31 Oct 2021 09:33:38 +0800
Subject: [PATCH] Support Java16+


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 3370e8771564d2656b9dd5004f98a202253f6073..fb03995b2c2d94779a9ea8d5faed82adb71e78cb 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -231,8 +231,7 @@ public class Main {
                 return;
             }
             if (javaVersion > 60.0) {
-                System.err.println("Unsupported Java detected (" + javaVersion + "). Only up to Java 16 is supported.");
-                if (!Boolean.getBoolean("Paper.IgnoreJavaVersion")) return; // Paper
+                System.err.println("Unsupported Java detected (" + javaVersion + "). "); // Leaves - support Java16+
             }
 
             try {
