From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 27 Jan 2022 20:11:05 +0800
Subject: [PATCH] Player can edit sign


diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index 4c290eb2d4abf9e9b923b8d0878f319328b7cf8c..d6abf25f70f6f5a19787c72e87fddd2801b47def 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -93,7 +93,9 @@ public final class LeavesConfig {
 
     private static void updateConfigVersion(final YamlConfiguration config) {
         if (configVersion < CURRENT_CONFIG_VERSION) {
+            playerCanEditSign = config.getBoolean("settings.player-can-edit-sign", playerCanEditSign);
 
+            config.set("settings.player-can-edit-sign", null);
         }
     }
 
@@ -132,6 +134,11 @@ public final class LeavesConfig {
         return LeavesConfig.config.getString(path, dfl);
     }
 
+    public static boolean playerCanEditSign = true;
+    private static void playerCanEditSign() {
+        playerCanEditSign = getBoolean("settings.modify.player-can-edit-sign", playerCanEditSign);
+    }
+
     public static final class WorldConfig {
 
         public final String worldName;
