From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 2 Feb 2022 00:43:25 +0800
Subject: [PATCH] Allow vanilla properties def


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 88a4dda44e59fbe6215d7ac2e5af0c54527a2fc7..0a65617d5804636ee8c7a739d21b886ec635e075 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -472,10 +472,10 @@ public class PaperConfig {
         useRgbForNamedTextColors = getBoolean("settings.loggers.use-rgb-for-named-text-colors", useRgbForNamedTextColors);
     }
 
-    public static boolean allowBlockPermanentBreakingExploits = false;
+    public static boolean allowBlockPermanentBreakingExploits = true;
     private static void allowBlockPermanentBreakingExploits() {
         if (config.contains("allow-perm-block-break-exploits")) {
-            allowBlockPermanentBreakingExploits = config.getBoolean("allow-perm-block-break-exploits", false);
+            allowBlockPermanentBreakingExploits = config.getBoolean("allow-perm-block-break-exploits", true);
             config.set("allow-perm-block-break-exploits", null);
         }
 
@@ -488,10 +488,10 @@ public class PaperConfig {
         consoleHasAllPermissions = getBoolean("settings.console-has-all-permissions", consoleHasAllPermissions);
     }
 
-    public static boolean allowPistonDuplication;
+    public static boolean allowPistonDuplication = true;
     private static void allowPistonDuplication() {
         config.set("settings.unsupported-settings.allow-piston-duplication-readme", "This setting controls if player should be able to use TNT duplication, but this also allows duplicating carpet, rails and potentially other items");
-        allowPistonDuplication = getBoolean("settings.unsupported-settings.allow-piston-duplication", config.getBoolean("settings.unsupported-settings.allow-tnt-duplication", false));
+        allowPistonDuplication = getBoolean("settings.unsupported-settings.allow-piston-duplication", config.getBoolean("settings.unsupported-settings.allow-tnt-duplication", true));
         set("settings.unsupported-settings.allow-tnt-duplication", null);
     }
 
@@ -512,10 +512,10 @@ public class PaperConfig {
         }
     }
 
-    public static boolean allowHeadlessPistons;
+    public static boolean allowHeadlessPistons = true;
     private static void allowHeadlessPistons() {
         config.set("settings.unsupported-settings.allow-headless-pistons-readme", "This setting controls if players should be able to create headless pistons.");
-        allowHeadlessPistons = getBoolean("settings.unsupported-settings.allow-headless-pistons", false);
+        allowHeadlessPistons = getBoolean("settings.unsupported-settings.allow-headless-pistons", true);
     }
 
     public static int itemValidationDisplayNameLength = 8192;
