From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 2 Feb 2022 00:43:25 +0800
Subject: [PATCH] Allow vanilla properties def


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 63ec2ebb71aa0e0dbb64bbce7cd3c9494e9ce2e7..c04b68b679f8fde29bb246b5fd0262e0bd1d7356 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -135,12 +135,12 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     public class UnsupportedSettings extends ConfigurationPart {
         @Comment("This setting controls if players should be able to break bedrock, end portals and other intended to be permanent blocks.")
-        public boolean allowPermanentBlockBreakExploits = false;
+        public boolean allowPermanentBlockBreakExploits = true;
         @Comment("This setting controls if player should be able to use TNT duplication, but this also allows duplicating carpet, rails and potentially other items")
-        public boolean allowPistonDuplication = false;
+        public boolean allowPistonDuplication = true;
         public boolean performUsernameValidation = true;
         @Comment("This setting controls if players should be able to create headless pistons.")
-        public boolean allowHeadlessPistons = false;
+        public boolean allowHeadlessPistons = true;
         @Comment("This setting controls if grindstones should be able to output overstacked items (such as cursed books).")
         public boolean allowGrindstoneOverstacking = false;
     }
