From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 27 Feb 2022 14:07:57 +0800
Subject: [PATCH] Make shears in dispenser can unlimited use


diff --git a/src/main/java/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
index 2366d411bf64f88c7296e888cd3bf584825ae4a9..7da29029038d5a007735dcb2be0114581d85d098 100644
--- a/src/main/java/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
@@ -63,7 +63,7 @@ public class ShearsDispenseItemBehavior extends OptionalDispenseItemBehavior {
             BlockPos blockposition = pointer.getPos().relative((Direction) pointer.getBlockState().getValue(DispenserBlock.FACING));
 
             this.setSuccess(ShearsDispenseItemBehavior.tryShearBeehive((ServerLevel) worldserver, blockposition) || ShearsDispenseItemBehavior.tryShearLivingEntity((ServerLevel) worldserver, blockposition, bukkitBlock, craftItem)); // CraftBukkit
-            if (this.isSuccess() && stack.hurt(1, worldserver.getRandom(), (ServerPlayer) null)) {
+            if (this.isSuccess() && stack.hurt(1, worldserver.getRandom(), (ServerPlayer) null) && !top.leavesmc.leaves.LeavesConfig.shearsInDispenserCanZeroAmount) { // Leaves - Make shears in dispenser can unlimited use
                 stack.setCount(0);
             }
         }
diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index d4000ba0a8cce6f3da43e813a283665c5c8ba8ab..766d2bb1eceb570f6ebdfec81bc88c6b4d4620b1 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -95,27 +95,6 @@ public final class LeavesConfig {
         LeavesConfig.config.set(path, value);
     }
 
-    public static boolean playerCanEditSign = true;
-    private static void playerCanEditSign() {
-        playerCanEditSign = getBoolean("settings.player-can-edit-sign", playerCanEditSign);
-    }
-
-    public static boolean snowballAndEggCanKnockback = true;
-    private static void snowballAndEggCanKnockback() {
-        snowballAndEggCanKnockback = getBoolean("settings.snowball-and-egg-can-knockback-player", snowballAndEggCanKnockback);
-    }
-
-    public static boolean fakeplayerSupport = true;
-    private static void fakeplayerSupport() {
-        fakeplayerSupport = getBoolean("settings.fakeplayer-support", fakeplayerSupport);
-    }
-
-    public static void registerCommands() {
-        for (Map.Entry<String, Command> entry : commands.entrySet()) {
-            MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Leaves", entry.getValue());
-        }
-    }
-
     static boolean getBoolean(final String path, final boolean dfl) {
         LeavesConfig.config.addDefault(path, Boolean.valueOf(dfl));
         return LeavesConfig.config.getBoolean(path, dfl);
@@ -141,6 +120,32 @@ public final class LeavesConfig {
         return LeavesConfig.config.getString(path, dfl);
     }
 
+    public static boolean playerCanEditSign = true;
+    private static void playerCanEditSign() {
+        playerCanEditSign = getBoolean("settings.player-can-edit-sign", playerCanEditSign);
+    }
+
+    public static boolean snowballAndEggCanKnockback = true;
+    private static void snowballAndEggCanKnockback() {
+        snowballAndEggCanKnockback = getBoolean("settings.snowball-and-egg-can-knockback-player", snowballAndEggCanKnockback);
+    }
+
+    public static boolean fakeplayerSupport = true;
+    private static void fakeplayerSupport() {
+        fakeplayerSupport = getBoolean("settings.fakeplayer-support", fakeplayerSupport);
+    }
+
+    public static boolean shearsInDispenserCanZeroAmount = false;
+    private static void shearsInDispenserCanZeroAmount() {
+        shearsInDispenserCanZeroAmount = getBoolean("settings.shears-in-dispenser-can-zero-amount", shearsInDispenserCanZeroAmount);
+    }
+
+    public static void registerCommands() {
+        for (Map.Entry<String, Command> entry : commands.entrySet()) {
+            MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Leaves", entry.getValue());
+        }
+    }
+
     public static final class WorldConfig {
 
         public final String worldName;
