From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 3 Apr 2022 11:31:04 +0800
Subject: [PATCH] Budding Amethyst can push by piston


diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index b71496a3c0d9f9d69dadb652e2f35c09cf906bc0..da2fc9a9e1ff9d8c8bcb1e325916ae0eaff29b1e 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -184,7 +184,7 @@ public final class LeavesConfig {
 
     public static boolean buddingAmethystCanPushByPiston = false;
     private static void buddingAmethystCanPushByPiston() {
-        buddingAmethystCanPushByPiston = getBoolean("settings.modify.budding-amethyst-can-push-by-piston", buddingAmethystCanPushByPiston);
+        buddingAmethystCanPushByPiston = getBoolean("settings.modify.budding-amethyst-can-push-by-piston", buddingAmethystCanPushByPiston); // void
     }
 
     public static boolean spectatorDontGetAdvancement = false;
