From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 18:04:35 +0800
Subject: [PATCH] Spectator dont get Advancement


diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index 6de539d8819b9c25fbf96cd2ac8ef4a87c0fb231..d8ae25c5c012b18f5b2ca7d5ec92b0965fb2087a 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -49,6 +49,7 @@ import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.world.level.GameRules;
+import net.minecraft.world.level.GameType;
 import org.slf4j.Logger;
 import top.leavesmc.leaves.bot.Bot;
 
@@ -280,6 +281,11 @@ public class PlayerAdvancements {
             return false;
         }
         // Leaves end - bot can't get advancement
+        // Leaves start - spectator don't get advancement
+        if (top.leavesmc.leaves.LeavesConfig.spectatorDontGetAdvancement && player.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
+            return false;
+        }
+        // Leaves end - spectator don't get advancement
         boolean flag = false;
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
         boolean flag1 = advancementprogress.isDone();
diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index af924f19ca0c77484ae3cdeb2994547b63213767..17f5bc4b4197868bf571c3bb53525dd21e6d0a46 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -175,6 +175,11 @@ public final class LeavesConfig {
     private static void buddingAmethystCanPushByPiston() {
         buddingAmethystCanPushByPiston = getBoolean("settings.budding-amethyst-can-push-by-piston", buddingAmethystCanPushByPiston);
     }
+
+    public static boolean spectatorDontGetAdvancement = false;
+    private static void  spectatorDontGetAdvancement() {
+        spectatorDontGetAdvancement = getBoolean("settings.spectator-dont-get-advancement", spectatorDontGetAdvancement);
+    }
     
     public static final class WorldConfig {
 
