From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH] Stick can change ArmorStand arm status


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 2f398750bfee5758ad8b1367b6fc14364e4de776..df29a54063ee957c2b88a12ef228c7d8541a2f2c 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -392,6 +392,12 @@ public class ArmorStand extends LivingEntity {
                         return InteractionResult.SUCCESS;
                     }
                 } else {
+                    // Leaves start - stick can change ArmorStand arm status
+                    if (org.leavesmc.leaves.LeavesConfig.stickChangeArmorStandArmStatus && itemstack.is(Items.STICK) && player.isShiftKeyDown()) {
+                        setShowArms(!isShowArms());
+                    }
+                    // Leaves end - stick can change ArmorStand arm status
+
                     if (this.isDisabled(enumitemslot)) {
                         return InteractionResult.FAIL;
                     }
