From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 25 Jun 2022 19:54:23 +0800
Subject: [PATCH] Stick can change ArmorStand arm status


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 150afceb491cfd254c0f1b84800e6df14cf26676..e108e757a6dab69784324625e9be00afbef6689e 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -384,6 +384,12 @@ public class ArmorStand extends LivingEntity {
                         return InteractionResult.SUCCESS;
                     }
                 } else {
+                    // Leaves start - stick can change ArmorStand arm status
+                    if (top.leavesmc.leaves.LeavesConfig.stickChangeArmorStandArmStatus && itemstack.is(Items.STICK) && player.isShiftKeyDown()) {
+                        setShowArms(!isShowArms());
+                    }
+                    // Leaves end - stick can change ArmorStand arm status
+
                     if (this.isDisabled(enumitemslot)) {
                         return InteractionResult.FAIL;
                     }
diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index 17f5bc4b4197868bf571c3bb53525dd21e6d0a46..6b5551df69952d57006c2303e5f13224b22c5d95 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -180,6 +180,11 @@ public final class LeavesConfig {
     private static void  spectatorDontGetAdvancement() {
         spectatorDontGetAdvancement = getBoolean("settings.spectator-dont-get-advancement", spectatorDontGetAdvancement);
     }
+
+    public static boolean stickChangeArmorStandArmStatus = true;
+    private static void  stickChangeArmorStandHasArm() {
+        stickChangeArmorStandArmStatus = getBoolean("settings.stick-change-armorstand-arm-status", stickChangeArmorStandArmStatus);
+    }
     
     public static final class WorldConfig {
 
