From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 30 Jul 2022 01:11:30 +0800
Subject: [PATCH] Fix tripwire update


diff --git a/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java b/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
index 8bf94fcf5c46ffe6d90f8431338b768ef1a2dd61..d9b3877257b31ca1b5acc4a47fbf5b993de69ae0 100644
--- a/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -201,8 +201,9 @@ public class TripWireHookBlock extends Block {
                     BlockState iblockdata4 = aiblockdata[l];
 
                     if (iblockdata4 != null) {
-                        if (world.getBlockState(blockposition2).is(Blocks.TRIPWIRE)) { // Paper - Fix tripwire disarming not working as intended
                         world.setBlock(blockposition2, (BlockState) iblockdata4.trySetValue(TripWireHookBlock.ATTACHED, flag4), 3);
+                        if (!world.getBlockState(blockposition2).isAir()) {
+                            ;
                         }
                     }
                 }
