From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 9 Jul 2023 16:51:47 +0800
Subject: [PATCH] Raider die skip self raid check


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raider.java b/src/main/java/net/minecraft/world/entity/raid/Raider.java
index ab132041982df2a701e4baea8195873f31b4a5fb..9b911254b24bc77930c518a9c61916983ba72e3c 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raider.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raider.java
@@ -124,7 +124,7 @@ public abstract class Raider extends PatrollingMonster {
             Raid raid = this.getCurrentRaid();
 
             if (raid != null) {
-                if (this.isPatrolLeader()) {
+                if (!org.leavesmc.leaves.LeavesConfig.modify.skipSelfRaidCheck && this.isPatrolLeader()) { // Leaves - skip self raid check
                     raid.removeLeader(this.getWave());
                 }
 
