From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 20 May 2023 21:46:13 +0800
Subject: [PATCH] Fix trapdoor feature


diff --git a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
index 29e43313ff6bd9b10cf6409e3519c0b2603b431e..1646170dc41f8cefac46d95378b45c919a23cad4 100644
--- a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -125,26 +125,7 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                     flag1 = eventRedstone.getNewCurrent() > 0;
                 }
                 // CraftBukkit end
-                boolean open = (Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1; // Paper - break redstone on trapdoors early
-                // Paper start - break redstone on trapdoors early
-                // note: this must run before any state for this block/its neighborus are written to the world
-                // we allow the redstone event to fire so that plugins can block
-                if (flag1 && open) { // if we are now powered and it caused the trap door to open
-                    // in this case, first check for the redstone on top first
-                    BlockPos abovePos = pos.above();
-                    BlockState above = world.getBlockState(abovePos);
-                    if (above.getBlock() instanceof RedStoneWireBlock) {
-                        world.setBlock(abovePos, Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS | Block.UPDATE_NEIGHBORS);
-                        Block.popResource(world, abovePos, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.REDSTONE));
-                        // now check that this didn't change our state
-                        if (world.getBlockState(pos) != state) {
-                            // our state was changed, so we cannot propagate this update
-                            return;
-                        }
-                    }
-                }
-                // Paper end - break redstone on trapdoors early
-                if (open) { // Paper - break redstone on trapdoors early
+                if ((Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1) { // Leaves - it is feature, not bug!
                     state = (BlockState) state.setValue(TrapDoorBlock.OPEN, flag1);
                     this.playSound((Player) null, world, pos, flag1);
                 }
