From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 7 Jul 2023 16:53:32 +0800
Subject: [PATCH] Lava riptide


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index fa876ddf54780728e7f3ecfe02aa8a16b8ef6f8d..2dad3b64673186c8776f414d68ae392818a2d749 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -66,7 +66,7 @@ public class TridentItem extends Item implements Vanishable {
             if (j >= 10) {
                 int k = EnchantmentHelper.getRiptide(stack);
 
-                if (k <= 0 || entityhuman.isInWaterOrRain()) {
+                if (k <= 0 || entityhuman.isInWaterOrRain() || (top.leavesmc.leaves.LeavesConfig.lavaRiptide && entityhuman.isInLava())) { // Leaves - lava riptide
                     if (!world.isClientSide) {
                         // CraftBukkit - moved down
                         /*
@@ -163,7 +163,7 @@ public class TridentItem extends Item implements Vanishable {
 
         if (itemstack.getDamageValue() >= itemstack.getMaxDamage() - 1) {
             return InteractionResultHolder.fail(itemstack);
-        } else if (EnchantmentHelper.getRiptide(itemstack) > 0 && !user.isInWaterOrRain()) {
+        } else if (EnchantmentHelper.getRiptide(itemstack) > 0 && !user.isInWaterOrRain() && !(top.leavesmc.leaves.LeavesConfig.lavaRiptide && user.isInLava())) { // Leaves - lava riptide
             return InteractionResultHolder.fail(itemstack);
         } else {
             user.startUsingItem(hand);
