From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 17 Jul 2023 22:45:48 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index a2441f300345c3936bce69d6d3f21fdd6626136d..77d6f5eeeb9918a70e0003e97671778303c33ce9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -70,6 +70,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -189,6 +190,8 @@ fun TaskContainer.registerRunTask(
     minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
+    jvmArgs("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
+
     doFirst {
         workingDir.mkdirs()
     }
