From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Jul 2023 17:47:58 +0800
Subject: [PATCH] Disable timings default


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 4ac3fa45cd155ae8a852e26d4d4d1f16b28efdc2..cacd808f1d03c9729bcf2110bffceb940c0e11c0 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -50,8 +50,8 @@ public class GlobalConfiguration extends ConfigurationPart {
     public Timings timings;
 
     public class Timings extends ConfigurationPart.Post {
-        public boolean enabled = true;
-        public boolean verbose = true;
+        public boolean enabled = false; // Leaves - def false
+        public boolean verbose = false; // Leaves - def false
         public String url = "https://timings.aikar.co/";
         public boolean serverNamePrivacy = false;
         public List<String> hiddenConfigEntries = List.of(
