From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 17 Jul 2023 22:45:48 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index fe1f6f1a97ab35c44f596596d16765b8b81753a9..28baa9a192a6fd83563b57a411e9bc905ba6b7e1 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -73,6 +73,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -182,6 +183,8 @@ fun TaskContainer.registerRunTask(
     minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
+    jvmArgs("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
+
     doFirst {
         workingDir.mkdirs()
     }
