From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 17 Jul 2023 22:45:48 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index ef5867918101868020d7bc14949b3b49fecaeecc..84a2c4a77895c4c08a754962d837b98f55aa2e59 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -71,6 +71,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -195,6 +196,8 @@ fun TaskContainer.registerRunTask(
     minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
+    jvmArgs("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
+
     doFirst {
         workingDir.mkdirs()
     }
