From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 7 Jul 2023 16:53:32 +0800
Subject: [PATCH] Lava riptide


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index 8078f127ff4b6e0aafb5804b9c02e237f79445b5..801f066878d6ffe5dabe01d20513db9022b323ec 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -66,7 +66,7 @@ public class TridentItem extends Item implements Vanishable {
             if (j >= 10) {
                 int k = EnchantmentHelper.getRiptide(stack);
 
-                if (k <= 0 || entityhuman.isInWaterOrRain()) {
+                if (k <= 0 || entityhuman.isInWaterOrRain() || (top.leavesmc.leaves.LeavesConfig.lavaRiptide && entityhuman.isInLava())) { // Leaves - lava riptide
                     if (!world.isClientSide) {
                         // CraftBukkit - moved down
                         /*
@@ -162,7 +162,7 @@ public class TridentItem extends Item implements Vanishable {
 
         if (itemstack.getDamageValue() >= itemstack.getMaxDamage() - 1) {
             return InteractionResultHolder.fail(itemstack);
-        } else if (EnchantmentHelper.getRiptide(itemstack) > 0 && !user.isInWaterOrRain()) {
+        } else if (EnchantmentHelper.getRiptide(itemstack) > 0 && !user.isInWaterOrRain() && !(top.leavesmc.leaves.LeavesConfig.lavaRiptide && user.isInLava())) { // Leaves - lava riptide
             return InteractionResultHolder.fail(itemstack);
         } else {
             user.startUsingItem(hand);
diff --git a/src/main/java/top/leavesmc/leaves/LeavesConfig.java b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
index 871ab6819f3c98c2f705689939115f0f53cedbc8..837c76768e6e93c5f8df7817cf644f7e0c53237a 100644
--- a/src/main/java/top/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/top/leavesmc/leaves/LeavesConfig.java
@@ -851,6 +851,7 @@ public final class LeavesConfig {
 
     public static void registerLeavesFeatures() {
         LeavesFeatureSet.register(LeavesFeature.of("mspt_sync", msptSyncProtocol));
+        LeavesFeatureSet.register(LeavesFeature.of("lava_riptide", lavaRiptide));
     }
 
     public static boolean hopperCounter = false;
