From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sat, 22 Jul 2023 12:00:59 +0800
Subject: [PATCH] Fix villagers dont release memory


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f62310f824bd842601fa0fb5526e06b2bd26baf2..dbef7a7b01fb1f609a92d2fdb10ae20d436cd6e8 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4057,7 +4057,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this;
     }
 
-    private Entity teleportCrossDimension(ServerLevel world, TeleportTransition teleportTarget) {
+    protected Entity teleportCrossDimension(ServerLevel world, TeleportTransition teleportTarget) { // Leaves - private -> protected
         List<Entity> list = this.getPassengers();
         List<Entity> list1 = new ArrayList(list.size());
 
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index b7a34f1c4d7b5ef3f7a843d152e33c839dcdedd5..8d3959e86ef1357e49aac6ef3c8285c770f2856a 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -1045,4 +1045,19 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             return worldTime - olong < 24000L;
         }).isPresent();
     }
+
+    // Leaves start - fixes a memory leak when villagers get moved to another world
+    @Override
+    public Entity teleportCrossDimension(ServerLevel world, net.minecraft.world.level.portal.TeleportTransition transition) {
+        if (org.leavesmc.leaves.LeavesConfig.performance.villagersDontReleaseMemoryFix) {
+            this.releaseAllPois();
+            this.getBrain().eraseMemory(MemoryModuleType.HOME);
+            this.getBrain().eraseMemory(MemoryModuleType.JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.MEETING_POINT);
+            this.refreshBrain(transition.newLevel());
+        }
+        return super.teleportCrossDimension(world, transition);
+    }
+    // Leaves end - fixes a memory leak when villagers get moved to another world
 }
