From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 17 Jul 2023 22:45:48 +0800
Subject: [PATCH] SIMD support


diff --git a/build.gradle.kts b/build.gradle.kts
index 955362ab4155718807859978d840d9f4d68a8ef4..3f3e60ecbd24c1c40ebbdaec7402192e111f5407 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -71,6 +71,7 @@ tasks.withType<JavaCompile> {
     compilerArgs.add("-Xlint:-module")
     compilerArgs.add("-Xlint:-removal")
     compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
 }
 // Leaves end - hide irrelevant compilation warnings
 
@@ -188,6 +189,8 @@ fun TaskContainer.registerRunTask(
     minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
+    jvmArgs("--add-modules=jdk.incubator.vector") // Leaves - SIMD support
+
     doFirst {
         workingDir.mkdirs()
     }
