From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 30 Jul 2023 12:20:16 +0800
Subject: [PATCH] Zero tick plants


diff --git a/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java b/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
index 5e88bd02f5c53124f1aeec3eae727a1f83cc8238..6949c3be76be46197bf281b2cb282f6b03eb80f0 100644
--- a/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -122,8 +122,11 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(world, pos)) {
             world.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+            // Leaves end - zero tick plants
         }
-
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/CactusBlock.java b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
index c045b1cccf0047dbef8c04d5a28d31d53389054f..e891218eda6d832f504753aad16dfeb0a1825b12 100644
--- a/src/main/java/net/minecraft/world/level/block/CactusBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
@@ -47,8 +47,11 @@ public class CactusBlock extends Block {
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(world, pos)) {
             world.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+            // Leaves end - zero tick plants
         }
-
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java b/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 6d0d13e70a82c4db7848e1007f5b6d670dd5acad..8de0c3415b356d6976117c28417faf879f54f927 100644
--- a/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -52,8 +52,11 @@ public class ChorusFlowerBlock extends Block {
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(world, pos)) {
             world.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+            // Leaves end - zero tick plants
         }
-
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java b/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
index 9b424d7661fedf8ee1eb9f3167c62e563f04d4d1..8bbed1f339c5d859dcee79542c07e2c335f7534b 100644
--- a/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -142,4 +142,15 @@ public abstract class GrowingPlantHeadBlock extends GrowingPlantBlock implements
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    // Leaves start - zero tick plants
+    @Override
+    public void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        if (!state.canSurvive(world, pos)) {
+            world.destroyBlock(pos, true);
+        } else if (org.leavesmc.leaves.LeavesConfig.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+        }
+    }
+    // Leaves end - zero tick plants
 }
diff --git a/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java b/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
index 547ea09ed84595286c97c128b3b96f6d387ae25f..11b70f0b1d6e001dec4a90c9692f7b232e7ace05 100644
--- a/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -46,8 +46,11 @@ public class SugarCaneBlock extends Block {
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!state.canSurvive(world, pos)) {
             world.destroyBlock(pos, true);
+            // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.zeroTickPlants) {
+            this.randomTick(state, world, pos, random);
+            // Leaves end - zero tick plants
         }
-
     }
 
     @Override
