From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 16 Aug 2023 13:10:58 +0800
Subject: [PATCH] Fix minecraft hopper not work without player


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
index d81a6874e8b25f098df619f84c359e146c7f64de..6c2045844bb18767c0645c24e6629b59beb25f4e 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -99,6 +99,13 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
         }
     }
 
+    // Leaves start - tick every
+    @Override
+    public void inactiveTick() {
+        this.tick();
+    }
+    // Leaves start - tick every
+
     public boolean suckInItems() {
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             this.immunize();  // Paper
