From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 24 Jul 2023 15:54:18 +0800
Subject: [PATCH] Bow infinity fix


diff --git a/src/main/java/net/minecraft/world/item/BowItem.java b/src/main/java/net/minecraft/world/item/BowItem.java
index 6eb5c0f23d9dc61e69ad5ad493c89602a9dcd4b5..0b8a1c4ba8df45b0a162a468ff58357e048d43ca 100644
--- a/src/main/java/net/minecraft/world/item/BowItem.java
+++ b/src/main/java/net/minecraft/world/item/BowItem.java
@@ -18,6 +18,8 @@ public class BowItem extends ProjectileWeaponItem {
     public static final int MAX_DRAW_DURATION = 20;
     public static final int DEFAULT_RANGE = 15;
 
+    private static final ItemStack EMPTY_ARROW = new ItemStack(Items.ARROW); // Leaves - Bow infinity fix
+
     public BowItem(Item.Properties settings) {
         super(settings);
     }
@@ -81,7 +83,8 @@ public class BowItem extends ProjectileWeaponItem {
     @Override
     public InteractionResultHolder<ItemStack> use(Level world, Player user, InteractionHand hand) {
         ItemStack itemStack = user.getItemInHand(hand);
-        boolean bl = !user.getProjectile(itemStack).isEmpty();
+        boolean bl = !user.getProjectile(itemStack).isEmpty()
+            || (org.leavesmc.leaves.LeavesConfig.bowInfinityFix && net.minecraft.world.item.enchantment.EnchantmentHelper.processAmmoUse((ServerLevel) world, itemStack, EMPTY_ARROW, 1) <= 0); // Leaves - Bow infinity fix
         if (!user.hasInfiniteMaterials() && !bl) {
             return InteractionResultHolder.fail(itemStack);
         } else {
