From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 28 Sep 2023 17:07:02 +0800
Subject: [PATCH] CCE update suppression


diff --git a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
index b7c1050dfafa79f3ed86524ce37cd06651cbbdd9..33da8d67db6926596a5851d25701665741f6d4b8 100644
--- a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -239,17 +239,21 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
 
     @Override
     protected int getAnalogOutputSignal(BlockState state, Level world, BlockPos pos) {
-        // Leaves start - fix update suppression crash
+        // Leaves start - fix update suppression crash - and cce fix
         if (org.leavesmc.leaves.LeavesConfig.modify.updateSuppressionCrashFix) {
             try {
-                return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
+                return org.leavesmc.leaves.LeavesConfig.cceUpdateSuppression ?
+                    AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) world.getBlockEntity(pos)) :
+                    AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
             } catch (ClassCastException ex) {
                 throw new org.leavesmc.leaves.util.UpdateSuppressionException(pos, this);
             }
         } else {
-            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
+            return org.leavesmc.leaves.LeavesConfig.cceUpdateSuppression ?
+                AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) world.getBlockEntity(pos)) :
+                AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
         }
-        // Leaves end - fix update suppression crash
+        // Leaves end - fix update suppression crash - and cce fix
     }
 
     @Override
