From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 14:15:54 +0800
Subject: [PATCH] Crafter 1 gt delay


diff --git a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
index 0e609b1e3abd50b415d8376dc550375a8a0251b6..37954c43b30c64d2d4748e1fd190784ed342e4c5 100644
--- a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
@@ -91,7 +91,7 @@ public class CrafterBlock extends BaseEntityBlock {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
         if (flag1 && !flag2) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !org.leavesmc.leaves.LeavesConfig.modify.oldMC.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
             world.setBlock(pos, (BlockState) state.setValue(CrafterBlock.TRIGGERED, true), 2);
             this.setBlockEntityTriggered(tileentity, true);
         } else if (!flag1 && flag2) {
@@ -157,7 +157,7 @@ public class CrafterBlock extends BaseEntityBlock {
     @Override
     public void setPlacedBy(Level world, BlockPos pos, BlockState state, LivingEntity placer, ItemStack itemStack) {
         if ((Boolean) state.getValue(CrafterBlock.TRIGGERED)) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !org.leavesmc.leaves.LeavesConfig.modify.oldMC.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
         }
 
     }
