From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 12:09:16 +0800
Subject: [PATCH] Copper Bulb 1 gt delay


diff --git a/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java b/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
index 7c6481fc6e84f6477a4d113a222cae51cfa9053a..4e05441e3d78559b372f2a7109aece8c254ba126 100644
--- a/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -32,14 +32,26 @@ public class CopperBulbBlock extends Block {
     @Override
     protected void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify) {
         if (oldState.getBlock() != state.getBlock() && world instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt delay
+            if (!org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                world.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
     @Override
     protected void neighborChanged(BlockState state, Level world, BlockPos pos, Block sourceBlock, @Nullable Orientation wireOrientation, boolean notify) {
         if (world instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt
+            if (!org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                world.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
@@ -56,6 +68,15 @@ public class CopperBulbBlock extends Block {
         }
     }
 
+    // Leaves start - copper bulb 1 gt delay
+    @Override
+    public void tick(BlockState state, ServerLevel world, BlockPos pos, net.minecraft.util.RandomSource random) {
+        if (org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+            checkAndFlip(state, world, pos);
+        }
+    }
+    // Leaves end - copper bulb 1 gt delay
+
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(LIT, POWERED);
