From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 5 Sep 2023 08:49:01 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index e675f1e3e5b6f9e1aa0d928ebb9abe76458edb38..902119c7501e98b2a4a67504216ad2cded138f87 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -145,6 +145,18 @@ public class Spider extends Monster {
         this.entityData.set(Spider.DATA_FLAGS_ID, b0);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(damageSource, causedByPlayer);
+        if (org.leavesmc.leaves.LeavesConfig.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < org.leavesmc.leaves.LeavesConfig.spiderJockeysDropGapples) {
+                this.spawnAtLocation(new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Nullable
     @Override
     public SpawnGroupData finalizeSpawn(ServerLevelAccessor world, DifficultyInstance difficulty, MobSpawnType spawnReason, @Nullable SpawnGroupData entityData) {
