From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 15 Aug 2023 22:40:42 +0800
Subject: [PATCH] Fix vehicle teleport by end gateway


diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 4af071ff2d0a29ad9315e1076436f70f848c89b8..8809755d0e3e6199e2db6c061e53375d17753b70 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -215,7 +215,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
 
                     entity1.setPortalCooldown();
                     ((ServerPlayer) entity1).connection.teleport(teleEvent.getTo());
-                    entity1.teleportPassengers(); // Paper - teleport passengers as well, preventing invisible passengers post teleport.
+                    // entity1.teleportPassengers(); // Paper - teleport passengers as well, preventing invisible passengers post teleport. // Leaves - vanilla
                     TheEndGatewayBlockEntity.triggerCooldown(world, pos, state, blockEntity); // CraftBukkit - call at end of method
                     return;
 
