From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 28 Sep 2023 17:07:02 +0800
Subject: [PATCH] CCE update suppression


diff --git a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
index c6737d4beb558da68b6fa39fd7595b9e453989b1..e70bf963cd960098709262b0a0b34e62562313d8 100644
--- a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -230,17 +230,21 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
 
     @Override
     protected int getAnalogOutputSignal(BlockState state, Level world, BlockPos pos) {
-        // Leaves start - fix update suppression crash
+        // Leaves start - fix update suppression crash - and cce fix
         if (org.leavesmc.leaves.LeavesConfig.updateSuppressionCrashFix) {
             try {
-                return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
+                return org.leavesmc.leaves.LeavesConfig.cceUpdateSuppression ?
+                    AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) world.getBlockEntity(pos)) :
+                    AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
             } catch (ClassCastException ex) {
                 throw new org.leavesmc.leaves.util.UpdateSuppressionException(null, pos);
             }
         } else {
-            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
+            return org.leavesmc.leaves.LeavesConfig.cceUpdateSuppression ?
+                AbstractContainerMenu.getRedstoneSignalFromContainer((net.minecraft.world.Container) world.getBlockEntity(pos)) :
+                AbstractContainerMenu.getRedstoneSignalFromBlockEntity(world.getBlockEntity(pos));
         }
-        // Leaves end - fix update suppression crash
+        // Leaves end - fix update suppression crash - and cce fix
     }
 
     @Override
