From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 15 Aug 2023 22:40:42 +0800
Subject: [PATCH] Fix vehicle teleport by end gateway


diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 0bf810971464d27b3e26b24cb8579d450a1fd37b..8cf9db4bf550cfec402fdfbfcc098cd18ea499d7 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -108,7 +108,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
             if (!list.isEmpty()) {
                 // Paper start - Ensure safe gateway teleport
                 for (Entity entity : list) {
-                    if (entity.canChangeDimensions()) {
+                    if (true || entity.canChangeDimensions()) { // Leaves - be vanilla
                         TheEndGatewayBlockEntity.teleportEntity(world, pos, state, entity, blockEntity);
                         break;
                     }
