From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 15 Aug 2023 22:40:42 +0800
Subject: [PATCH] Fix vehicle teleport by end gateway


diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 1193c4809d551c12329c7f4b380a134609cbe85d..96f323b1ce7ffa0e1a4c58b39da1f01a356310ad 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -109,7 +109,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
             if (!list.isEmpty()) {
                 // Paper start - Ensure safe gateway teleport
                 for (Entity entity : list) {
-                    if (entity.canChangeDimensions()) {
+                    if (true || entity.canChangeDimensions()) { // Leaves - be vanilla
                         TheEndGatewayBlockEntity.teleportEntity(world, pos, state, entity, blockEntity);
                         break;
                     }
