From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 16 Aug 2023 13:10:58 +0800
Subject: [PATCH] Fix minecraft hopper not work without player


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
index 761142374f793a1cd4228936b21a68d7a0458894..97367882f7643b9abe221ccb907e220c05ab588c 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -86,6 +86,13 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
         }
     }
 
+    // Leaves start - tick every
+    @Override
+    public void inactiveTick() {
+        this.tick();
+    }
+    // Leaves start - tick every
+
     public boolean suckInItems() {
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             this.immunize();  // Paper
