From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 14:15:54 +0800
Subject: [PATCH] Crafter 1 gt delay


diff --git a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
index deb5b4fab94618c08a51e432c33e49b1627bd717..cd141680fec975b1a286974178b90d89ed8e34f5 100644
--- a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
@@ -90,7 +90,7 @@ public class CrafterBlock extends BaseEntityBlock {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
         if (flag1 && !flag2) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !org.leavesmc.leaves.LeavesConfig.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
             world.setBlock(pos, (BlockState) state.setValue(CrafterBlock.TRIGGERED, true), 2);
             this.setBlockEntityTriggered(tileentity, true);
         } else if (!flag1 && flag2) {
@@ -156,7 +156,7 @@ public class CrafterBlock extends BaseEntityBlock {
     @Override
     public void setPlacedBy(Level world, BlockPos pos, BlockState state, LivingEntity placer, ItemStack itemStack) {
         if ((Boolean) state.getValue(CrafterBlock.TRIGGERED)) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !org.leavesmc.leaves.LeavesConfig.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
         }
 
     }
