From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 14:15:54 +0800
Subject: [PATCH] Crafter 1 gt delay


diff --git a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
index 391a6af36d4c27a04000b31583297a25b89125d3..b94c5ba94f1e08bc173d8e35cffb67ba736aee18 100644
--- a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
@@ -80,7 +80,7 @@ public class CrafterBlock extends BaseEntityBlock {
         boolean bl2 = state.getValue(TRIGGERED);
         BlockEntity blockEntity = world.getBlockEntity(pos);
         if (bl && !bl2) {
-            world.scheduleTick(pos, this, 4);
+            world.scheduleTick(pos, this, !org.leavesmc.leaves.LeavesConfig.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
             world.setBlock(pos, state.setValue(TRIGGERED, Boolean.valueOf(true)), 2);
             this.setBlockEntityTriggered(blockEntity, true);
         } else if (!bl && bl2) {
@@ -130,7 +130,7 @@ public class CrafterBlock extends BaseEntityBlock {
     @Override
     public void setPlacedBy(Level world, BlockPos pos, BlockState state, LivingEntity placer, ItemStack itemStack) {
         if (state.getValue(TRIGGERED)) {
-            world.scheduleTick(pos, this, 4);
+            world.scheduleTick(pos, this, !org.leavesmc.leaves.LeavesConfig.crafter1gt ? 4 : 1); // Leaves - crafter 1 gt delay
         }
     }
 
