From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 15 Aug 2023 22:40:42 +0800
Subject: [PATCH] Fix vehicle teleport by end gateway


diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index b0acc2f37a8ebc5d93dcf33a0b89bd9c4444fccf..c06092aa35d44ba98b9680dc72eeba09f6d07822 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -108,7 +108,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
             if (!list.isEmpty()) {
                 // Paper start
                 for (Entity entity : list) {
-                    if (entity.canChangeDimensions()) {
+                    if (true || entity.canChangeDimensions()) { // Leaves - be vanilla
                         TheEndGatewayBlockEntity.teleportEntity(world, pos, state, entity, blockEntity);
                         break;
                     }
