From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 3 Aug 2023 20:36:38 +0800
Subject: [PATCH] Photographer


diff --git a/src/main/java/top/leavesmc/leaves/replay/Photographer.java b/src/main/java/top/leavesmc/leaves/replay/Photographer.java
new file mode 100644
index 0000000000000000000000000000000000000000..1cf7022d56626779bb9fe8174beeb1738499acfa
--- /dev/null
+++ b/src/main/java/top/leavesmc/leaves/replay/Photographer.java
@@ -0,0 +1,4 @@
+package top.leavesmc.leaves.replay;
+
+public class Photographer {
+}
