From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 5 Sep 2023 08:49:01 +0800
Subject: [PATCH] Spider jockeys drop gapples


diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index 71b5a9c97a13f703073c0122742ff9e8a0e49df2..15b1cdc7e2bf53998187c12ffff17c6653098890 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -154,6 +154,18 @@ public class Spider extends Monster {
         this.entityData.set(Spider.DATA_FLAGS_ID, b0);
     }
 
+    // Leaves start - spider jockeys drop gapples
+    @Override
+    protected void dropFromLootTable(DamageSource damageSource, boolean causedByPlayer) {
+        super.dropFromLootTable(damageSource, causedByPlayer);
+        if (top.leavesmc.leaves.LeavesConfig.spiderJockeysDropGapples > 0.0D) {
+            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < top.leavesmc.leaves.LeavesConfig.spiderJockeysDropGapples) {
+                this.spawnAtLocation(new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
+            }
+        }
+    }
+    // Leaves end - spider jockeys drop gapples
+
     @Nullable
     @Override
     public SpawnGroupData finalizeSpawn(ServerLevelAccessor world, DifficultyInstance difficulty, MobSpawnType spawnReason, @Nullable SpawnGroupData entityData, @Nullable CompoundTag entityNbt) {
