From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 22 Jul 2024 18:30:42 +0800
Subject: [PATCH] Can disable LivingEntity aiStep alive check


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 55cce94ebee920a64fdd80248fe10c231c8023d0..4358734f20a597dbb92e346c5ad9c3d4dce0b6a7 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3305,7 +3305,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
             }
         }
 
-        if (!this.isRemoved()) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.disableLivingEntityAiStepAliveCheck || !this.isRemoved()) { // Leaves - can disable check
             this.aiStep();
         }
 
