From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 16 Aug 2023 13:10:58 +0800
Subject: [PATCH] Fix minecraft hopper not work without player


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
index fc35cfc9d045f3e5b6a50af1d0ba83b6e322091f..48ae92dc7c4ec06d73fd1616d24ecd9c2cbda49e 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -88,6 +88,13 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
 
     }
 
+    // Leaves start - tick every
+    @Override
+    public void inactiveTick() {
+        this.tick();
+    }
+    // Leaves start - tick every
+
     public boolean suckInItems() {
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             this.immunize();  // Paper
