From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 16 Feb 2024 19:50:03 +0800
Subject: [PATCH] Fix falling block's block location


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 22f1ccd7f0055cc6d68803111666e2df59a54aa4..99ad282dcc026b45cbe5b4af22db3da02e837d10 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4751,6 +4751,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             int j = Mth.floor(y);
             int k = Mth.floor(z);
 
+            // Leaves start - fix falling block location
+            if (this instanceof net.minecraft.world.entity.item.FallingBlockEntity) {
+                if (y < 0.0 && y + 1e-10 > 0.0) {
+                    j = 0;
+                }
+            }
+            // Leaves end - fix falling block location
+
             if (i != this.blockPosition.getX() || j != this.blockPosition.getY() || k != this.blockPosition.getZ()) {
                 this.blockPosition = new BlockPos(i, j, k);
                 this.inBlockState = null;
