From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 12 Dec 2023 12:09:16 +0800
Subject: [PATCH] Copper Bulb 1 gt delay


diff --git a/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java b/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
index b4250059a9068124b76268913f675babac379670..e6cfa8c1ce45c5e4b647ac6faceb838961e69dad 100644
--- a/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -30,14 +30,26 @@ public class CopperBulbBlock extends Block {
     @Override
     protected void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify) {
         if (oldState.getBlock() != state.getBlock() && world instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt delay
+            if (!org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                world.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
     @Override
     protected void neighborChanged(BlockState state, Level world, BlockPos pos, Block sourceBlock, BlockPos sourcePos, boolean notify) {
         if (world instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // Leaves start - copper bulb 1 gt
+            if (!org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                world.scheduleTick(pos, this, 1);
+            }
+            // Leaves end - copper bulb 1 gt delay
         }
     }
 
@@ -54,6 +66,15 @@ public class CopperBulbBlock extends Block {
         }
     }
 
+    // Leaves start - copper bulb 1 gt delay
+    @Override
+    public void tick(BlockState state, ServerLevel world, BlockPos pos, net.minecraft.util.RandomSource random) {
+        if (org.leavesmc.leaves.LeavesConfig.copperBulb1gt) {
+            checkAndFlip(state, world, pos);
+        }
+    }
+    // Leaves end - copper bulb 1 gt delay
+
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(LIT, POWERED);
