From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 30 Oct 2023 10:43:44 +0800
Subject: [PATCH] Make Item tick vanilla


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index ffa9fe429ed773db25a5e1020283b12597a69070..feb3fb984f6e6ed70992a038649249a9d3a9c938 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -252,6 +252,9 @@ public class ItemEntity extends Entity implements TraceableEntity {
     // Spigot start - copied from above
     @Override
     public void inactiveTick() {
+        // Leaves start - vanilla
+        this.tick();
+        /*
         // CraftBukkit start - Use wall time for pickup and despawn timers
         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
         if (this.pickupDelay != 32767) this.pickupDelay -= elapsedTicks;
@@ -269,6 +272,8 @@ public class ItemEntity extends Entity implements TraceableEntity {
             // CraftBukkit end
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
+         */
+        // Leaves end - vanilla
     }
     // Spigot end
 
