From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Wed, 14 Aug 2024 01:48:14 +0800
Subject: [PATCH] Old BlockEntity behaviour


diff --git a/src/main/java/net/minecraft/world/level/block/ChiseledBookShelfBlock.java b/src/main/java/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
index 5a95ecd7e94652d2af3eb42ee9cf61913fb95a69..cadd3362e9993715cd47e5e5c6d7a3eb60d11d33 100644
--- a/src/main/java/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
@@ -194,6 +194,13 @@ public class ChiseledBookShelfBlock extends BaseEntityBlock {
                     }
 
                     chiseledBookShelfBlockEntity.clearContent();
+                    // Leaves start - behaviour 1.21.1-
+                    if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) {
+                        bl = false;
+                        world.updateNeighbourForOutputSignal(pos, this);
+                        break label32;
+                    }
+                    // Leaves end - behaviour 1.21.1-
                     bl = true;
                     break label32;
                 }
diff --git a/src/main/java/net/minecraft/world/level/block/LecternBlock.java b/src/main/java/net/minecraft/world/level/block/LecternBlock.java
index 70f2e6278e2d970245ca5b46fbd9ffae4727b47b..cb1ef444efaaff0579154537db151d16efae357e 100644
--- a/src/main/java/net/minecraft/world/level/block/LecternBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LecternBlock.java
@@ -229,10 +229,11 @@ public class LecternBlock extends BaseEntityBlock {
                 this.popBook(state, world, pos);
             }
 
-            super.onRemove(state, world, pos, newState, moved);
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) super.onRemove(state, world, pos, newState, moved); // Leaves - behaviour 1.21.1-
             if ((Boolean) state.getValue(LecternBlock.POWERED)) {
                 LecternBlock.updateBelow(world, pos, state);
             }
+            if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) super.onRemove(state, world, pos, newState, moved); // Leaves - behaviour 1.21.1-
 
         }
     }
diff --git a/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java b/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
index 0ed449a188d98f87dbddd2d76009fed02a29ed25..e812d158cf56d4d1d67ea335433a23c7be0e4292 100644
--- a/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -151,10 +151,11 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
     @Override
     protected void onRemove(BlockState state, Level world, BlockPos pos, BlockState newState, boolean moved) {
         if (!state.is(newState.getBlock())) {
-            super.onRemove(state, world, pos, newState, moved);
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) super.onRemove(state, world, pos, newState, moved); // Leaves - behaviour 1.21.1-
             if (SculkSensorBlock.getPhase(state) == SculkSensorPhase.ACTIVE) {
                 SculkSensorBlock.updateNeighbours(world, pos, state);
             }
+            if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) super.onRemove(state, world, pos, newState, moved); // Leaves - behaviour 1.21.1-
 
         }
     }
diff --git a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
index b01918fc97926a3182c21145bb7411e7bc409d35..0646a861cc04ad29ba850f34cd6a629228f2c640 100644
--- a/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -186,6 +186,7 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
             if (blockEntity instanceof ShulkerBoxBlockEntity) {
                 world.updateNeighbourForOutputSignal(pos, state.getBlock());
             }
+            if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour) super.onRemove(state, world, pos, newState, moved); // Leaves - behaviour 1.21.1-
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index 1f929b467a0ece3143af58a657cf5983c07a8d51..e8ab0901393844c9d2ea3595a28306bfd8bd40fb 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -67,7 +67,7 @@ public abstract class BlockEntity {
     }
 
     public boolean isValidBlockState(BlockState state) {
-        return this.type.isValid(state);
+        return org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldBlockEntityBehaviour || this.type.isValid(state); // Leaves - behaviour 1.21.1-
     }
 
     public static BlockPos getPosFromTag(CompoundTag nbt) {
