From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 4 Sep 2023 22:09:10 +0800
Subject: [PATCH] Loot world random


diff --git a/src/main/java/net/minecraft/world/level/storage/loot/LootContext.java b/src/main/java/net/minecraft/world/level/storage/loot/LootContext.java
index 17df8b341c65ca24fd2a18424f221918d0a7a111..d5eea5d36b9fac6001cca262e9a7c397852e4bdb 100644
--- a/src/main/java/net/minecraft/world/level/storage/loot/LootContext.java
+++ b/src/main/java/net/minecraft/world/level/storage/loot/LootContext.java
@@ -113,6 +113,11 @@ public class LootContext {
             ServerLevel serverLevel = this.getLevel();
             MinecraftServer minecraftServer = serverLevel.getServer();
             RandomSource randomSource;
+            // Leaves start - world random
+            if (top.leavesmc.leaves.LeavesConfig.lootWorldRandom) {
+                return new LootContext(this.params, serverLevel.getRandom(), minecraftServer.getLootData());
+            }
+            // Leaves end - world random
             if (this.random != null) {
                 randomSource = this.random;
             } else if (randomSequenceId != null) {
