From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 1 Jul 2024 22:09:33 +0800
Subject: [PATCH] Disable end gateway portal entity ticking


diff --git a/src/main/java/net/minecraft/world/level/block/EndGatewayBlock.java b/src/main/java/net/minecraft/world/level/block/EndGatewayBlock.java
index 3087c60589de8fc38e12d322e58886ba76749507..d16f1defac4030248ad24b4c4a38ca1717908cd4 100644
--- a/src/main/java/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -128,7 +128,7 @@ public class EndGatewayBlock extends BaseEntityBlock implements Portal {
             }
             // Leaves end - force void trade
 
-            return vec3d != null ? new DimensionTransition(world, vec3d, EndGatewayBlock.calculateExitMovement(entity), entity.getYRot(), entity.getXRot(), DimensionTransition.PLACE_PORTAL_TICKET, PlayerTeleportEvent.TeleportCause.END_GATEWAY) : null; // CraftBukkit
+            return vec3d != null ? new DimensionTransition(world, vec3d, EndGatewayBlock.calculateExitMovement(entity), entity.getYRot(), entity.getXRot(), !org.leavesmc.leaves.LeavesConfig.disableGatewayPortalEntityTicking ? DimensionTransition.PLACE_PORTAL_TICKET : DimensionTransition.DO_NOTHING, PlayerTeleportEvent.TeleportCause.END_GATEWAY) : null; // CraftBukkit // Leaves - disable-gateway-portal-entity-ticking
         } else {
             return null;
         }
