From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Wed, 26 Jun 2024 17:59:56 +0800
Subject: [PATCH] Allow grindstone overstacking


diff --git a/src/main/java/net/minecraft/world/inventory/GrindstoneMenu.java b/src/main/java/net/minecraft/world/inventory/GrindstoneMenu.java
index 1678f6c8b2c7db761783e53043169bf12bc2cb29..13cfda7c065e0332133a1e1b60b454ae44e87d35 100644
--- a/src/main/java/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -202,7 +202,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
             byte b0 = 1;
 
             if (!firstInput.isDamageableItem()) {
-                if (firstInput.getMaxStackSize() < 2 || !ItemStack.matches(firstInput, secondInput)) {
+                if (!org.leavesmc.leaves.LeavesConfig.allowGrindstoneOverstacking && firstInput.getMaxStackSize() < 2 || !ItemStack.matches(firstInput, secondInput)) { // Leaves - allowGrindstoneOverstaking
                     return ItemStack.EMPTY;
                 }
 
@@ -286,7 +286,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
             ItemStack itemstack3 = this.repairSlots.getItem(1);
 
             if (slot == 2) {
-                if (!this.moveItemStackTo(itemstack1, 3, 39, true)) {
+                if (!this.moveItemStackTo(itemstack1, 3, 39, true, org.leavesmc.leaves.LeavesConfig.allowGrindstoneOverstacking)) { // Leaves - allowGrindstoneOverstacking: Disable stack check
                     return ItemStack.EMPTY;
                 }
 
