From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Wed, 26 Jun 2024 18:15:57 +0800
Subject: [PATCH] MCOld-MC-67


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b450c19fbb6cc7266db4ddc4bc86546ca2054d5a..69d2ee3bd4e89968da219f27231af241c88b3600 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3753,7 +3753,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public boolean canChangeDimensions(Level from, Level to) {
-        return this.isAlive() && this.valid; // Paper - Fix item duplication and teleport issues
+        return !(this.isPassenger() || this.isVehicle() && !org.leavesmc.leaves.LeavesConfig.allowEntityPortalWithPassenger) && this.isAlive() && this.valid; // Paper - Fix item duplication and teleport issues // Leaves - allowEntityPortalWithPassenger
     }
 
     public float getBlockExplosionResistance(Explosion explosion, BlockGetter world, BlockPos pos, BlockState blockState, FluidState fluidState, float max) {
diff --git a/src/main/java/org/leavesmc/leaves/LeavesConfig.java b/src/main/java/org/leavesmc/leaves/LeavesConfig.java
index 6ad969d47898199bfa2d73ecda3d26d3c4719520..799fa0e780431cd690b86f3340850dd85a2e4884 100644
--- a/src/main/java/org/leavesmc/leaves/LeavesConfig.java
+++ b/src/main/java/org/leavesmc/leaves/LeavesConfig.java
@@ -921,5 +921,7 @@ public final class LeavesConfig {
     @GlobalConfig(name = "allow-grindstone-overstacking", category = "fix")
     public static boolean allowGrindstoneOverstacking = false;
 
+    @GlobalConfig(name = "allow-entity-portal-with-passenger", category = "fix")
+    public static boolean allowEntityPortalWithPassenger = true;
     // Leaves end - fix
 }
