From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huanli233 <392352840@qq.com>
Date: Sun, 27 Oct 2024 15:57:37 +0800
Subject: [PATCH] Disable vault blacklist


diff --git a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index 30544783d959fbdf4b6fd53c10bd3806e718a066..4319e2108b59aa78ed00d88859ba742e663edbd3 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -57,11 +57,12 @@ public class VaultServerData {
     }
 
     boolean hasRewardedPlayer(Player player) {
-        return this.rewardedPlayers.contains(player.getUUID());
+        return !org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist && this.rewardedPlayers.contains(player.getUUID()); // Leaves - disable vault blacklist
     }
 
     @VisibleForTesting
     public void addToRewardedPlayers(Player player) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist) return; // Leaves - disable vault blacklist
         this.rewardedPlayers.add(player.getUUID());
         if (this.rewardedPlayers.size() > 128) {
             Iterator<UUID> iterator = this.rewardedPlayers.iterator();
diff --git a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index f09ce717869741bb2f027ab1216fad9d86425683..a32c8920ef6c1ab4b6ec7a68b9ed10c37afdfbf3 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -68,8 +68,8 @@ public class VaultSharedData {
         Set<UUID> set = config.playerDetector()
             .detect(world, config.entitySelector(), pos, radius, false)
             .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+            .filter(uuid -> org.leavesmc.leaves.LeavesConfig.modify.disableVaultBlacklist || !serverData.getRewardedPlayers().contains(uuid))
+            .collect(Collectors.toSet()); // Leaves - disable vault blacklist
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
