From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 23 Jan 2024 01:08:41 +0800
Subject: [PATCH] Renewable coral


diff --git a/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java b/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
index 7701683000132479d8c8eaa77ac3f93e41a960be..74e87b040ceb46836a0eb4c053d5f8ccf78fabfe 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
@@ -13,7 +13,7 @@ import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
 
-public class CoralFanBlock extends BaseCoralFanBlock {
+public class CoralFanBlock extends BaseCoralFanBlock implements top.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
 
     public static final MapCodec<CoralFanBlock> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
         return instance.group(CoralBlock.DEAD_CORAL_FIELD.forGetter((blockcoralfan) -> {
@@ -63,4 +63,11 @@ public class CoralFanBlock extends BaseCoralFanBlock {
             return super.updateShape(state, direction, neighborState, world, pos, neighborPos);
         }
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return top.leavesmc.leaves.LeavesConfig.renewableCoral == top.leavesmc.leaves.LeavesConfig.RenewableCoralType.EXPANDED;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java b/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
index 4c4aa2a63be84ff614a3dc0db2864266755545a2..2aa1a65c8c8864ffd52eaf6e3b0a18c242f97de0 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -16,7 +16,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
+public class CoralPlantBlock extends BaseCoralPlantTypeBlock implements top.leavesmc.leaves.util.FertilizableCoral { // Leaves - renewable coral
 
     public static final MapCodec<CoralPlantBlock> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
         return instance.group(CoralBlock.DEAD_CORAL_FIELD.forGetter((blockcoralplant) -> {
@@ -73,4 +73,12 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
     protected VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
         return CoralPlantBlock.SHAPE;
     }
+
+    // Leaves start - renewable coral
+    @Override
+    public boolean isEnabled() {
+        return top.leavesmc.leaves.LeavesConfig.renewableCoral == top.leavesmc.leaves.LeavesConfig.RenewableCoralType.EXPANDED
+            || top.leavesmc.leaves.LeavesConfig.renewableCoral == top.leavesmc.leaves.LeavesConfig.RenewableCoralType.TRUE;
+    }
+    // Leaves end - renewable coral
 }
diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/CoralFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/CoralFeature.java
index 7bc5ff8eb1174834dcc27363af4a5cef19017b3d..e2fc5f9c5cc4ce6b5c1a0d0c43b753c1ca87bf72 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/CoralFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/CoralFeature.java
@@ -31,6 +31,12 @@ public abstract class CoralFeature extends Feature<NoneFeatureConfiguration> {
         return !optional.isEmpty() && this.placeFeature(worldGenLevel, randomSource, blockPos, optional.get().defaultBlockState());
     }
 
+    // Leaves start - renewable coral
+    public boolean growSpecific(LevelAccessor world, RandomSource random, BlockPos pos, BlockState state) {
+        return placeFeature(world, random, pos, state);
+    }
+    // Leaves end - renewable coral
+
     protected abstract boolean placeFeature(LevelAccessor world, RandomSource random, BlockPos pos, BlockState state);
 
     protected boolean placeCoralBlock(LevelAccessor world, RandomSource random, BlockPos pos, BlockState state) {
