From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Sat, 20 Jan 2024 02:50:56 +0800
Subject: [PATCH] Placing locked hopper no longer send NC updates


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index ce85d69027cbfb06666c9242576d3abbd7cc2831..bf2cf885552249765f005086750a8457c0949c49 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -51,6 +51,7 @@ import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.HopperBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.TickingBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
@@ -1029,7 +1030,9 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
             }
 
             if ((i & 1) != 0) {
-                this.blockUpdated(blockposition, iblockdata1.getBlock());
+                if (!(iblockdata.getBlock() instanceof HopperBlock) || iblockdata.getValue(HopperBlock.ENABLED)) { // Leaves - Placing locked hopper doesn't send NC updates.
+                    this.blockUpdated(blockposition, iblockdata1.getBlock());
+                } // Leaves
                 if (!this.isClientSide && iblockdata.hasAnalogOutputSignal()) {
                     this.updateNeighbourForOutputSignal(blockposition, newBlock.getBlock());
                 }
