From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 16 Aug 2024 02:02:06 +0800
Subject: [PATCH] Fix Spigot EndPlatform destroy


diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
index 0bc659a8427b89b5e3211220c55b52eec6a20494..3746721baa1897111c2d30c9e0f388dddca099da 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
@@ -44,7 +44,13 @@ public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
                     // CraftBukkit start
                     if (!blockList.getBlockState(blockposition_mutableblockposition1).is(block)) {
                         if (flag) {
-                            blockList.destroyBlock(blockposition_mutableblockposition1, true, (Entity) null);
+                            // Leaves start - spigot!
+                            if (!org.leavesmc.leaves.LeavesConfig.spigotEndPlatformDestroy) {
+                                blockList.destroyBlock(blockposition_mutableblockposition1, true, (Entity) null);
+                            } else {
+                                worldaccess.destroyBlock(blockposition_mutableblockposition1, true, (Entity) null);
+                            }
+                            // Leaves end - spigot!
                         }
 
                         blockList.setBlock(blockposition_mutableblockposition1, block.defaultBlockState(), 3);
