From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 2 Feb 2024 13:18:22 +0800
Subject: [PATCH] Force minecraft command


diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
index 4b1ac1fe7ea07f419ae2818251900e7ba434ee16..fefe60c252148cd563c292581367fa6a277d88a5 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
@@ -14,4 +14,18 @@ public class CraftCommandMap extends SimpleCommandMap {
     public Map<String, Command> getKnownCommands() {
         return this.knownCommands;
     }
+
+    // Leaves start - force minecraft command
+    @Override
+    public Command getCommand(String name) {
+        Command target = super.getCommand(name);
+        if (top.leavesmc.leaves.LeavesConfig.forceMinecraftCommand && !(target instanceof org.bukkit.craftbukkit.command.VanillaCommandWrapper)) {
+            Command minecraftTarget = super.getCommand("minecraft:" + name);
+            if (minecraftTarget != null) {
+                return minecraftTarget;
+            }
+        }
+        return target;
+    }
+    // Leaves end - force minecraft command
 }
