From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7studio@foxmail.com>
Date: Sat, 11 May 2024 00:41:58 +0800
Subject: [PATCH] Pass SignedMessageChain Build


diff --git a/src/main/java/net/minecraft/network/chat/SignedMessageChain.java b/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
index 9756e8b560281796081a6ee157f9d92240bd2d56..8842f824d2f7a8297c67feaa5d27c7f5dcc9aedc 100644
--- a/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
+++ b/src/main/java/net/minecraft/network/chat/SignedMessageChain.java
@@ -66,6 +66,7 @@ public class SignedMessageChain {
                         }
                     }
                 }
+                return null; // Leaves - pass build
             }
 
             @Override
