From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: serverbread <serverbread@163.com>
Date: Fri, 5 Apr 2024 02:36:11 +0800
Subject: [PATCH] Fixed /bot command.


diff --git a/src/main/java/top/leavesmc/leaves/bot/BotCommand.java b/src/main/java/top/leavesmc/leaves/bot/BotCommand.java
index fd8a1306f3179d0729a4a543405ac210acefe8d9..dcdefc6f2d9a4a6de80e9de5940f9f01205538f1 100644
--- a/src/main/java/top/leavesmc/leaves/bot/BotCommand.java
+++ b/src/main/java/top/leavesmc/leaves/bot/BotCommand.java
@@ -168,7 +168,7 @@ public class BotCommand extends Command {
             return false;
         }
 
-        if (Bukkit.getPlayer(name) != null || ServerBot.getBot(name) != null) {
+        if (Bukkit.getPlayer(name) != null && Bukkit.getPlayer(name).equals(name) || ServerBot.getBot(name) != null) {
             sender.sendMessage(ChatColor.RED + "This player is in server");
             return false;
         }
