From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: silvery0113 <silvery0113@foxmail.com>
Date: Fri, 5 Apr 2024 12:32:17 +0800
Subject: [PATCH] fix next steps


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 7d1cd7b98e44e0b6c0d1f29df25ad45c0946548d..e29667480aec47ebf729321747e43bacf45b43f2 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1155,15 +1155,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             long tickSection = Util.getNanos();
             long currentTime;
             // Paper end - further improve server tick loop
-            // Paper start - Add onboarding message for initial server start
+            // Leaves start - Add onboarding message for initial server start
             if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
                 LOGGER.info("*************************************************************************************");
                 LOGGER.info("This is the first time you're starting this server.");
                 LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
-                LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
+                LOGGER.info("View this and more helpful information here: https://docs.leavesmc.org/leaves/guides/next-steps");
                 LOGGER.info("*************************************************************************************");
             }
-            // Paper end - Add onboarding message for initial server start
+            // Leaves end - Add onboarding message for initial server start
 
             while (this.running) {
                 // Paper start - rewrite chunk system
diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 7704a5951ac3d02020ed0f40d76500dd6ba005af..7814f5f5dccccabe0a79fa2fe94da29208905aab 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -116,9 +116,9 @@ public class MinecraftServerGui extends JComponent {
         return jpanel;
     }
 
-    // Paper start - Add onboarding message for initial server start
+    // Leaves start - Add onboarding message for initial server start
     private JComponent buildOnboardingPanel() {
-        String onboardingLink = "https://docs.papermc.io/paper/next-steps";
+        String onboardingLink = "https://docs.leavesmc.org/leaves/guides/next-steps";
         JPanel jPanel = new JPanel();
 
         javax.swing.JLabel jLabel = new javax.swing.JLabel("If you need help setting up your server you can visit:");
@@ -147,7 +147,7 @@ public class MinecraftServerGui extends JComponent {
 
         return jPanel;
     }
-    // Paper end - Add onboarding message for initial server start
+    // Leaves end - Add onboarding message for initial server start
 
     private JComponent buildPlayerPanel() {
         JList<?> jlist = new PlayerListComponent(this.server);
