Subject: [PATCH] fix next steps
---
Index: src/main/java/net/minecraft/server/MinecraftServer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
--- a/src/main/java/net/minecraft/server/MinecraftServer.java	(revision 6488c8038d4ae7dfbab02f8e963fc22148314e64)
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java	(revision 9af8bd1bca85f336d1769ff13ffe48c61a34e727)
@@ -1155,15 +1155,15 @@
             long tickSection = Util.getNanos();
             long currentTime;
             // Paper end - further improve server tick loop
-            // Paper start - Add onboarding message for initial server start
+            // Leaves start - Add onboarding message for initial server start
             if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
                 LOGGER.info("*************************************************************************************");
                 LOGGER.info("This is the first time you're starting this server.");
                 LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
-                LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
+                LOGGER.info("View this and more helpful information here: https://docs.leavesmc.org/leaves/guides/next-steps");
                 LOGGER.info("*************************************************************************************");
             }
-            // Paper end - Add onboarding message for initial server start
+            // Leaves end - Add onboarding message for initial server start
 
             while (this.running) {
                 // Paper start - rewrite chunk system
Index: src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java	(revision 6488c8038d4ae7dfbab02f8e963fc22148314e64)
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java	(revision 9af8bd1bca85f336d1769ff13ffe48c61a34e727)
@@ -116,9 +116,9 @@
         return jpanel;
     }
 
-    // Paper start - Add onboarding message for initial server start
+    // Leaves start - Add onboarding message for initial server start
     private JComponent buildOnboardingPanel() {
-        String onboardingLink = "https://docs.papermc.io/paper/next-steps";
+        String onboardingLink = "https://docs.leavesmc.org/leaves/guides/next-steps";
         JPanel jPanel = new JPanel();
 
         javax.swing.JLabel jLabel = new javax.swing.JLabel("If you need help setting up your server you can visit:");
@@ -147,7 +147,7 @@
 
         return jPanel;
     }
-    // Paper end - Add onboarding message for initial server start
+    // Leaves end - Add onboarding message for initial server start
 
     private JComponent buildPlayerPanel() {
         JList<?> jlist = new PlayerListComponent(this.server);
